// @sapappgyver/appgyver-formulas/functions/schema/SCHEMA_TYPE.js
const execute = require('./SCHEMA_TYPE.runtime');

module.exports = {
  name: 'SCHEMA_TYPE',
  category: 'Schema',
  title: 'Get schema type',
  description: 'Returns a special type keyword text describing what kind of value the schema accepts. The most typical keywords are these:\n'
    + '\n'
    + '- `string`: text value, e.g. `"Hello!"`\n'
    + '- `number`: number value, e.g. `123`\n'
    + '- `boolean`: either `true` or `false`\n'
    + '- `array`: list of values, e.g. `[1, 2, 3]` or `["foo", "bar"]`\n'
    + '- `object`: object value, e.g. `{name: "John", year: 1987}`\n'
    + '\n'
    + 'Returns `null` if the schema accepts more than one type, or if the type is unknown.',
  parameters: [{
    type: 'schema',
    title: 'Schema',
    description: 'Schema from which to get the type',
  }],
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    { expression: 'SCHEMA_TYPE({ type: "number" })', result: 'number' },
    { expression: 'SCHEMA_TYPE({ type: "string" })', result: 'string' },
  ],
};
