// @sapappgyver/appgyver-formulas/functions/schema/SCHEMA_LIST_ITEM.js
const execute = require('./SCHEMA_LIST_ITEM.runtime');

module.exports = {
  name: 'SCHEMA_LIST_ITEM',
  category: 'Schema',
  title: 'Get list item schema',
  description: 'Returns a schema of the items in the given schema representing a list.\n\n'
    + 'For example, if you have a schema representing a list of strings, you can use this function to extract the type of the items.\n\n'
    + 'Returns `null` if the parameter is not a schema representing an list, or if the schema is otherwise unknown.',
  parameters: [{
    type: 'schema',
    title: 'Schema',
    description: 'Schema representing a list from which to get the item type',
    schema: {
      type: 'array',
    },
  }],
  analyze: () => ({ schema: { type: 'schema' } }),
  execute,
  examples: [
    {
      context: {
        tagsSchema: {
          type: 'array',
          items: {
            type: 'string',
          },
        },
      },
      expression: 'SCHEMA_LIST_ITEM(tagsSchema)',
      result: { type: 'string' },
    },
  ],
};
