// ts-patch/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setOptions = exports.patch = exports.uninstall = exports.install = exports.getTSModule = exports.getTSPackage = void 0;
var file_utils_1 = require("./lib/file-utils");
Object.defineProperty(exports, "getTSPackage", { enumerable: true, get: function () { return file_utils_1.getTSPackage; } });
Object.defineProperty(exports, "getTSModule", { enumerable: true, get: function () { return file_utils_1.getTSModule; } });
var actions_1 = require("./lib/actions");
Object.defineProperty(exports, "install", { enumerable: true, get: function () { return actions_1.install; } });
Object.defineProperty(exports, "uninstall", { enumerable: true, get: function () { return actions_1.uninstall; } });
Object.defineProperty(exports, "patch", { enumerable: true, get: function () { return actions_1.patch; } });
Object.defineProperty(exports, "setOptions", { enumerable: true, get: function () { return actions_1.setOptions; } });
__exportStar(require("./plugin-types"), exports);
