// @sapappgyver/appgyver-formulas/functions/text/APPEND_URL_PARAMETERS.js
const execute = require('./APPEND_URL_PARAMETERS.runtime');

module.exports = {
  name: 'APPEND_URL_PARAMETERS',
  category: 'Text',
  title: 'Append parameters to URL',
  description: 'Appends the given URL parameters (given as a dictionary object) to the URL text. Can result in multiple URL parameters of the same name.',
  parameters: [{
    type: 'string',
    format: 'url',
    title: 'Base URL',
    description: 'Base URL text',
  }, {
    type: 'object',
    title: 'New parameters object',
    description: 'Dictionary for parameters',
  }],
  analyze: () => ({ schema: { type: 'string', format: 'url' } }),
  execute,
  examples: [
    {
      expression: 'APPEND_URL_PARAMETERS("http://www.example.com", { test1: "abc", test2: 123})',
      result: 'http://www.example.com/?test1=abc&test2=123',
    },
    {
      expression: 'APPEND_URL_PARAMETERS("http://www.example.com/?abc=123&test=abc", { test1: "abc", test2: 123})',
      result: 'http://www.example.com/?abc=123&test=abc&test1=abc&test2=123',
    },
    {
      expression: 'APPEND_URL_PARAMETERS("http://www.example.com/?abc=123&test=abc", {})',
      result: 'http://www.example.com/?abc=123&test=abc',
    },
  ],
};
