// @sapappgyver/appgyver-formulas/functions/text/FORMAT_LOCALIZED_DECIMAL.js
const execute = require('./FORMAT_LOCALIZED_DECIMAL.runtime');
const { supportedLocales } = require('../../utils/locale');

module.exports = {
  name: 'FORMAT_LOCALIZED_DECIMAL',
  category: 'Text',
  title: 'Returns localized number representation',
  description: 'Returns a text with a language-/locale-sensitive representation of the given decimal number.',
  parameters: [{
    type: 'number',
    title: 'Number',
    description: 'Number to localize',
  }, {
    type: 'string',
    title: 'Locale',
    description: 'A text with a BCP 47 language tag. '
      + 'Examples: "en", "de-DE", "de-AT", "en-IN" , "fi-FI"',
    enum: supportedLocales,
  }, {
    type: 'number',
    title: 'minDecimalCount',
    description: 'Minimal decimal count',
  }, {
    type: 'number',
    title: 'maxDecimalCount',
    default: null,
    description: 'Maximum decimal count',
  }],
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    { expression: 'FORMAT_LOCALIZED_DECIMAL(123450.673, "ru-RU", 2)', result: '123 450,67' },
    { expression: 'FORMAT_LOCALIZED_DECIMAL(123450.673, "en-IN", 4)', result: '1,23,450.6730' },
    { expression: 'FORMAT_LOCALIZED_DECIMAL(123450.673, "de-CH", 3)', result: '123\'450.673' },
    { expression: 'FORMAT_LOCALIZED_DECIMAL(1234.5678, "fi-FI", 3, 8)', result: '1 234,5678' },
  ],
};
