// @sapappgyver/appgyver-formulas/functions/text/QUERY_PARAMETERS_OBJECT.js
const execute = require('./QUERY_PARAMETERS_OBJECT.runtime');

module.exports = {
  name: 'QUERY_PARAMETERS_OBJECT',
  category: 'Text',
  title: 'Convert query parameters in object',
  description: `Given an URL, return its query parameters as an object with schema:

\`{ param1Key: param1Value, param2Key: param2Value }\`

Numbers and booleans are parsed to their correct types.`,
  parameters: [{
    title: 'URL',
    description: 'URL to convert',
    type: 'string',
  }],
  analyze: () => ({
    schema: {
      type: 'object',
      properties: {},
      additionalProperties: { type: 'string' },
    },
  }),
  execute,
  examples: [
    {
      expression: 'QUERY_PARAMETERS_OBJECT("https://api.myapp.com/leads?search=all&limit=20")',
      result: { limit: '20', search: 'all' },
    },
    {
      expression: 'QUERY_PARAMETERS_OBJECT("http://example.com/path?name=Branch&products=Journeys&products=Email&products=Universal%20Ads")',
      result: {
        name: 'Branch',
        products: 'Universal Ads',
      },
    },
  ],
};
