// @sapappgyver/appgyver-formulas/functions/text/QUERY_PARAMETERS_ARRAY.js
const execute = require('./QUERY_PARAMETERS_ARRAY.runtime');

module.exports = {
  name: 'QUERY_PARAMETERS_ARRAY',
  category: 'Text',
  title: 'Get query parameters from URL as list',
  description: `Given an URL, return its query parameters as a list of objects with schema:

\`{ key: queryParamKeyName, value: queryParamValue }\`

Numbers and booleans are parsed to their correct types.`,
  parameters: [{
    title: 'URL',
    description: 'URL to convert',
    type: 'string',
  }],
  analyze: () => ({ schema: { type: 'array', items: { type: 'object', properties: { key: { type: 'string' }, value: { type: 'string' } } } } }),
  execute,
  examples: [
    {
      expression: 'QUERY_PARAMETERS_ARRAY("https://api.myapp.com/leads?search=all&limit=20")',
      result: [
        { key: 'limit', value: '20' },
        { key: 'search', value: 'all' },
      ],
    },
    {
      expression: 'QUERY_PARAMETERS_ARRAY("http://example.com/path?name=Branch&products=Journeys&products=Email&products=Universal%20Ads")',
      result: [
        { key: 'name', value: 'Branch' },
        { key: 'products', value: 'Journeys' },
        { key: 'products', value: 'Email' },
        { key: 'products', value: 'Universal Ads' },
      ],
    },
  ],
};
