// @sapappgyver/appgyver-formulas/functions/text/REPLACE_URL_PARAMETERS.js
const execute = require('./REPLACE_URL_PARAMETERS.runtime');

module.exports = {
  name: 'REPLACE_URL_PARAMETERS',
  category: 'Text',
  title: 'Set URL parameters',
  description: 'Sets the given URL texts query parameters to the given parameter object.',
  parameters: [{
    type: 'string',
    format: 'url',
    title: 'Base URL',
    description: 'Base URL text',
  }, {
    type: 'object',
    title: 'Parameters object',
    description: 'Dictionary for parameters',
  }],
  analyze: () => ({ schema: { type: 'string', format: 'url' } }),
  execute,
  examples: [
    {
      expression: 'REPLACE_URL_PARAMETERS("http://www.example.com", { test1: "abc", test2: 123})',
      result: 'http://www.example.com/?test1=abc&test2=123',
    },
    {
      expression: 'REPLACE_URL_PARAMETERS("http://www.example.com/?abc=123&test=abc", { test1: "abc", test2: 123})',
      result: 'http://www.example.com/?test1=abc&test2=123',
    },
    {
      expression: 'REPLACE_URL_PARAMETERS("http://www.example.com/?abc=123&test=abc", {})',
      result: 'http://www.example.com/',
    },
  ],
};
