// @sapappgyver/appgyver-formulas/functions/text/SPLIT.js
const execute = require('./SPLIT.runtime');

module.exports = {
  name: 'SPLIT',
  category: 'Text',
  title: 'Split a text',
  description: 'Splits a text into a list of sliced texts.'
    + ' The separator parameter specifies the characters within the text at which the splits will be done.'
    + ' If the separator is an empty text, the result is an array of single characters.',
  parameters: [{
    type: 'string',
    title: 'Text',
    description: 'Text to split',
    inversible: true,
  }, {
    type: 'string',
    title: 'Pattern',
    description: 'Token that splits the text',
  }, {
    type: 'number',
    title: 'Max slices',
    description: 'Maximum number of slices to which the list is split',
    default: null,
  }],
  maxInversions: 1,
  analyze: () => ({
    schema: {
      type: 'array',
      items: {
        type: 'string',
      },
    },
  }),
  execute,
  examples: [
    { expression: 'SPLIT("Foo Bar", " ")', result: ['Foo', 'Bar'] },
    { expression: 'SPLIT("Foo", "")', result: ['F', 'o', 'o'] },
    { expression: 'SPLIT("Foo Bar", "bbb")', result: ['Foo Bar'] },
    { expression: 'SPLIT("Foo Bar Baz xxx", " ", 3)', result: ['Foo', 'Bar', 'Baz xxx'] },
  ],
};
