// @sapappgyver/appgyver-formulas/functions/text/SUBSTRING.js
const execute = require('./SUBSTRING.runtime');

module.exports = {
  name: 'SUBSTRING',
  category: 'Text',
  title: 'Takes a slice of a text',
  description: `Returns a slice of the given text between the given indices.

The slice of text starts at (and includes) the index given as the 2nd parameter. It ends at (and excludes) the index given as the 3rd parameter.

If both indices are the same, the function returns an empty text. Negative indices are counted backwards, starting from the end of the text.`,
  parameters: [{
    type: 'string',
    title: 'Text',
    description: 'Text which to get a slice from',
    inversible: true,
  }, {
    type: 'number',
    title: 'Start',
    description: 'The index (starting from 0) indicating the first character to include.',
  }, {
    default: null,
    type: 'number',
    title: 'Stop',
    description: 'The index (starting from 0) indicating the end of the slice. This character will not be included in the result.'
        + 'If omitted, then the rest of the text will be included.',
  }],
  maxInversions: 1,
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    { expression: 'SUBSTRING("foo bar", 0, 3)', result: 'foo' },
    { expression: 'SUBSTRING("foo bar", 4)', result: 'bar' },
    { expression: 'SUBSTRING("foo bar", 0, -1)', result: 'foo ba' },
  ],
};
