// @sapappgyver/appgyver-formulas/functions/text/TRUNCATE.js
const execute = require('./TRUNCATE.runtime');

module.exports = {
  name: 'TRUNCATE',
  category: 'Text',
  title: 'Truncate the text',
  description: 'If the given text is longer than the given length parameter, it is shortened and appended with an ellipsis character, so that the resultant text length is always less or equal than the given length parameter. Returns the truncated text.',
  parameters: [{
    type: 'string',
    title: 'Text',
    description: 'Text',
  }, {
    type: 'number',
    title: 'Max length',
    description: 'Max length',
  }, {
    type: 'string',
    title: 'Ellipsis',
    description: 'Ellipsis character (default is "…")',
    default: '…',
  }],
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    { expression: 'TRUNCATE("Wissenschaft is a science", 5)', result: 'Wiss…' },
    { expression: 'TRUNCATE("This is a simple text", 4)', result: 'Thi…' },
    { expression: 'TRUNCATE("This is a simple text", 4, "...")', result: 'T...' },
    { expression: 'TRUNCATE("This is a simple text", 4, "")', result: 'This' },
  ],
};
