// @sapappgyver/appgyver-formulas/functions/inverts.js
/* eslint-disable global-require */
/* eslint sort-keys: "error" */
module.exports = {
  ADD: require('./math/ADD.invert'),
  DIVIDE: require('./math/DIVIDE.invert'),
  JOIN: require('./text/JOIN.invert'),
  LOOKUP: require('./object/LOOKUP.invert'),
  MAP: require('./list/MAP.invert'),
  MULTIPLY: require('./math/MULTIPLY.invert'),
  NOT: require('./utility/NOT.invert'),
  OPPOSITE: require('./math/OPPOSITE.invert'),
  ORDER: require('./list/ORDER.invert'),
  PICK_ITEM: require('./list/PICK_ITEM.invert'),
  PLUCK: require('./list/PLUCK.invert'),
  REVERSE: require('./list/REVERSE.invert'),
  SELECT: require('./list/SELECT.invert'),
  SLICE: require('./list/SLICE.invert'),
  SORT_BY_KEY: require('./list/SORT_BY_KEY.invert'),
  SORT_BY_KEY_CASE_SENSITIVE: require('./list/SORT_BY_KEY_CASE_SENSITIVE.invert'),
  SPLIT: require('./text/SPLIT.invert'),
  SUBSTRING: require('./text/SUBSTRING.invert'),
  SUBTRACT: require('./math/SUBTRACT.invert'),
  VALUES: require('./object/VALUES.invert'),
  WITHOUT_ITEM: require('./list/WITHOUT_ITEM.invert'),
  WITH_ITEM: require('./list/WITH_ITEM.invert'),
  ZIP: require('./list/ZIP.invert'),
};
