// @sapappgyver/appgyver-formulas/functions/runtime.js
/* eslint-disable global-require */
/* eslint sort-keys: "error" */
module.exports = {
  ABS: require('./math/ABS.runtime'),
  ACCRINT: require('./financial/ACCRINT.runtime'),
  ACOS: require('./math/ACOS.runtime'),
  ACOSEC: require('./math/ACOSEC.runtime'),
  ACOSH: require('./math/ACOSH.runtime'),
  ACOT: require('./math/ACOT.runtime'),
  ACOTH: require('./math/ACOTH.runtime'),
  ADD: require('./math/ADD.runtime'),
  ADD_DURATION: require('./date/ADD_DURATION.runtime'),
  ALPHA: require('./color/ALPHA.runtime'),
  AMORLINC: require('./financial/AMORLINC.runtime'),
  APPEND_URL_PARAMETERS: require('./text/APPEND_URL_PARAMETERS.runtime'),
  ASEC: require('./math/ASEC.runtime'),
  ASIN: require('./math/ASIN.runtime'),
  ASINH: require('./math/ASINH.runtime'),
  ATAN: require('./math/ATAN.runtime'),
  ATAN2: require('./math/ATAN2.runtime'),
  ATANH: require('./math/ATANH.runtime'),
  AVEDEV: require('./statistical/AVEDEV.runtime'),
  AVERAGE: require('./statistical/AVERAGE.runtime'),
  AVERAGE_WEIGHTED: require('./statistical/AVERAGE_WEIGHTED.runtime'),
  BESSELI: require('./engineering/BESSELI.runtime'),
  BESSELJ: require('./engineering/BESSELJ.runtime'),
  BESSELK: require('./engineering/BESSELK.runtime'),
  BESSELY: require('./engineering/BESSELY.runtime'),
  BETA: require('./statistical/BETA.runtime'),
  BETA_DIST: require('./statistical/BETA_DIST.runtime'),
  BETA_INV: require('./statistical/BETA_INV.runtime'),
  BINOMDIST: require('./statistical/BINOMDIST.runtime'),
  BIT_AND: require('./bitwise/BIT_AND.runtime'),
  BIT_NOT: require('./bitwise/BIT_NOT.runtime'),
  BIT_OR: require('./bitwise/BIT_OR.runtime'),
  BIT_SHIFT_LEFT: require('./bitwise/BIT_SHIFT_LEFT.runtime'),
  BIT_SHIFT_RIGHT: require('./bitwise/BIT_SHIFT_RIGHT.runtime'),
  BIT_SHIFT_RIGHT_ARITH: require('./bitwise/BIT_SHIFT_RIGHT_ARITH.runtime'),
  BIT_XOR: require('./bitwise/BIT_XOR.runtime'),
  BLUE: require('./color/BLUE.runtime'),
  BOOLEAN: require('./utility/BOOLEAN.runtime'),
  BUILD_OBJECT: require('./list/BUILD_OBJECT.runtime'),
  CAPITALIZE: require('./text/CAPITALIZE.runtime'),
  CEIL: require('./math/CEIL.runtime'),
  CHISQ_DIST: require('./statistical/CHISQ_DIST.runtime'),
  CHISQ_INV: require('./statistical/CHISQ_INV.runtime'),
  CHISQ_RT_DIST: require('./statistical/CHISQ_RT_DIST.runtime'),
  CHISQ_RT_INV: require('./statistical/CHISQ_RT_INV.runtime'),
  CHI_TEST: require('./statistical/CHI_TEST.runtime'),
  COLOR: require('./color/COLOR.runtime'),
  COMBIN: require('./math/COMBIN.runtime'),
  COMBINA: require('./math/COMBINA.runtime'),
  CONCAT: require('./list/CONCAT.runtime'),
  CONFIDENCE: require('./statistical/CONFIDENCE.runtime'),
  CONFIDENCE_T: require('./statistical/CONFIDENCE_T.runtime'),
  CONTAINS: require('./text/CONTAINS.runtime'),
  CORREL: require('./statistical/CORREL.runtime'),
  COS: require('./math/COS.runtime'),
  COSEC: require('./math/COSEC.runtime'),
  COSECH: require('./math/COSECH.runtime'),
  COSH: require('./math/COSH.runtime'),
  COT: require('./math/COT.runtime'),
  COTH: require('./math/COTH.runtime'),
  COUNT: require('./list/COUNT.runtime'),
  COUPDAYBS: require('./financial/COUPDAYBS.runtime'),
  COUPDAYS: require('./financial/COUPDAYS.runtime'),
  COUPDAYSNC: require('./financial/COUPDAYSNC.runtime'),
  COUPNCD: require('./financial/COUPNCD.runtime'),
  COUPNUM: require('./financial/COUPNUM.runtime'),
  COUPPCD: require('./financial/COUPPCD.runtime'),
  COVAR: require('./statistical/COVAR.runtime'),
  COVAR_P: require('./statistical/COVAR_P.runtime'),
  CUMIPMT: require('./financial/CUMIPMT.runtime'),
  CUMPRINC: require('./financial/CUMPRINC.runtime'),
  CURRENCY_DECIMAL: require('./financial/CURRENCY_DECIMAL.runtime'),
  CURRENCY_FRACTIONAL: require('./financial/CURRENCY_FRACTIONAL.runtime'),
  DARKEN: require('./color/DARKEN.runtime'),
  DATE: require('./deprecated/DATE.runtime'),
  DATETIME: require('./date/DATETIME.runtime'),
  DATETIME_DIFFERENCE: require('./date/DATETIME_DIFFERENCE.runtime'),
  DATETIME_IS_AFTER: require('./date/DATETIME_IS_AFTER.runtime'),
  DATETIME_IS_BEFORE: require('./date/DATETIME_IS_BEFORE.runtime'),
  DATETIME_IS_BETWEEN: require('./date/DATETIME_IS_BETWEEN.runtime'),
  DATETIME_IS_SAME: require('./date/DATETIME_IS_SAME.runtime'),
  DATETIME_IS_SAME_OR_AFTER: require('./date/DATETIME_IS_SAME_OR_AFTER.runtime'),
  DATETIME_IS_SAME_OR_BEFORE: require('./date/DATETIME_IS_SAME_OR_BEFORE.runtime'),
  DECLINING_BALANCE: require('./financial/DECLINING_BALANCE.runtime'),
  DECODE_BASE64: require('./utility/DECODE_BASE64.runtime'),
  DECODE_JSON: require('./utility/DECODE_JSON.runtime'),
  DEFAULT: require('./utility/DEFAULT.runtime'),
  DEFAULT_FILE_EXTENSION: require('./utility/DEFAULT_FILE_EXTENSION.runtime'),
  DEGREES: require('./math/DEGREES.runtime'),
  DELETE_KEY: require('./object/DELETE_KEY.runtime'),
  DESATURATE: require('./color/DESATURATE.runtime'),
  DEVSQ: require('./statistical/DEVSQ.runtime'),
  DIFFERENCE: require('./list/DIFFERENCE.runtime'),
  DISC: require('./financial/DISC.runtime'),
  DIV: require('./math/DIV.runtime'),
  DIVIDE: require('./math/DIVIDE.runtime'),
  DOUBLE_DECLINING_BALANCE: require('./financial/DOUBLE_DECLINING_BALANCE.runtime'),
  EFFECT: require('./financial/EFFECT.runtime'),
  ENCODE_BASE64: require('./utility/ENCODE_BASE64.runtime'),
  ENCODE_FOR_URL: require('./utility/ENCODE_FOR_URL.runtime'),
  ENCODE_JSON: require('./utility/ENCODE_JSON.runtime'),
  ENCODE_OBJECT_FOR_URL: require('./utility/ENCODE_OBJECT_FOR_URL.runtime'),
  ERF: require('./statistical/ERF.runtime'),
  ERFC: require('./statistical/ERFC.runtime'),
  ERFC_INV: require('./statistical/ERFC_INV.runtime'),
  EULER: require('./math/EULER.runtime'),
  EVAL: require('./deprecated/EVAL.runtime'),
  EVEN: require('./math/EVEN.runtime'),
  EVERY: require('./list/EVERY.runtime'),
  EXACTLY_ONE: require('./list/EXACTLY_ONE.runtime'),
  EXCLUDE_BY_KEY: require('./list/EXCLUDE_BY_KEY.runtime'),
  EXCLUDE_BY_REGEX: require('./list/EXCLUDE_BY_REGEX.runtime'),
  EXP: require('./math/EXP.runtime'),
  EXPON_DIST: require('./statistical/EXPON_DIST.runtime'),
  EXPON_INV: require('./statistical/EXPON_INV.runtime'),
  FACTORIAL: require('./math/FACTORIAL.runtime'),
  FADE_IN: require('./color/FADE_IN.runtime'),
  FADE_OUT: require('./color/FADE_OUT.runtime'),
  FILENAME_IS_IMAGE: require('./utility/FILENAME_IS_IMAGE.runtime'),
  FIND: require('./list/FIND.runtime'),
  FIND_BY_KEY: require('./list/FIND_BY_KEY.runtime'),
  FIND_BY_REGEX: require('./list/FIND_BY_REGEX.runtime'),
  FIRST_ITEM: require('./list/FIRST_ITEM.runtime'),
  FISHER: require('./statistical/FISHER.runtime'),
  FISHER_INV: require('./statistical/FISHER_INV.runtime'),
  FLATTEN: require('./list/FLATTEN.runtime'),
  FLOOR: require('./math/FLOOR.runtime'),
  FORECAST: require('./statistical/FORECAST.runtime'),
  FORMAT_DATE: require('./deprecated/FORMAT_DATE.runtime'),
  FORMAT_DATETIME_LOCAL: require('./date/FORMAT_DATETIME_LOCAL.runtime'),
  FORMAT_DATETIME_WITH_TIMEZONE: require('./date/FORMAT_DATETIME_WITH_TIMEZONE.runtime'),
  FORMAT_LOCALIZED_DECIMAL: require('./text/FORMAT_LOCALIZED_DECIMAL.runtime'),
  FUTURE_VALUE: require('./financial/FUTURE_VALUE.runtime'),
  FUTURE_VALUE_SCHEDULE: require('./financial/FUTURE_VALUE_SCHEDULE.runtime'),
  F_DIST: require('./statistical/F_DIST.runtime'),
  F_DIST_RT: require('./statistical/F_DIST_RT.runtime'),
  F_INV: require('./statistical/F_INV.runtime'),
  F_INV_RT: require('./statistical/F_INV_RT.runtime'),
  F_TEST: require('./statistical/F_TEST.runtime'),
  GAMMA: require('./statistical/GAMMA.runtime'),
  GAMMALN: require('./statistical/GAMMALN.runtime'),
  GAMMA_DIST: require('./statistical/GAMMA_DIST.runtime'),
  GAMMA_INV: require('./statistical/GAMMA_INV.runtime'),
  GAUSS: require('./statistical/GAUSS.runtime'),
  GENERATE_RANGE: require('./list/GENERATE_RANGE.runtime'),
  GENERATE_UUID: require('./text/GENERATE_UUID.runtime'),
  GEOMEAN: require('./statistical/GEOMEAN.runtime'),
  GET_DATETIME_COMPONENT: require('./date/GET_DATETIME_COMPONENT.runtime'),
  GREATEST_COMMON_DIVISOR: require('./math/GREATEST_COMMON_DIVISOR.runtime'),
  GREEN: require('./color/GREEN.runtime'),
  GROUP: require('./list/GROUP.runtime'),
  HARMEAN: require('./statistical/HARMEAN.runtime'),
  HSL: require('./color/HSL.runtime'),
  HUE: require('./color/HUE.runtime'),
  HYPGEOM_DIST: require('./statistical/HYPGEOM_DIST.runtime'),
  HYPOT: require('./math/HYPOT.runtime'),
  IF: require('./utility/IF.runtime'),
  INDEX_OF: require('./list/INDEX_OF.runtime'),
  INSERT_ITEMS_AT: require('./list/INSERT_ITEMS_AT.runtime'),
  INSERT_ITEM_AT: require('./list/INSERT_ITEM_AT.runtime'),
  INTEGER: require('./utility/INTEGER.runtime'),
  INTERCEPT: require('./statistical/INTERCEPT.runtime'),
  INTERSECTION: require('./list/INTERSECTION.runtime'),
  INTRATE: require('./financial/INTRATE.runtime'),
  INVESTMENT_DURATION: require('./financial/INVESTMENT_DURATION.runtime'),
  INVESTMENT_MDURATION: require('./financial/INVESTMENT_MDURATION.runtime'),
  INVESTMENT_PDURATION: require('./financial/INVESTMENT_PDURATION.runtime'),
  IPMT: require('./financial/IPMT.runtime'),
  IRR: require('./financial/IRR.runtime'),
  ISPMT: require('./financial/ISPMT.runtime'),
  IS_DATETIME: require('./date/IS_DATETIME.runtime'),
  IS_EMAIL: require('./utility/IS_EMAIL.runtime'),
  IS_EMPTY: require('./utility/IS_EMPTY.runtime'),
  IS_EQUAL: require('./utility/IS_EQUAL.runtime'),
  IS_EVEN: require('./math/IS_EVEN.runtime'),
  IS_IN_ARRAY: require('./list/IS_IN_ARRAY.runtime'),
  IS_IN_ARRAY_BY_KEY: require('./list/IS_IN_ARRAY_BY_KEY.runtime'),
  IS_NAN: require('./math/IS_NAN.runtime'),
  IS_NULL: require('./utility/IS_NULL.runtime'),
  IS_NULLY: require('./utility/IS_NULLY.runtime'),
  IS_NUMBER: require('./utility/IS_NUMBER.runtime'),
  IS_ODD: require('./math/IS_ODD.runtime'),
  IS_SAME: require('./utility/IS_SAME.runtime'),
  IS_UNDEFINED: require('./utility/IS_UNDEFINED.runtime'),
  JOIN: require('./text/JOIN.runtime'),
  KECCAK: require('./cryptography/KECCAK.runtime'),
  KEYS: require('./object/KEYS.runtime'),
  KURT: require('./statistical/KURT.runtime'),
  LAST_ITEM: require('./list/LAST_ITEM.runtime'),
  LEAST_COMMON_MULTIPLE: require('./math/LEAST_COMMON_MULTIPLE.runtime'),
  LEET: require('./text/LEET.runtime'),
  LEFT_PAD: require('./text/LEFT_PAD.runtime'),
  LEFT_STRIP: require('./text/LEFT_STRIP.runtime'),
  LENGTH: require('./text/LENGTH.runtime'),
  LIGHTEN: require('./color/LIGHTEN.runtime'),
  LIGHTNESS: require('./color/LIGHTNESS.runtime'),
  LN: require('./math/LN.runtime'),
  LOCALIZE_NUMBER: require('./deprecated/LOCALIZE_NUMBER.runtime'),
  LOCAL_TIMEZONE: require('./date/LOCAL_TIMEZONE.runtime'),
  LOG: require('./math/LOG.runtime'),
  LOG10: require('./math/LOG10.runtime'),
  LOG2: require('./math/LOG2.runtime'),
  LOGNORM_DIST: require('./statistical/LOGNORM_DIST.runtime'),
  LOGNORM_INV: require('./statistical/LOGNORM_INV.runtime'),
  LOOKUP: require('./object/LOOKUP.runtime'),
  LOOKUP_PATH: require('./object/LOOKUP_PATH.runtime'),
  LOWERCASE: require('./text/LOWERCASE.runtime'),
  MAP: require('./list/MAP.runtime'),
  MATCHES_REGEX: require('./text/MATCHES_REGEX.runtime'),
  MAX: require('./math/MAX.runtime'),
  MD5: require('./cryptography/MD5.runtime'),
  MEDIAN: require('./statistical/MEDIAN.runtime'),
  MERGE: require('./object/MERGE.runtime'),
  MIME_TYPE: require('./utility/MIME_TYPE.runtime'),
  MIME_TYPE_IS_IMAGE: require('./utility/MIME_TYPE_IS_IMAGE.runtime'),
  MIN: require('./math/MIN.runtime'),
  MIRR: require('./financial/MIRR.runtime'),
  MIX: require('./color/MIX.runtime'),
  MOD: require('./math/MOD.runtime'),
  MODE: require('./statistical/MODE.runtime'),
  MULTINOMIAL: require('./math/MULTINOMIAL.runtime'),
  MULTIPLY: require('./math/MULTIPLY.runtime'),
  NEG_BINOM_DIST: require('./statistical/NEG_BINOM_DIST.runtime'),
  NET_PRESENT_VALUE: require('./financial/NET_PRESENT_VALUE.runtime'),
  NOMINAL: require('./financial/NOMINAL.runtime'),
  NORM_DIST: require('./statistical/NORM_DIST.runtime'),
  NORM_INV: require('./statistical/NORM_INV.runtime'),
  NORM_S_DIST: require('./statistical/NORM_S_DIST.runtime'),
  NORM_S_INV: require('./statistical/NORM_S_INV.runtime'),
  NOT: require('./utility/NOT.runtime'),
  NOW: require('./date/NOW.runtime'),
  NPER: require('./financial/NPER.runtime'),
  NTH_LARGEST: require('./statistical/NTH_LARGEST.runtime'),
  NTH_SMALLEST: require('./statistical/NTH_SMALLEST.runtime'),
  NUMBER: require('./utility/NUMBER.runtime'),
  NUMERIC_BASE: require('./utility/NUMERIC_BASE.runtime'),
  ODD: require('./math/ODD.runtime'),
  OMIT_KEYS: require('./object/OMIT_KEYS.runtime'),
  OPPOSITE: require('./math/OPPOSITE.runtime'),
  ORDER: require('./list/ORDER.runtime'),
  PERCENTILE: require('./statistical/PERCENTILE.runtime'),
  PERCENTILE_EXC: require('./statistical/PERCENTILE_EXC.runtime'),
  PERCENT_RANK: require('./statistical/PERCENT_RANK.runtime'),
  PERCENT_RANK_EXC: require('./statistical/PERCENT_RANK_EXC.runtime'),
  PERMUT: require('./statistical/PERMUT.runtime'),
  PERMUTATIONA: require('./statistical/PERMUTATIONA.runtime'),
  PHI: require('./statistical/PHI.runtime'),
  PI: require('./math/PI.runtime'),
  PICK_ITEM: require('./list/PICK_ITEM.runtime'),
  PICK_KEYS: require('./object/PICK_KEYS.runtime'),
  PLUCK: require('./list/PLUCK.runtime'),
  PMT: require('./financial/PMT.runtime'),
  POISSON_DIST: require('./statistical/POISSON_DIST.runtime'),
  POLYNOMIAL: require('./math/POLYNOMIAL.runtime'),
  POWER: require('./math/POWER.runtime'),
  PPMT: require('./financial/PPMT.runtime'),
  PRESENT_VALUE: require('./financial/PRESENT_VALUE.runtime'),
  PRICE: require('./financial/PRICE.runtime'),
  PRICEDISC: require('./financial/PRICEDISC.runtime'),
  PRICEMAT: require('./financial/PRICEMAT.runtime'),
  PROB: require('./statistical/PROB.runtime'),
  PRODUCT: require('./math/PRODUCT.runtime'),
  QUARTILE: require('./statistical/QUARTILE.runtime'),
  QUARTILE_EXC: require('./statistical/QUARTILE_EXC.runtime'),
  QUERY_PARAMETERS_ARRAY: require('./text/QUERY_PARAMETERS_ARRAY.runtime'),
  QUERY_PARAMETERS_OBJECT: require('./text/QUERY_PARAMETERS_OBJECT.runtime'),
  RADIANS: require('./math/RADIANS.runtime'),
  RANDOM: require('./math/RANDOM.runtime'),
  RANDOM_ARRAY: require('./math/RANDOM_ARRAY.runtime'),
  RANDOM_INTEGER_BETWEEN: require('./math/RANDOM_INTEGER_BETWEEN.runtime'),
  RANDOM_NUMBER_BETWEEN: require('./math/RANDOM_NUMBER_BETWEEN.runtime'),
  RANK: require('./statistical/RANK.runtime'),
  RANK_AVG: require('./statistical/RANK_AVG.runtime'),
  RATE: require('./financial/RATE.runtime'),
  RECEIVED: require('./financial/RECEIVED.runtime'),
  RED: require('./color/RED.runtime'),
  REDUCE: require('./list/REDUCE.runtime'),
  REDUCE_INIT: require('./list/REDUCE_INIT.runtime'),
  REMOVE_ITEM: require('./list/REMOVE_ITEM.runtime'),
  REMOVE_ITEMS_AT: require('./list/REMOVE_ITEMS_AT.runtime'),
  REMOVE_ITEM_BY_KEY: require('./list/REMOVE_ITEM_BY_KEY.runtime'),
  REPEAT: require('./list/REPEAT.runtime'),
  REPLACE_ALL: require('./text/REPLACE_ALL.runtime'),
  REPLACE_ALL_REGEX: require('./text/REPLACE_ALL_REGEX.runtime'),
  REPLACE_ONE: require('./text/REPLACE_ONE.runtime'),
  REPLACE_ONE_REGEX: require('./text/REPLACE_ONE_REGEX.runtime'),
  REPLACE_URL_PARAMETERS: require('./text/REPLACE_URL_PARAMETERS.runtime'),
  REVERSE: require('./list/REVERSE.runtime'),
  REVERSE_LOOKUP: require('./object/REVERSE_LOOKUP.runtime'),
  REVERSE_LOOKUP_ALL: require('./object/REVERSE_LOOKUP_ALL.runtime'),
  RGB: require('./color/RGB.runtime'),
  RIGHT_PAD: require('./text/RIGHT_PAD.runtime'),
  RIGHT_STRIP: require('./text/RIGHT_STRIP.runtime'),
  RIPEMD160: require('./cryptography/RIPEMD160.runtime'),
  ROUND: require('./math/ROUND.runtime'),
  RRI: require('./financial/RRI.runtime'),
  RSQ: require('./statistical/RSQ.runtime'),
  SATURATE: require('./color/SATURATE.runtime'),
  SATURATION: require('./color/SATURATION.runtime'),
  SCHEMA_DESCRIPTION: require('./schema/SCHEMA_DESCRIPTION.runtime'),
  SCHEMA_LIST_ITEM: require('./schema/SCHEMA_LIST_ITEM.runtime'),
  SCHEMA_OBJECT_PROPERTY: require('./schema/SCHEMA_OBJECT_PROPERTY.runtime'),
  SCHEMA_POSSIBLE_OBJECT_KEYS: require('./schema/SCHEMA_POSSIBLE_OBJECT_KEYS.runtime'),
  SCHEMA_REQUIRED_OBJECT_KEYS: require('./schema/SCHEMA_REQUIRED_OBJECT_KEYS.runtime'),
  SCHEMA_TITLE: require('./schema/SCHEMA_TITLE.runtime'),
  SCHEMA_TYPE: require('./schema/SCHEMA_TYPE.runtime'),
  SEC: require('./math/SEC.runtime'),
  SECH: require('./math/SECH.runtime'),
  SELECT: require('./list/SELECT.runtime'),
  SELECT_BY_KEY: require('./list/SELECT_BY_KEY.runtime'),
  SELECT_BY_REGEX: require('./list/SELECT_BY_REGEX.runtime'),
  SEQUENCE: require('./math/SEQUENCE.runtime'),
  SERIES_SUM: require('./math/SERIES_SUM.runtime'),
  SET_ALPHA: require('./color/SET_ALPHA.runtime'),
  SET_BLUE: require('./color/SET_BLUE.runtime'),
  SET_DATETIME_COMPONENT: require('./date/SET_DATETIME_COMPONENT.runtime'),
  SET_GREEN: require('./color/SET_GREEN.runtime'),
  SET_HUE: require('./color/SET_HUE.runtime'),
  SET_ITEM: require('./list/SET_ITEM.runtime'),
  SET_ITEM_AT: require('./list/SET_ITEM_AT.runtime'),
  SET_KEY: require('./object/SET_KEY.runtime'),
  SET_LIGHTNESS: require('./color/SET_LIGHTNESS.runtime'),
  SET_RED: require('./color/SET_RED.runtime'),
  SET_SATURATION: require('./color/SET_SATURATION.runtime'),
  SET_TIMEZONE: require('./date/SET_TIMEZONE.runtime'),
  SHA1: require('./cryptography/SHA1.runtime'),
  SHA256: require('./cryptography/SHA256.runtime'),
  SHA512: require('./cryptography/SHA512.runtime'),
  SHIFT_HUE: require('./color/SHIFT_HUE.runtime'),
  SHUFFLE: require('./list/SHUFFLE.runtime'),
  SIGN: require('./math/SIGN.runtime'),
  SIN: require('./math/SIN.runtime'),
  SINH: require('./math/SINH.runtime'),
  SKEW: require('./statistical/SKEW.runtime'),
  SKEW_P: require('./statistical/SKEW_P.runtime'),
  SLICE: require('./list/SLICE.runtime'),
  SLN: require('./financial/SLN.runtime'),
  SLOPE: require('./statistical/SLOPE.runtime'),
  SOME: require('./list/SOME.runtime'),
  SORT: require('./list/SORT.runtime'),
  SORT_BY_KEY: require('./list/SORT_BY_KEY.runtime'),
  SORT_BY_KEY_CASE_SENSITIVE: require('./list/SORT_BY_KEY_CASE_SENSITIVE.runtime'),
  SORT_CASE_SENSITIVE: require('./list/SORT_CASE_SENSITIVE.runtime'),
  SPLICE: require('./list/SPLICE.runtime'),
  SPLIT: require('./text/SPLIT.runtime'),
  SQRT: require('./math/SQRT.runtime'),
  SQRT_PI: require('./math/SQRT_PI.runtime'),
  STANDARDIZE: require('./statistical/STANDARDIZE.runtime'),
  STDEV: require('./statistical/STDEV.runtime'),
  STDEV_P: require('./statistical/STDEV_P.runtime'),
  STEYX: require('./statistical/STEYX.runtime'),
  STRING: require('./utility/STRING.runtime'),
  STRIP_HTML_TAGS: require('./text/STRIP_HTML_TAGS.runtime'),
  SUBSTRING: require('./text/SUBSTRING.runtime'),
  SUBTRACT: require('./math/SUBTRACT.runtime'),
  SUBTRACT_DURATION: require('./date/SUBTRACT_DURATION.runtime'),
  SUM: require('./math/SUM.runtime'),
  SUM_BY_KEY: require('./list/SUM_BY_KEY.runtime'),
  SUM_SQUARED: require('./math/SUM_SQUARED.runtime'),
  SYD: require('./financial/SYD.runtime'),
  SYMMETRIC_DIFFERENCE: require('./list/SYMMETRIC_DIFFERENCE.runtime'),
  TAN: require('./math/TAN.runtime'),
  TANH: require('./math/TANH.runtime'),
  TBILLPRICE: require('./financial/TBILLPRICE.runtime'),
  TBILLYIELD: require('./financial/TBILLYIELD.runtime'),
  TIMESTAMP: require('./date/TIMESTAMP.runtime'),
  TIMEZONE: require('./date/TIMEZONE.runtime'),
  TRIM_MEAN: require('./statistical/TRIM_MEAN.runtime'),
  TRIM_WHITESPACE: require('./text/TRIM_WHITESPACE.runtime'),
  TRUNCATE: require('./text/TRUNCATE.runtime'),
  TYPE_OF: require('./utility/TYPE_OF.runtime'),
  T_DIST: require('./statistical/T_DIST.runtime'),
  T_DIST_2T: require('./statistical/T_DIST_2T.runtime'),
  T_DIST_RT: require('./statistical/T_DIST_RT.runtime'),
  T_INV: require('./statistical/T_INV.runtime'),
  T_INV_2T: require('./statistical/T_INV_2T.runtime'),
  T_TEST: require('./statistical/T_TEST.runtime'),
  UNION: require('./list/UNION.runtime'),
  UNIQUE_BY_KEY: require('./list/UNIQUE_BY_KEY.runtime'),
  UPPERCASE: require('./text/UPPERCASE.runtime'),
  URL: require('./utility/URL.runtime'),
  VALUES: require('./object/VALUES.runtime'),
  VARIANCE: require('./statistical/VARIANCE.runtime'),
  VARIANCE_B: require('./statistical/VARIANCE_B.runtime'),
  VARIANCE_P: require('./statistical/VARIANCE_P.runtime'),
  VDB: require('./financial/VDB.runtime'),
  WEIBULL_DIST: require('./statistical/WEIBULL_DIST.runtime'),
  WEIBULL_INV: require('./statistical/WEIBULL_INV.runtime'),
  WITHOUT_ITEM: require('./list/WITHOUT_ITEM.runtime'),
  WITH_ITEM: require('./list/WITH_ITEM.runtime'),
  WITH_UNIQUE_ITEM: require('./list/WITH_UNIQUE_ITEM.runtime'),
  WITH_UNIQUE_ITEM_BY_KEY: require('./list/WITH_UNIQUE_ITEM_BY_KEY.runtime'),
  XIRR: require('./financial/XIRR.runtime'),
  XNPV: require('./financial/XNPV.runtime'),
  YEARFRAC: require('./date/YEARFRAC.runtime'),
  YIELDDISC: require('./financial/YIELDDISC.runtime'),
  YIELDMAT: require('./financial/YIELDMAT.runtime'),
  ZIP: require('./list/ZIP.runtime'),
  Z_TEST: require('./statistical/Z_TEST.runtime'),
};
