// @sapappgyver/appgyver-formulas/functions/math/ASINH.js
const execute = require('./ASINH.runtime');

module.exports = {
  name: 'ASINH',
  category: 'Math',
  title: 'Calculate hyperbolic arcsine',
  description: 'Calculates the inverse hyberbolic sine of the given number. The value is returned in radians.',
  parameters: [{
    type: 'number',
    title: 'Number',
    description: 'The value for which to calculate the inverse hyperbolic sine.',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'ASINH(0)', result: 0 },
    { expression: 'ASINH(1)', result: 0.881373587019543 },
    { expression: 'ASINH(10)', result: 2.99822295029797 },
    { expression: 'ASINH(10000)', result: 9.903487550036129 },
    { expression: 'ASINH(EULER())', result: 1.725382558852315 },
    { expression: 'ASINH(-1)', result: -0.881373587019543 },
    { expression: 'ASINH(-10)', result: -2.99822295029797 },
    { expression: 'ASINH(-10000)', result: -9.903487550036129 },
    { expression: 'ASINH(-EULER())', result: -1.725382558852315 },
  ],
};
