// @sapappgyver/appgyver-formulas/functions/math/POLYNOMIAL.js
const execute = require('./POLYNOMIAL.runtime');

module.exports = {
  name: 'POLYNOMIAL',
  category: 'Math',
  title: 'Calculate polynomial series',
  description: `Given a list of numbers such as \`[a, b, c]\`, calculates the sum \`c*x⁰ + b*x¹ + a*x²\`. If there are more elements in the list, the series is continued.

If the given list is empty, returns 0.`,
  parameters: [{
    type: 'number',
    title: 'X value',
    description: 'The input value to the power series.',
  }, {
    type: 'array',
    title: 'Coefficients A',
    description: 'The coefficients by which each successive power of x-value is multiplied.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'POLYNOMIAL(10, [1, 3, 9])', result: 139 },
    { expression: 'POLYNOMIAL(5, [1, 3, 9])', result: 49 },
    { expression: 'POLYNOMIAL(-5, [1, 3, 9])', result: 19 },
    { expression: 'POLYNOMIAL(5, [-1, 3, 9])', result: -1 },
    { expression: 'POLYNOMIAL(10, [1])', result: 1 },
    { expression: 'POLYNOMIAL(10, [])', result: 0 },
  ],
};
