// @sapappgyver/appgyver-formulas/functions/math/SERIES_SUM.js
const execute = require('./SERIES_SUM.runtime');

module.exports = {
  name: 'SERIES_SUM',
  category: 'Math',
  title: 'Returns SERIES_SUM',
  description: `Calculates the sum of a power series. Given a list of numbers \`[a₀, a₁, a₂, ... , aᵢ]\`, and the constants \`x\`, \`n\` and \`m\`, the sum is calculated using the formula:

\`a₀ × x^n + a₁ × x^(n+m) + a₂ × x^(n + 2 × m) + ... + aᵢ × x^(n + i × m)\``,
  parameters: [{
    type: 'number',
    title: 'Number x',
    description: 'The `x` value input to the power series.',
  }, {
    type: 'number',
    title: 'Number n',
    description: 'The initial power to which to raise `x` in the power series.',
  }, {
    type: 'number',
    title: 'Number m',
    description: 'The additive increment by which to increase the power to which `x` is raised.',
  }, {
    type: 'array',
    items: {
      type: 'number',
    },
    title: 'List',
    description: 'The list or range containing the coefficients of the power series.',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'SERIES_SUM(1, 2, 3, [1, 2, 3])', result: 6 },
    { expression: 'SERIES_SUM(2, 2, 3, [1, 2, 3])', result: 836 },
    { expression: 'SERIES_SUM(3, 2, 3, [1, 2, 3])', result: 20178 },
  ],
};
