// @sapappgyver/appgyver-formulas/functions/list/EVERY.js
const execute = require('./EVERY.runtime');

module.exports = {
  name: 'EVERY',
  category: 'List',
  title: 'Check if all items in a list match',
  description: `Checks if every item in the given list match the given criteria. Returns \`true\` each value matches, otherwise \`false\`.
Returns \`true\` also if the list is empty.

The criteria is defined as a formula at the second parameter. The formula is evaluated with each item in the list.
The formula must result in a true/false value. When the formula results in \`false\`, then the function will also return \`false\`.`,
  searchNames: ['ALL'],
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'List of items to search',
    items: {
      $id: '#item', // Make this a generic type
    },
  }, {
    type: 'function',
    title: 'Selection formula',
    description: 'Formula used to determine if an item matches the criteria',
    parameters: [{
      name: 'item',
      title: 'Current item',
      description: 'The current item from the original list',
      $ref: '#item', // Refers to the generic type
    }, {
      name: 'index',
      type: 'number',
      minimum: 0,
      title: 'Current index',
      description: 'The position of the current item in the list as an index number starting from 0',
      examples: [0, 1, 2],
    }],
    returnValue: {
      type: 'boolean',
      title: 'Whether the item matches',
      description: 'Whether or not the current item matches the criteria',
    },
  }],
  analyze: () => ({
    schema: {
      type: 'boolean',
    },
  }),
  execute,
  examples: [
    {
      expression: 'EVERY(animals, item.weight > 5)',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: true,
    },
    {
      expression: 'EVERY(animals, item.weight < 100)',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: false,
    },
    {
      expression: 'EVERY(animals, item.name != "horse")',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: true,
    },
    {
      expression: 'EVERY<product>(products, product.category == "Utility")',
      context: {
        products: [
          { name: 'Duct tape', category: 'Utility' },
          { name: 'Swiss army knife', category: 'Tools' },
          { name: 'Glue', category: 'Utility' },
        ],
      },
      result: false,
    },
  ],
};
