// @sapappgyver/appgyver-formulas/functions/list/EXACTLY_ONE.js
const execute = require('./EXACTLY_ONE.runtime');

module.exports = {
  name: 'EXACTLY_ONE',
  category: 'List',
  title: 'Check if exactly one item in a list matches',
  description: `Checks if there is exactly one item in the given list that matches the given criteria. Returns \`true\` if one match was found, otherwise \`false\`.

The criteria is defined as a formula at the second parameter. The formula is evaluated with each item in the list.
The formula must result in a true/false value. When the formula results in \`true\` exactly once, then the function will also return \`true\`.`,
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'List of items to search',
    items: {
      $id: '#item', // Make this a generic type
    },
  }, {
    type: 'function',
    title: 'Selection formula',
    description: 'Formula used to determine if an item matches the criteria',
    parameters: [{
      name: 'item',
      title: 'Current item',
      description: 'The current item from the original list',
      $ref: '#item', // Refers to the generic type
    }, {
      name: 'index',
      type: 'number',
      minimum: 0,
      title: 'Current index',
      description: 'The position of the current item in the list as an index number starting from 0',
      examples: [0, 1, 2],
    }],
    returnValue: {
      type: 'boolean',
      title: 'Whether the item matches',
      description: 'Whether or not the current item matches the criteria',
    },
  }],
  analyze: () => ({
    schema: {
      type: 'boolean',
    },
  }),
  execute,
  examples: [
    {
      expression: 'EXACTLY_ONE(animals, item.name == "cat")',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: true,
    },
    {
      expression: 'EXACTLY_ONE(animals, item.name == "horse")',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: false,
    },
    {
      expression: 'EXACTLY_ONE(animals, item.weight > 100)',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: false,
    },
    {
      expression: 'EXACTLY_ONE<product>(products, product.category == "Utility")',
      context: {
        products: [
          { name: 'Duct tape', category: 'Utility' },
          { name: 'Swiss army knife', category: 'Tools' },
          { name: 'Glue', category: 'Utility' },
        ],
      },
      result: false,
    },
    {
      expression: 'EXACTLY_ONE<product, position>(products, position == 1)',
      context: {
        products: [
          { name: 'Duct tape', category: 'Utility' },
          { name: 'Swiss army knife', category: 'Tools' },
          { name: 'Glue', category: 'Utility' },
        ],
      },
      result: true,
    },
  ],
};
