// @sapappgyver/appgyver-formulas/functions/list/FIND.js
const { getArrayItemSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./FIND.runtime');

module.exports = {
  name: 'FIND',
  category: 'List',
  title: 'Finds an item from list',
  description: `Finds the first item from the given list that matches the given criteria. Returns that item, or \`null\` if not found.

The criteria is defined as a formula at the second parameter. The formula is evaluated with each item in the list.
The formula must result in a true/false value. If \`true\`, then item will be included to the result, otherwise it will be excluded.`,
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'List of items to search',
    items: {
      $id: '#item', // Make this a generic type
    },
  }, {
    type: 'function',
    title: 'Selection formula',
    description: 'Formula used to determine if an item is the one to find from the list',
    parameters: [{
      name: 'item',
      title: 'Current item',
      description: 'The current item from the original list',
      $ref: '#item', // Refers to the generic type
    }, {
      name: 'index',
      type: 'number',
      minimum: 0,
      title: 'Current index',
      description: 'The position of the current item in the list as an index number starting from 0',
      examples: [0, 1, 2],
    }],
    returnValue: {
      type: 'boolean',
      title: 'Whether the item matches',
      description: 'Whether or not the current item is the one to be found from the list',
    },
  }],
  analyze: ([arrayParam]) => ({
    schema: getArrayItemSchema(arrayParam && arrayParam.schema) || {},
  }),
  execute,
  examples: [
    {
      expression: 'FIND(animals, item.name == "cat")',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: { name: 'cat', weight: 48 },
    },
    {
      expression: 'FIND(animals, item.name == "horse")',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: undefined,
    },
    {
      expression: 'FIND(animals, index == 2)',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: { name: 'rat', weight: 10 },
    },
    {
      expression: 'FIND<product>(products, product.name == "Duct tape")',
      context: {
        products: [
          { name: 'Duct tape', category: 'Utility' },
          { name: 'Swiss army knife', category: 'Tools' },
          { name: 'Glue', category: 'Utility' },
        ],
      },
      result: { name: 'Duct tape', category: 'Utility' },
    },
    {
      expression: 'FIND<product, position>(products, position == 1)',
      context: {
        products: [
          { name: 'Duct tape', category: 'Utility' },
          { name: 'Swiss army knife', category: 'Tools' },
          { name: 'Glue', category: 'Utility' },
        ],
      },
      result: { name: 'Swiss army knife', category: 'Tools' },
    },
  ],
};
