// @sapappgyver/appgyver-formulas/functions/list/FIND_BY_KEY.js
const { resolveObjectMemberAccessor, getSchemaAssignmentErrors, getArrayItemSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./FIND_BY_KEY.runtime');

module.exports = {
  name: 'FIND_BY_KEY',
  category: 'List',
  title: 'Find an object in list by a key',
  description: `Finds an object in the given list by a key-value pair.

For example, given **Key** \`"id"\` and **Value** \`123\`, will return the object in the list whose \`id\` property is \`123\`, or \`null\` if such an object was not found.

If multiple matches exist, the first matching object is returned.`,
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'List to find object in',
    items: {
      type: 'object',
    },
  }, {
    title: 'Key',
    description: 'Key whose value to match',
    type: 'string',
  }, {
    title: 'Value',
    description: 'Value to match',
  }],
  analyze: ([arrayParam, keyParam, itemParam], offset, parameterErrors) => {
    const arraySchema = (arrayParam && arrayParam.schema) || {};
    const itemsSchema = getArrayItemSchema(arraySchema) || { type: 'object' };
    const keySchema = (keyParam && keyParam.schema) || {};
    const {
      errors: keyErrors,
      schema: keyItemSchema,
    } = resolveObjectMemberAccessor(itemsSchema, keySchema);
    const itemSchema = (itemParam && itemParam.schema) || {};
    const assignmentErrors = getSchemaAssignmentErrors(itemSchema, keyItemSchema);
    return {
      schema: itemsSchema,
      errors: parameterErrors.length ? [] : [
        ...keyErrors.map(error => ({ ...error, offset: keyParam.offset })),
        ...assignmentErrors.map(error => ({ ...error, offset: itemParam.offset })),
      ],
    };
  },
  execute,
  examples: [
    {
      context: {
        array1: [{ a: 3 }, { a: 2 }],
        key: 'a',
        value: 2,
      },
      expression: 'FIND_BY_KEY(array1, key, value)',
      result: { a: 2 },
    },
    {
      context: {
        array: [{ a: 3 }, { a: { b: 10 } }],
        key: 'a',
        objValue: { b: 10 },
      },
      expression: 'FIND_BY_KEY(array, key, objValue)',
      result: { a: { b: 10 } },
    },
  ],
};
