// @sapappgyver/appgyver-formulas/functions/list/IS_IN_ARRAY.js
const execute = require('./IS_IN_ARRAY.runtime');

module.exports = {
  name: 'IS_IN_ARRAY',
  category: 'List',
  title: 'Is the given item in the source list?',
  description: 'Given a list and an item whose schema matches the list\'s schema, return `true` if exactly that item can be found in the list, or `false` if it is not found.',
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'List to check item',
    items: {
      $id: '#item', // Make this a generic type
    },
  }, {
    $ref: '#item', // Refer to the generic type
    title: 'Item',
    description: 'Item to check',
  }],
  analyze: () => ({ schema: { type: 'boolean' } }),
  execute,
  examples: [
    {
      expression: 'IS_IN_ARRAY(array, 3)',
      context: {
        array: [100, 3, 20],
      },
      result: true,
    },
    {
      expression: 'IS_IN_ARRAY(array, 1)',
      context: {
        array: [100, 3, 20],
      },
      result: false,
    },
    {
      expression: 'IS_IN_ARRAY([100, 3, { a: 10 }], { a: 10 })',
      result: true,
    },
  ],
};
