// @sapappgyver/appgyver-formulas/functions/list/REMOVE_ITEM.js
const { getArrayItemSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./REMOVE_ITEM.runtime');

module.exports = {
  name: 'REMOVE_ITEM',
  category: 'List',
  title: 'Removes a matching item from list',
  description: `Removes the first item from a list that matches the given criteria. Returns a list all other items except the matching one, in their corresponding order.

The criteria is defined as a formula at the second parameter. The formula is evaluated with each item in the list until a matching item is found.
The formula must result in a true/false value. If \`true\`, then item will be excluded from the result, otherwise it will be included.
Note that only first matching item is removed; if multiple items matches the criteria, then only first one is removed.
If no matching item is found, then the original list is returned unchanged.`,
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'List of items',
    items: {
      $id: '#item', // Make this a generic type
    },
  }, {
    type: 'function',
    title: 'Search formula',
    description: 'Formula to be evaluated for each item, deciding whether the item is the one to remove',
    parameters: [{
      name: 'item',
      title: 'Current item',
      description: 'The current item from the original list',
      $ref: '#item', // Refers to the generic type
    }, {
      name: 'index',
      type: 'number',
      minimum: 0,
      title: 'Current index',
      description: 'The position of the current item in the list as an index number starting from 0',
      examples: [0, 1, 2],
    }],
    returnValue: {
      type: 'boolean',
      title: 'Whether item is excluded',
      description: 'Whether or not the item should be removed from the result',
    },
  }],
  analyze: ([arrayParam]) => {
    const arraySchema = arrayParam && arrayParam.schema;
    // NOTE: Not using the parameter type directly, as it might contain enum or examples
    // that are no longer valid.
    const schema = { type: 'array' };
    if (arraySchema) {
      schema.items = getArrayItemSchema(arraySchema) || {};
    }
    return { schema };
  },
  execute,
  examples: [
    {
      expression: 'REMOVE_ITEM(animals, item.name == "rat")',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: [
        { name: 'dog', weight: 123 },
        { name: 'cat', weight: 48 },
        { name: 'pig', weight: 230 },
      ],
    },
    {
      expression: 'REMOVE_ITEM(animals, index == 1)',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
          { name: 'pig', weight: 230 },
        ],
      },
      result: [
        { name: 'dog', weight: 123 },
        { name: 'rat', weight: 10 },
        { name: 'pig', weight: 230 },
      ],
    },
    {
      expression: 'REMOVE_ITEM<product>(products, product.name == "Swiss army knife")',
      context: {
        products: [
          { name: 'Duct tape', category: 'Utility' },
          { name: 'Swiss army knife', category: 'Tools' },
          { name: 'Glue', category: 'Utility' },
        ],
      },
      result: [
        { name: 'Duct tape', category: 'Utility' },
        { name: 'Glue', category: 'Utility' },
      ],
    },
    {
      expression: 'REMOVE_ITEM<product, position>(products, position > 1 && product.category == "Utility")',
      context: {
        products: [
          { name: 'Duct tape', category: 'Utility' },
          { name: 'Swiss army knife', category: 'Tools' },
          { name: 'Glue', category: 'Utility' },
        ],
      },
      result: [
        { name: 'Duct tape', category: 'Utility' },
        { name: 'Swiss army knife', category: 'Tools' },
      ],
    },
  ],
};
