// @sapappgyver/appgyver-formulas/functions/list/REMOVE_ITEMS_AT.js
const { getArrayItemSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./REMOVE_ITEMS_AT.runtime');

module.exports = {
  name: 'REMOVE_ITEMS_AT',
  category: 'List',
  title: 'Removes a number of items from a list at an index',
  description: 'Return given list wherein removes the given number of items from a list at the given index. If the removal count parameter is omitted, then one (1) item is removed.',
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'Given list',
  }, {
    title: 'Start index',
    description: 'Start index at which the items will be removed',
    type: 'number',
  }, {
    title: 'Removal count',
    default: 1,
    minimum: 0,
    examples: [1, 2, 3],
    description: 'Number of items to remove at the given index.',
    type: 'number',
  }],
  analyze: ([arrayParam]) => {
    const schema = { type: 'array' };
    const itemSchema = arrayParam && getArrayItemSchema(arrayParam.schema);
    if (itemSchema) {
      schema.items = itemSchema;
    }
    return { schema };
  },
  execute,
  examples: [
    {
      context: {
        array: [{ a: 1 }, { a: 2 }, { a: 3 }],
      },
      expression: 'REMOVE_ITEMS_AT(array, 2, 1)',
      result: [{ a: 1 }, { a: 2 }],
    },
    {
      context: {
        array: [{ a: 1 }, { a: 2 }, { a: 3 }],
      },
      expression: 'REMOVE_ITEMS_AT(array, -10)',
      result: [{ a: 2 }, { a: 3 }],
    },
    {
      context: {
        array: [{ a: 1 }, { a: 2 }, { a: 3 }],
      },
      expression: 'REMOVE_ITEMS_AT(array, -1, 1)',
      result: [{ a: 1 }, { a: 2 }],
    },
    {
      context: {
        numbers: [1, 2, 3],
      },
      expression: 'REMOVE_ITEMS_AT(numbers, 1)',
      result: [1, 3]
      ,
    },
  ],
};
