// @sapappgyver/appgyver-formulas/functions/list/SELECT_BY_REGEX.js
const { resolveObjectMemberAccessor, getArrayItemSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./SELECT_BY_REGEX.runtime');

module.exports = {
  name: 'SELECT_BY_REGEX',
  category: 'List',
  title: 'Find objects from list by a key-regex match',
  description: `Finds multiple objects from the given list by a key-regex pair.

For example, given **Key** \`"name"\` and **Pattern** \`/Agnus/\`, will return an array containing all objects in the array whose \`name\` property contains \`Agnus\`, or an empty list if no matching objects were found.`,
  parameters: [{
    type: 'array',
    title: 'Array',
    description: 'List to find objects in',
    items: {
      type: 'object',
    },
  }, {
    title: 'Key',
    description: 'Key whose value to match',
    type: 'string',
  }, {
    title: 'Pattern',
    description: 'Regex pattern to match',
    type: 'string',
  }],
  analyze: ([arrayParam, keyParam], offset, parameterErrors) => {
    const arraySchema = (arrayParam && arrayParam.schema) || { type: 'array' };
    const itemsSchema = getArrayItemSchema(arraySchema) || { type: 'object' };
    const keySchema = (keyParam && keyParam.schema) || {};
    const {
      errors: keyErrors,
    } = resolveObjectMemberAccessor(itemsSchema, keySchema);
    return {
      schema: arraySchema,
      errors: parameterErrors.length ? [] : keyErrors.map(error => ({
        ...error, offset: keyParam.offset,
      })),
    };
  },
  execute,
  examples: [
    {
      expression: 'SELECT_BY_REGEX(array1, "a", "^foo")',
      context: {
        array1: [{ a: 2 }, { a: 'foo bar' }, { a: 'foo', b: 3 }],
      },
      result: [{ a: 'foo bar' }, { a: 'foo', b: 3 }],
    },
    {
      expression: 'SELECT_BY_REGEX(array1, "a", "^foo")',
      context: {
        array1: [{ a: 2 }, { a: 'foo bar' }, { a: 'foo', b: 3 }],
      },
      result: [{ a: 'foo bar' }, { a: 'foo', b: 3 }],
    },
  ],
};
