// @sapappgyver/appgyver-formulas/functions/list/SLICE.js
const { getArrayItemSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./SLICE.runtime');

module.exports = {
  name: 'SLICE',
  category: 'List',
  title: 'Select a range of items from list',
  description: `Selects the items starting at the given **Start** index and ending at the given **End** index.

Note that the element at the given end index is *not* included.

You can use a negative end index to select from the end of the list.`,
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'List to slice',
    inversible: true,
  }, {
    title: 'Start',
    description: 'Index of the first item',
    type: 'number',
  }, {
    title: 'End',
    description: 'Index where the range ends, not including item at this index. Negative index will be calculated from the end of the list.',
    type: 'number',
    default: null,
  }],
  maxInversions: 1,
  analyze: ([arrayParam]) => {
    const schema = { type: 'array' };
    const itemSchema = arrayParam && getArrayItemSchema(arrayParam.schema);
    if (itemSchema) {
      schema.items = itemSchema;
    }
    return { schema };
  },
  execute,
  examples: [
    {
      expression: 'SLICE(["a", "b", "c", "d"], 1, 3)',
      result: ['b', 'c'],
    },
    {
      expression: 'SLICE(["a", "b", "c", "d"], 0, -1)',
      result: ['a', 'b', 'c'],
    },
    {
      expression: 'SLICE(["a", "b", "c", "d"], 2)',
      result: ['c', 'd'],
    },
  ],
};
