// @sapappgyver/appgyver-formulas/functions/list/SORT.js
const execute = require('./SORT.runtime');

module.exports = {
  name: 'SORT',
  category: 'List',
  title: 'Sort a list of texts or numbers',
  description: 'Returns an array containing all the values in the given list of texts or numbers, sorted in order. The function ignores letter case.',
  parameters: [{
    title: 'List',
    description: 'List to sort',
    anyOf: [
      {
        type: 'array',
        items: { type: 'number' },
      },
      {
        type: 'array',
        items: { type: 'string' },
      },
    ],
  }, {
    type: 'string',
    title: 'Order',
    default: 'asc',
    enum: ['asc', 'desc'],
    description: 'Order to sort in (`"asc"` for ascending, `"desc"` for descending)',
  }],
  analyze: ([arrayParam]) => {
    const arraySchema = (arrayParam && arrayParam.schema) || { type: 'array' };
    return { schema: arraySchema };
  },
  execute,
  examples: [
    {
      expression: 'SORT(array)',
      context: {
        array: [10, 1, 2, 6, -1, 10, 12],
      },
      result: [-1, 1, 2, 6, 10, 10, 12],
    },
    {
      expression: 'SORT(array, "desc")',
      context: {
        array: [10, 1, 2, 6, -1, 10, 12],
      },
      result: [12, 10, 10, 6, 2, 1, -1],
    },
    {
      expression: 'SORT(textArray)',
      context: {
        textArray: ['ABCD', 'CC', 'bb', 'AA', 'a', 'aa', 'abcd'],
      },
      result: ['a', 'AA', 'aa', 'ABCD', 'abcd', 'bb', 'CC'],
    },
  ],
};
