// @sapappgyver/appgyver-formulas/functions/list/SORT_CASE_SENSITIVE.js
const execute = require('./SORT_CASE_SENSITIVE.runtime');

module.exports = {
  name: 'SORT_CASE_SENSITIVE',
  category: 'List',
  title: 'Sort a list of texts or numbers (case-sensitive)',
  description: `Returns a list containing all the values in the given list of texts or numbers, sorted in order.
  
This function performs a **case-sensitive** sort, meaning that all the upper case letters follow all the lower case letters.
This is for advanced use only, because typically you would like to perfrom a case insensitive sort, so unless you have
a reason to sort case-sensitively then you should use \`SORT\` instead!`,
  parameters: [{
    title: 'List',
    description: 'List to sort',
    anyOf: [
      {
        type: 'array',
        items: { type: 'number' },
      },
      {
        type: 'array',
        items: { type: 'string' },
      },
    ],
  }, {
    type: 'string',
    title: 'Order',
    default: 'asc',
    enum: ['asc', 'desc'],
    description: 'Order to sort in (`"asc"` for ascending, `"desc"` for descending)',
  }],
  analyze: ([arrayParam]) => {
    const arraySchema = (arrayParam && arrayParam.schema) || { type: 'array' };
    return { schema: arraySchema };
  },
  execute,
  examples: [
    {
      expression: 'SORT_CASE_SENSITIVE(textArray)',
      context: {
        textArray: ['ABCD', 'CC', 'bb', 'AA', 'a', 'aa', 'abcd'],
      },
      result: ['AA', 'ABCD', 'CC', 'a', 'aa', 'abcd', 'bb'],
    },
    {
      expression: 'SORT_CASE_SENSITIVE(array)',
      context: {
        array: [10, 1, 2, 6, -1, 10, 12],
      },
      result: [-1, 1, 2, 6, 10, 10, 12],
    },
    {
      expression: 'SORT_CASE_SENSITIVE(array, "desc")',
      context: {
        array: [10, 1, 2, 6, -1, 10, 12],
      },
      result: [12, 10, 10, 6, 2, 1, -1],
    },
  ],
};
