// @sapappgyver/appgyver-formulas/functions/list/SUM_BY_KEY.js
const { resolveObjectMemberAccessor, getArrayItemSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./SUM_BY_KEY.runtime');

module.exports = {
  name: 'SUM_BY_KEY',
  category: 'List',
  title: 'Calculate the sum of numeric key values of objects',
  description: 'Calculates the sum of all numeric key values of objects in the given list. Returns `0` if the list is empty or all key values are non-numeric.',
  parameters: [{
    type: 'array',
    title: 'Objects',
    description: 'List of objects',
    items: {
      type: 'object',
    },
  }, {
    type: 'string',
    title: 'Key name',
    description: 'Name of the key whose values to sum',
  }],
  analyze: ([arrayParam, propNameParam], offset, parameterErrors) => {
    const itemSchema = getArrayItemSchema(arrayParam && arrayParam.schema) || {};
    const accessorSchema = (propNameParam && propNameParam.schema) || {};
    const { errors } = resolveObjectMemberAccessor(itemSchema, accessorSchema);
    return {
      schema: { type: 'number' },
      errors: parameterErrors.length ? [] : errors.map(error => ({
        ...error, offset: propNameParam.offset,
      })),
    };
  },
  execute,
  examples: [
    {
      expression: 'SUM_BY_KEY(products, "price")',
      context: {
        products: [
          { name: 'Duct tape', price: 5 },
          { name: 'Swiss army knife', price: 60 },
        ],
      },
      result: 65,
    },
    {
      expression: 'SUM_BY_KEY(animals, "weight")',
      context: {
        animals: [
          { name: 'dog', weight: 123 },
          { name: 'cat', weight: 48 },
          { name: 'rat', weight: 10 },
        ],
      },
      result: 123 + 48 + 10,
    },
  ],
};
