// @sapappgyver/appgyver-formulas/functions/list/WITHOUT_ITEM.js
const execute = require('./WITHOUT_ITEM.runtime');

module.exports = {
  name: 'WITHOUT_ITEM',
  category: 'List',
  title: 'Remove item from list',
  description: 'Returns a list which doesn\'t contain the given item.',
  searchNames: ['REMOVE'],
  parameters: [{
    type: 'array',
    title: 'List',
    description: 'List to remove item from',
    inversible: true,
    items: {
      $id: '#item', // Make this a generic type
    },
  }, {
    $ref: '#item', // Refer to the generic type
    title: 'Item',
    description: 'Item to remove',
  }],
  maxInversions: 1,
  analyze: ([arrayParam]) => {
    const arraySchema = (arrayParam && arrayParam.schema) || { type: 'array' };
    return {
      schema: arraySchema,
    };
  },
  execute,
  examples: [
    { expression: 'WITHOUT_ITEM([100, 3, 20], 3)', result: [100, 20] },
    { expression: 'WITHOUT_ITEM([[1, 2], [3, 4], [1, 2]], [1, 2])', result: [[3, 4], [1, 2]] },
    { expression: 'WITHOUT_ITEM([1, "2", 3, { a: 4 }], 3)', result: [1, '2', { a: 4 }] },
  ],
};
