// @sapappgyver/appgyver-formulas/functions/statistical/AVERAGE_WEIGHTED.js
const execute = require('./AVERAGE_WEIGHTED.runtime');

module.exports = {
  name: 'AVERAGE_WEIGHTED',
  category: 'Statistical',
  title: 'Weighted average',
  description: 'Given a list of numbers and a matching list of weights, calculates the numbers\' weighted average.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List of numeric values',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'Weights',
    description: 'List of weights',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 4],
        weights: [0.5, 0.5, 0.25],
      },
      expression: 'AVERAGE_WEIGHTED(numbers, weights)',
      result: 2,
    },
    {
      context: {
        numbersLarge: [1, 2, 4, 5, 6],
        weightsLarge: [0.5, 0.5, 0.25, 0.3, 0.4],
      },
      expression: 'AVERAGE_WEIGHTED(numbersLarge, weightsLarge)',
      result: 3.28205128,
    },
  ],
};
