// @sapappgyver/appgyver-formulas/functions/statistical/BETA_DIST.js
const execute = require('./BETA_DIST.runtime');

module.exports = {
  name: 'BETA_DIST',
  category: 'Statistical',
  title: 'Beta distribution',
  description: 'Calculates the probability of the given value using the beta distribution function. Optionally, you can use the cumulative distribution function.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The value at which to evaluate the probability function.',
  }, {
    type: 'number',
    title: 'Alpha',
    description: 'The first shape parameter of the distribution.',
  }, {
    type: 'number',
    title: 'Beta',
    description: 'The second shape parameter of the distribution.',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'BETA_DIST(0, 2, 3)', result: 0 },
    { expression: 'BETA_DIST(0.1, 2, 3)', result: 0.972 },
    { expression: 'BETA_DIST(0.1, 2, 3, false)', result: 0.972 },
    { expression: 'BETA_DIST(1, 2, 3)', result: 0 },
    { expression: 'BETA_DIST(0, 2, 3, true)', result: 0 },
    { expression: 'BETA_DIST(0.5, 2, 3, true)', result: 0.6875 },
    { expression: 'BETA_DIST(1, 2, 3, true)', result: 1 },
  ],
};
