// @sapappgyver/appgyver-formulas/functions/statistical/BETA_INV.js
const execute = require('./BETA_INV.runtime');

module.exports = {
  name: 'BETA_INV',
  category: 'Statistical',
  title: 'Beta inverse distribution',
  description: 'Given a probability, calculates the value of the inverse beta distribution function.',
  parameters: [{
    type: 'number',
    title: 'Probability',
    description: 'The probability at which to evaluate the function.',
  }, {
    type: 'number',
    title: 'Alpha',
    description: 'The first shape parameter of the distribution.',
  }, {
    type: 'number',
    title: 'Beta',
    description: 'The second shape parameter of the distribution.',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'BETA_INV(0.1, 2, 3)', result: 0.14255932 },
    { expression: 'BETA_INV(0.5, 2, 3)', result: 0.38572757 },
    { expression: 'BETA_INV(0.9, 2, 3)', result: 0.67953942 },
    { expression: 'BETA_INV(1, 2, 3)', result: 1 },
    { expression: 'BETA_INV(0.5, 2, 4)', result: 0.31381017 },
    { expression: 'BETA_INV(0.5, 3, 3)', result: 0.50000004 },
  ],
};
