// @sapappgyver/appgyver-formulas/functions/statistical/BINOMDIST.js
const execute = require('./BINOMDIST.runtime');

module.exports = {
  name: 'BINOMDIST',
  category: 'Statistical',
  title: 'Binomial distribution probability',
  description: 'Calculates the individual term binomial distribution probability.',
  parameters: [{
    type: 'number',
    title: 'Number successes',
    description: 'The number of successes for which to calculate the probability in number trials trials.',
  }, {
    type: 'number',
    title: 'Number trials',
    description: 'The number of independent trials',
  }, {
    type: 'number',
    title: 'Number probability of success',
    description: 'The probability of success in any given trial.',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the binomial cumulative distribution',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'BINOMDIST(9, 20, 0.5)', result: 0.16017914 },
    { expression: 'BINOMDIST(9, 20, 0.5, false)', result: 0.16017914 },
    { expression: 'BINOMDIST(9, 20, 0.5, true)', result: 0.41190147 },
  ],
};
