// @sapappgyver/appgyver-formulas/functions/statistical/CHI_TEST.js
const execute = require('./CHI_TEST.runtime');

module.exports = {
  name: 'CHI_TEST',
  category: 'Statistical',
  title: 'Chi test',
  description: 'Calculates a Pearson\'s chi-squared test on two sets of data\'s associated right-tailed probability.',
  parameters: [{
    type: 'array',
    title: 'Observed range',
    description: 'The counts associated with each category of data.',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'Expected range',
    description: 'The expected counts for each category under the null hypothesis.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 4],
        numbers2: [1, 3, 9],
      },
      expression: 'CHI_TEST(numbers, numbers2)',
      result: 0.21107209,
    },
    {
      context: {
        numbers: [1, 2, 4],
      },
      expression: 'CHI_TEST(numbers, numbers)',
      result: 1,
    },
    {
      context: {
        numbers: [1, 2, 4],
        numbersReverse: [4, 2, 1],
      },
      expression: 'CHI_TEST(numbers, numbersReverse)',
      result: 0.00360656,
    },
    {
      context: {
        numbersLarge: [1, 2, 4, 5, 6],
        numbersNoOrder: [23, 19, 50, 1, 1],
      },
      expression: 'CHI_TEST(numbersLarge, numbersNoOrder)',
      result: 0,
    },
  ],
};
