// @sapappgyver/appgyver-formulas/functions/statistical/CHISQ_DIST.js
const execute = require('./CHISQ_DIST.runtime');

module.exports = {
  name: 'CHISQ_DIST',
  category: 'Statistical',
  title: 'Chi-squared distribution',
  description: 'Calculates the chi-squared distribution (left-tailed).',
  parameters: [{
    type: 'number',
    title: 'Value',
    description: 'The input to the chi-squared probability distribution function. '
      + 'The value at which to evaluate the function. Must be a positive number.',
  }, {
    type: 'number',
    title: 'Degrees freedom',
    description: 'The number of degrees of freedom of the distribution',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'CHISQ_DIST(1, 1)', result: 0.24197072 },
    { expression: 'CHISQ_DIST(1, 1, false)', result: 0.24197072 },
    { expression: 'CHISQ_DIST(1, 2)', result: 0.30326533 },
    { expression: 'CHISQ_DIST(1, 1, true)', result: 0.68268949 },
    { expression: 'CHISQ_DIST(1, 2, true)', result: 0.39346934 },
  ],
};
