// @sapappgyver/appgyver-formulas/functions/statistical/CORREL.js
const execute = require('./CORREL.runtime');

module.exports = {
  name: 'CORREL',
  category: 'Statistical',
  title: 'Correlation coefficient',
  description: 'Given two lists of numbers as a dataset, calculates their r, or the Pearson product moment correlation coefficient.',
  parameters: [{
    type: 'array',
    title: 'Y dataset',
    description: 'List of dependent data.',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'X dataset',
    description: 'List of independent data.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 4],
        numbers2: [1, 3, 9],
      },
      expression: 'CORREL(numbers, numbers2)',
      result: 0.99587059,
    },
    {
      context: {
        numbers: [1, 2, 4],
        numbersReverse: [4, 2, 1],
      },
      expression: 'CORREL(numbers, numbersReverse)',
      result: -0.92857143,
    },
    {
      context: {
        numbers2: [1, 3, 9],
      },
      expression: 'CORREL(numbers2, numbers2)',
      result: 1,
    },
    {
      context: {
        numbersLarge: [1, 2, 4, 5, 6],
        numbersNoOrder: [23, 19, 50, 1, 1],
      },
      expression: 'CORREL(numbersLarge, numbersNoOrder)',
      result: -0.39719512,
    },
  ],
};
