// @sapappgyver/appgyver-formulas/functions/statistical/COVAR_P.js
const execute = require('./COVAR_P.runtime');

module.exports = {
  name: 'COVAR_P',
  category: 'Statistical',
  title: 'Covariant for population',
  description: 'Given two lists of numbers as a dataset, calculates the population covariance.',
  parameters: [{
    type: 'array',
    title: 'Y dataset',
    description: 'List of dependent data.',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'X dataset',
    description: 'List of independent data.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 4],
        numbers2: [1, 3, 9],
      },
      expression: 'COVAR_P(numbers, numbers2)',
      result: 4.22222222,
    },
    {
      context: {
        numbers: [1, 2, 4],
        numbersReverse: [4, 2, 1],
      },
      expression: 'COVAR_P(numbers, numbersReverse)',
      result: -1.44444444,
    },
    {
      context: {
        numbers2: [1, 3, 9],
      },
      expression: 'COVAR_P(numbers2, numbers2)',
      result: 11.55555556,
    },
    {
      context: {
        numbersLarge: [1, 2, 4, 5, 6],
        numbersNoOrder: [23, 19, 50, 1, 1],
      },
      expression: 'COVAR_P(numbersLarge, numbersNoOrder)',
      result: -13.28,
    },
  ],
};
