// @sapappgyver/appgyver-formulas/functions/statistical/COVAR.js
const execute = require('./COVAR.runtime');

module.exports = {
  name: 'COVAR',
  category: 'Statistical',
  title: 'Covariant for sample',
  description: 'Given two lists of numbers as a dataset, calculates the sample covariance.',
  parameters: [{
    type: 'array',
    title: 'Y dataset',
    description: 'List of dependent data.',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'X dataset',
    description: 'List of independent data.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 4],
        numbers2: [1, 3, 9],
      },
      expression: 'COVAR(numbers, numbers2)',
      result: 6.33333333,
    },
    {
      context: {
        numbers: [1, 2, 4],
        numbersReverse: [4, 2, 1],
      },
      expression: 'COVAR(numbers, numbersReverse)',
      result: -2.16666667,
    },
    {
      context: {
        numbers2: [1, 3, 9],
      },
      expression: 'COVAR(numbers2, numbers2)',
      result: 17.33333333,
    },
    {
      context: {
        numbersLarge: [1, 2, 4, 5, 6],
        numbersNoOrder: [23, 19, 50, 1, 1],
      },
      expression: 'COVAR(numbersLarge, numbersNoOrder)',
      result: -16.6,
    },
  ],
};
