// @sapappgyver/appgyver-formulas/functions/statistical/EXPON_DIST.js
const execute = require('./EXPON_DIST.runtime');

module.exports = {
  name: 'EXPON_DIST',
  category: 'Statistical',
  title: 'Exponential distribution',
  description: 'Given a specific lambda, calculates the value of the exponential distribution function at the given number.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the exponential distribution function.',
  }, {
    type: 'number',
    title: 'Lambda',
    description: 'The lambda to specify the exponential distribution function',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'EXPON_DIST(0, 2)', result: 2 },
    { expression: 'EXPON_DIST(1, 2)', result: 0.27067057 },
    { expression: 'EXPON_DIST(1, 2, false)', result: 0.27067057 },
    { expression: 'EXPON_DIST(0, 2, true)', result: 0 },
    { expression: 'EXPON_DIST(1, 2, true)', result: 0.86466472 },
  ],
};
