// @sapappgyver/appgyver-formulas/functions/statistical/EXPON_INV.js
const execute = require('./EXPON_INV.runtime');

module.exports = {
  name: 'EXPON_INV',
  category: 'Statistical',
  title: 'Exponential distribution, inverse',
  description: 'Given a specific lambda, calculates the value of the inverse exponential distribution function at the given number.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the exponential distribution function.',
  }, {
    type: 'number',
    title: 'Lambda',
    description: 'The lambda to specify the exponential distribution function',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'EXPON_INV(0.001, 2)', result: 0.00050025 },
    { expression: 'EXPON_INV(0.1, 2)', result: 0.05268026 },
    { expression: 'EXPON_INV(0.5, 2)', result: 0.34657359 },
    { expression: 'EXPON_INV(0.9, 2)', result: 1.15129255 },
    { expression: 'EXPON_INV(0.999, 2)', result: 3.45387764 },
    { expression: 'EXPON_INV(0.5, 3)', result: 0.23104906 },
  ],
};
