// @sapappgyver/appgyver-formulas/functions/statistical/F_DIST.js
const execute = require('./F_DIST.runtime');

module.exports = {
  name: 'F_DIST',
  category: 'Statistical',
  title: 'F distribution, left-tailed',
  description: 'Calculates F probability distribution (left-tailed).',
  parameters: [{
    type: 'number',
    title: 'Value',
    description: 'The input to the F probability distribution function. Must be a positive number.',
  }, {
    type: 'number',
    title: 'Degrees of freedom 1',
    description: 'The numerator degrees of freedom',
  }, {
    type: 'number',
    title: 'Degrees of freedom 2',
    description: 'The denominator degrees of freedom',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'F_DIST(1, 1, 1)', result: 0.1591549431 },
    { expression: 'F_DIST(1, 1, 1, false)', result: 0.1591549431 },
    { expression: 'F_DIST(1, 1, 2)', result: 0.1924500897 },
    { expression: 'F_DIST(1, 1, 2, true)', result: 0.5773502692 },
  ],
};
