// @sapappgyver/appgyver-formulas/functions/statistical/F_INV_RT.js
const execute = require('./F_INV_RT.runtime');

module.exports = {
  name: 'F_INV_RT',
  category: 'Statistical',
  title: 'Inverse F distribution, right-tailed',
  description: 'Calculates the inverse of the F probability distribution (right-tailed).',
  parameters: [{
    type: 'number',
    title: 'Probability',
    description: 'The probability associated with the right-tailed F-distribution.',
  }, {
    type: 'number',
    title: 'Degrees of freedom 1',
    description: 'The numerator degrees of freedom',
  }, {
    type: 'number',
    title: 'Degrees of freedom 2',
    description: 'The denominator degrees of freedom',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'F_INV_RT(0.01, 2, 3)', result: 30.81652035 },
    { expression: 'F_INV_RT(0.5, 2, 3)', result: 0.88110154 },
    { expression: 'F_INV_RT(0.99, 2, 3)', result: 0.01008408093 },
    { expression: 'F_INV_RT(0.5, 2, 4)', result: 0.82842712 },
    { expression: 'F_INV_RT(0.5, 3, 4)', result: 0.94053408 },
  ],
};
