// @sapappgyver/appgyver-formulas/functions/statistical/F_TEST.js
const execute = require('./F_TEST.runtime');

module.exports = {
  name: 'F_TEST',
  category: 'Statistical',
  title: 'F-test',
  description: 'Given two lists of numbers, calculates the probability associated with an F-test for the equality of variances.',
  parameters: [{
    type: 'array',
    title: 'Sample 1',
    description: 'The first sample for the F-test',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'Sample 2',
    description: 'The second sample for the F-test',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'F_TEST([1, 2, 3, 4, 5], [1, 2, 2, 2, 6])', result: 0.6948693646 },
    { expression: 'F_TEST([1, 2, 2, 2, 6], [1, 2, 3, 4, 5])', result: 0.6948693646 },
    { expression: 'F_TEST([1, 2, 3, 4, 5], [1, 5, 2, 2])', result: 0.83322082 },
    { expression: 'F_TEST([1, 5, 2, 2], [1, 2, 3, 4, 5])', result: 0.83322082 },
    { expression: 'F_TEST([1, 2, 3, 4, 5], [1, 2, 3, 4, 5])', result: 1 },
  ],
};
