// @sapappgyver/appgyver-formulas/functions/statistical/FORECAST.js
const execute = require('./FORECAST.runtime');

module.exports = {
  name: 'FORECAST',
  category: 'Statistical',
  title: 'Calculate forecast based on a linear regression',
  description: 'Given the dependent and independent datasets a list of numbers, calculates the expected y-value for the given input x using linear regression.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the linear regression function',
  }, {
    type: 'array',
    title: 'Dependant data',
    description: 'Dependant data as a list of numbers.',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'Independent data',
    description: 'Independent data as a list of numbers.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'FORECAST(3, [1, 2, 3], [1, 3, 3])', result: 2.5 },
    { expression: 'FORECAST(3, [1, 3, 3], [1, 2, 3])', result: 3.33333333 },
    { expression: 'FORECAST(20, [1, 3, 5, 10], [2, 5, 8, 15])', result: 13.41935484 },
  ],
};
