// @sapappgyver/appgyver-formulas/functions/statistical/GAMMA_DIST.js
const execute = require('./GAMMA_DIST.runtime');

module.exports = {
  name: 'GAMMA_DIST',
  category: 'Statistical',
  title: 'Gamma distribution',
  description: 'Calculates the value of the Gamma distribution function for the given input number.',
  parameters: [{
    type: 'number',
    title: 'Value',
    description: 'The input to the gamma probability distribution function (X). The value at which to evaluate the function.',
  }, {
    type: 'number',
    title: 'Shape',
    description: 'The shape parameter of the distribution (Alpha).',
  }, {
    type: 'number',
    title: 'Scale',
    description: 'The second parameter of the distribution (Beta).',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'GAMMA_DIST(1, 1, 1)', result: 0.36787944 },
    { expression: 'GAMMA_DIST(1, 1, 1, false)', result: 0.36787944 },
    { expression: 'GAMMA_DIST(1, 1, 2)', result: 0.30326533 },
    { expression: 'GAMMA_DIST(1, 1, 2, true)', result: 0.39346934 },
  ],
};
