// @sapappgyver/appgyver-formulas/functions/statistical/HYPGEOM_DIST.js
const execute = require('./HYPGEOM_DIST.runtime');

module.exports = {
  name: 'HYPGEOM_DIST',
  category: 'Statistical',
  title: 'Hypergeometric distribution',
  description: 'Given a population size and the number of successes in the population, calculates the probability of drawing a given number of successes in the given number of tries. The trial is performed without replacing draws.',
  parameters: [{
    type: 'number',
    title: 'Number of successes',
    description: 'The desired number of successes',
  }, {
    type: 'number',
    title: 'Number of draws',
    description: 'The number of permitted draws',
  }, {
    type: 'number',
    title: 'Successes in population',
    description: 'The total number of successes in the population',
  }, {
    type: 'number',
    title: 'Population size',
    description: 'The total size of the population',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'HYPGEOM_DIST(1, 1, 1, 1)', result: 1 },
    { expression: 'HYPGEOM_DIST(1, 2, 3, 4)', result: 0.5 },
    { expression: 'HYPGEOM_DIST(1, 2, 3, 5)', result: 0.6 },
    { expression: 'HYPGEOM_DIST(1, 2, 4, 5)', result: 0.4 },
    { expression: 'HYPGEOM_DIST(1, 3, 3, 5)', result: 0.3 },
    { expression: 'HYPGEOM_DIST(2, 3, 3, 5)', result: 0.6 },
  ],
};
