// @sapappgyver/appgyver-formulas/functions/statistical/INTERCEPT.js
const execute = require('./INTERCEPT.runtime');

module.exports = {
  name: 'INTERCEPT',
  category: 'Statistical',
  title: 'Calculate intercept based on a linear regression',
  description: 'Given two lists of numbers, representing the dependent and independent data, performs linear regression on the dataset, calculating the y-value at which the resultant line will intersect the y-axis.',
  parameters: [{
    type: 'array',
    title: 'Dependent data',
    description: 'Dependent data as a list of numbers.',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'Independent data',
    description: 'Independent data as a list of numbers.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'INTERCEPT([1, 2, 3], [1, 3, 3])', result: 0.25 },
    { expression: 'INTERCEPT([1, 3, 3], [1, 2, 3])', result: 0.33333333 },
    { expression: 'INTERCEPT([1, 3, 5, 10], [2, 5, 8, 15])', result: -0.4516129 },
  ],
};
