// @sapappgyver/appgyver-formulas/functions/statistical/LOGNORM_DIST.js
const execute = require('./LOGNORM_DIST.runtime');

module.exports = {
  name: 'LOGNORM_DIST',
  category: 'Statistical',
  title: 'Log-normal distribution',
  description: 'Given a mean and standard deviation, calculates the value of log-normal distribution at the given value.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the log-normal distribution function.',
  }, {
    type: 'number',
    title: 'Mean',
    description: 'The mean (mu) of the distribution function',
  }, {
    type: 'number',
    title: 'Standard deviation',
    description: 'The standard deviation (sigma) of the distribution function',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'LOGNORM_DIST(1, 2, 3)', result: 0.10648267 },
    { expression: 'LOGNORM_DIST(1, 2, 3, false)', result: 0.10648267 },
    { expression: 'LOGNORM_DIST(1, 2, 3, true)', result: 0.25249254 },
    { expression: 'LOGNORM_DIST(2, 2, 3, true)', result: 0.33155710 },
  ],
};
