// @sapappgyver/appgyver-formulas/functions/statistical/LOGNORM_INV.js
const execute = require('./LOGNORM_INV.runtime');

module.exports = {
  name: 'LOGNORM_INV',
  category: 'Statistical',
  title: 'Log-normal inverse distribution',
  description: 'Given a mean and standard deviation, calculates the value of the inverse log-normal distribution at the given value.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the inverse log-normal cumulative distribution function.',
  }, {
    type: 'number',
    title: 'Mean',
    description: 'The mean (mu) of the distribution function',
  }, {
    type: 'number',
    title: 'Standard deviation',
    description: 'The standard deviation (sigma) of the distribution function',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'LOGNORM_INV(0.01, 2, 3)', result: 0.00688064 },
    { expression: 'LOGNORM_INV(0.1, 2, 3)', result: 0.15807990 },
    { expression: 'LOGNORM_INV(0.5, 2, 3)', result: 7.38905610 },
    { expression: 'LOGNORM_INV(0.9, 2, 3)', result: 345.38325919 },
    { expression: 'LOGNORM_INV(0.99, 2, 3)', result: 7935.0395891 },
  ],
};
