// @sapappgyver/appgyver-formulas/functions/statistical/MODE.js
const execute = require('./MODE.runtime');

module.exports = {
  name: 'MODE',
  category: 'Statistical',
  title: 'Calculate mode of a list',
  description: 'Given a list of numbers, calculates their mode.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List of numeric values',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [2, 4, 6],
      },
      expression: 'MODE(numbers)',
      result: null,
    },
    {
      context: {
        mode2Numbers: [1, 1, 2, 2, 4, 6, 2, 2],
      },
      expression: 'MODE(mode2Numbers)',
      result: 2,
    },
    {
      context: {
        mode1and2Numbers: [1, 1, 1, 1, 2, 2, 4, 6, 2, 2],
      },
      expression: 'MODE(mode1and2Numbers)',
      result: 1,
    },
    {
      context: {
        mode2and1Numbers: [2, 2, 2, 1, 1, 1],
      },
      expression: 'MODE(mode2and1Numbers)',
      result: 1,
    },
    {
      context: {
        singleNumber: [1],
      },
      expression: 'MODE(singleNumber)',
      result: null,
    },
    {
      context: {
        floats: [1.1, 1.2, 1.1, 1.1, 0.99],
      },
      expression: 'MODE(floats)',
      result: 1.1,
    },
  ],
};
