// @sapappgyver/appgyver-formulas/functions/statistical/NEG_BINOM_DIST.js
const execute = require('./NEG_BINOM_DIST.runtime');

module.exports = {
  name: 'NEG_BINOM_DIST',
  category: 'Statistical',
  title: 'Calculate negative binomial distribution',
  description: 'Given the probability of success in independent trials, calculates the probability that a certain number of failures will be reached before a certain number of successes is reached.',
  parameters: [{
    title: 'Num failures',
    description: 'The number of failures to model.',
    type: 'number',
  }, {
    title: 'Num successes',
    description: 'The number of successes to model.',
    type: 'number',
  }, {
    title: 'Prob success',
    description: 'The probability of success in any given trial.',
    type: 'number',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'NEG_BINOM_DIST(5, 3, 0.4)', result: 0.07741440 },
    { expression: 'NEG_BINOM_DIST(2, 5, 0.2)', result: 0.07864320 },
    { expression: 'NEG_BINOM_DIST(4, 5, 0.4)', result: 0.11147674 },
    { expression: 'NEG_BINOM_DIST(10, 15, 0.3)', result: 0.03665440 },
  ],
};
