// @sapappgyver/appgyver-formulas/functions/statistical/NORM_INV.js
const execute = require('./NORM_INV.runtime');

module.exports = {
  name: 'NORM_INV',
  category: 'Statistical',
  title: 'Inverse normal distribution',
  description: `Calculates the value of the inverse normal distribution function for the given number input, mean, and standard deviation.

Optionally, the cumulative distribution function can be used.`,
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input of the inverse normal distribution function',
  }, {
    type: 'number',
    title: 'Mean',
    description: 'The mean (mu) of the normal distribution function',
  }, {
    type: 'number',
    title: 'Standard deviation',
    description: 'The standard deviation (sigma) of the normal distribution function',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'NORM_INV(0.997, 15, 2.5)', result: 21.86945346 },
    { expression: 'NORM_INV(0.23, 500, 5)', result: 496.30576575 },
    { expression: 'NORM_INV(0.977, 500, 5)', result: 509.97696655 },
  ],
};
