// @sapappgyver/appgyver-formulas/functions/statistical/NORM_DIST.js
const execute = require('./NORM_DIST.runtime');

module.exports = {
  name: 'NORM_DIST',
  category: 'Statistical',
  title: 'Normal distribution',
  description: `Calculates the value of the normal distribution function for the given number input, mean, and standard deviation.

Optionally, the cumulative distribution function can be used.`,
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the normal distribution function',
  }, {
    type: 'number',
    title: 'Mean',
    description: 'The mean (mu) of the normal distribution function',
  }, {
    type: 'number',
    title: 'Standard deviation',
    description: 'The standard deviation (sigma) of the normal distribution function',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the normal cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'NORM_DIST(22, 15, 2.5)', result: 0.00316618 },
    { expression: 'NORM_DIST(22, 15, 2.5, false)', result: 0.00316618 },
    { expression: 'NORM_DIST(22, 15, 2.5, true)', result: 0.99744487 },
  ],
};
