// @sapappgyver/appgyver-formulas/functions/statistical/NTH_SMALLEST.js
const execute = require('./NTH_SMALLEST.runtime');

module.exports = {
  name: 'NTH_SMALLEST',
  category: 'Statistical',
  title: 'Nth smallest number in list',
  description: 'Given a list of numbers, returns the nth smallest number.',
  parameters: [{
    title: 'List',
    description: 'List or range containing the dataset to consider',
    type: 'array',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Rank',
    description: 'The rank from smallest to smallest of the element to return.',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      expression: 'NTH_SMALLEST(numbers1, 1)',
      context: {
        numbers1: [10, 1, 2, 6, -1, 10, 12],
      },
      result: -1,
    }, {
      expression: 'NTH_SMALLEST(numbers2, 1)',
      context: {
        numbers2: [29, 44, 61, 1, 36, -3, 14, 10, -24],
      },
      result: -24,
    }, {
      expression: 'NTH_SMALLEST(numbers2, 9)',
      context: {
        numbers2: [29, 44, 61, 1, 36, -3, 14, 10, -24],
      },
      result: 61,
    },
  ],
};
