// @sapappgyver/appgyver-formulas/functions/statistical/PERCENTILE_EXC.js
const execute = require('./PERCENTILE_EXC.runtime');

module.exports = {
  name: 'PERCENTILE_EXC',
  category: 'Statistical',
  title: 'Percentile exclusive of dataset',
  description: `Given a list of numbers and a percentile value (a decimal greater than 0 and smaller than 1), returns the value at that percentile.

If a specific value is not found at the exact percentile given, the value is estimated based on other numbers in the list.`,
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List of numeric values',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Percentile',
    description: 'The percentile whose value within data will be calculated and returned ( 0 <= percentile <= 1).',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'PERCENTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 0.1)', result: 1.1 },
    { expression: 'PERCENTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 0.5)', result: 5.5 },
    { expression: 'PERCENTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 0.9)', result: 9.9 },
    { expression: 'PERCENTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9], 0.1)', result: 1 },
    { expression: 'PERCENTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9], 0.5)', result: 5 },
    { expression: 'PERCENTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9], 0.9)', result: 9 },
  ],
};
