// @sapappgyver/appgyver-formulas/functions/statistical/PERCENTILE.js
const execute = require('./PERCENTILE.runtime');

module.exports = {
  name: 'PERCENTILE',
  category: 'Statistical',
  title: 'Percentile of dataset',
  description: `Given a list of numbers and a percentile value (a decimal between 0 and 1), returns the value at that percentile.

If a specific value is not found at the exact percentile given, the value is estimated based on other numbers in the list.`,
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List of numeric values',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Percentile',
    description: 'The percentile whose value within data will be calculated and returned ( 0 <= percentile <= 1).',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
      },
      expression: 'PERCENTILE(numbers, 0)',
      result: 1,
    },
    {
      context: {
        numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
      },
      expression: 'PERCENTILE(numbers, 0.5)',
      result: 5.5,
    },
    {
      context: {
        numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
      },
      expression: 'PERCENTILE(numbers, 1)',
      result: 10,
    },
    {
      context: {
        numbersOdd: [1, 2, 3, 4, 5, 6, 7, 8, 9],
      },
      expression: 'PERCENTILE(numbersOdd, 0.5)',
      result: 5,
    },
    {
      context: {
        numbersUnsorted: [1, 4, 3, 2, 5, 10, 7, 8, 9, 6],
      },
      expression: 'PERCENTILE(numbersUnsorted, 0.5)',
      result: 5.5,
    },
  ],
};
