// @sapappgyver/appgyver-formulas/functions/statistical/PERCENT_RANK_EXC.js
const execute = require('./PERCENT_RANK_EXC.runtime');

module.exports = {
  name: 'PERCENT_RANK_EXC',
  category: 'Statistical',
  title: 'Percent rank of a number in list (exclusive 0 and 1)',
  description: `Given a list of numbers and a value in that list, gives the percent rank / percentile of the value. The percent rank is given as a decimal number between 0 and 1.

If the exact value is not found in the list, it's percent rank is estimated based on the other numbers.

Contrary to \`PERCENT_RANK\`, 0 and 1 percent ranks are excluded.

If the value is smaller or bigger than the smallest/biggest number in the list, the function returns \`null\`.`,
  parameters: [{
    title: 'List',
    description: 'The list containing the dataset to consider.',
    type: 'array',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Percentile',
    description: 'The percentile whose value within data will be calculated and returned',
  }, {
    type: 'number',
    title: 'Significant digits',
    description: 'The number of significant figures to use in the calculation',
    default: 3,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      expression: 'PERCENT_RANK_EXC(numbers, 1)',
      context: {
        numbers: [1, 2, 3, 4],
      },
      result: 0.2,
    }, {
      expression: 'PERCENT_RANK_EXC(numbersLarge, 24)',
      context: {
        numbersLarge: [20, 18, 3, 35, 67, 41, 72, 90],
      },
      result: 0.363,
    }, {
      expression: 'PERCENT_RANK_EXC(numbersLarge, 23, 5)',
      context: {
        numbersLarge: [20, 18, 3, 35, 67, 41, 72, 90],
      },
      result: 0.35556,
    },
  ],
};
