// @sapappgyver/appgyver-formulas/functions/statistical/PERCENT_RANK.js
const execute = require('./PERCENT_RANK.runtime');

module.exports = {
  name: 'PERCENT_RANK',
  category: 'Statistical',
  title: 'Percent rank of a number in list',
  description: `Given a list of numbers and a value in that list, gives the percent rank / percentile of the value. The percent rank is given as a decimal number between 0 and 1.

If the exact value is not found in the list, it's percent rank is estimated based on the other numbers.

If the value is at the start of the list, the function returns 0. If the value is at the end of the list, the function returns 1.

If the value is smaller or bigger than the smallest/biggest number in the list, the function returns \`null\`.`,
  parameters: [{
    title: 'List',
    description: 'List containing the dataset to consider',
    type: 'array',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Percentile',
    description: 'The percentile whose value within data will be calculated and returned',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      expression: 'PERCENT_RANK(numbers1, 1)',
      context: {
        numbers1: [1, 2, 3, 4],
      },
      result: 0,
    }, {
      expression: 'PERCENT_RANK(numbers1, 4)',
      context: {
        numbers1: [1, 2, 3, 4],
      },
      result: 1,
    }, {
      expression: 'PERCENT_RANK(numbers1, 2.5)',
      context: {
        numbers1: [1, 2, 3, 4],
      },
      result: 0.5,
    },
  ],
};
