// @sapappgyver/appgyver-formulas/functions/statistical/PERMUTATIONA.js
const execute = require('./PERMUTATIONA.runtime');

module.exports = {
  name: 'PERMUTATIONA',
  category: 'Statistical',
  title: 'Permutation',
  description: `Given the size of a pool of objects, returns the number of permutations for selecting a number of objects from that pool, with each object being replaced to the pool after being selected.

Mathematically, this is equivalent to raising the pool size to the power of number of objects selected.`,
  parameters: [{
    title: 'Number',
    description: 'The total number of objects to choose from.',
    type: 'number',
  }, {
    title: 'Number chosen',
    description: 'The number of objects to choose from number. Must be less than or equal to number.',
    type: 'number',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'PERMUTATIONA(5, 3)', result: 125 },
    { expression: 'PERMUTATIONA(6, 3)', result: 216 },
    { expression: 'PERMUTATIONA(5, 4)', result: 625 },
    { expression: 'PERMUTATIONA(5, 1)', result: 5 },
    { expression: 'PERMUTATIONA(5, 5)', result: 3125 },
  ],
};
