// @sapappgyver/appgyver-formulas/functions/statistical/POISSON_DIST.js
const execute = require('./POISSON_DIST.runtime');

module.exports = {
  name: 'POISSON_DIST',
  category: 'Statistical',
  title: 'Poisson distribution',
  description: 'Calculates the value of the Poisson distribution function for the given input number and mean. Optionally, the cumulative distribution function can be used.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the Poisson distribution function.',
  }, {
    type: 'number',
    title: 'Mean',
    description: 'The mean (mu) of the distribution function',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'POISSON_DIST(0, 2)', result: 0.13533528 },
    { expression: 'POISSON_DIST(1, 2)', result: 0.27067057 },
    { expression: 'POISSON_DIST(1, 2, false)', result: 0.27067057 },
    { expression: 'POISSON_DIST(0, 2, true)', result: 0.13533528 },
    { expression: 'POISSON_DIST(1, 2, true)', result: 0.40600585 },
  ],
};
