// @sapappgyver/appgyver-formulas/functions/statistical/PROB.js
const execute = require('./PROB.runtime');

module.exports = {
  name: 'PROB',
  category: 'Statistical',
  title: 'Probability',
  description: 'Given two lists of numbers that correspond to a set of values and their probabilities, calculates the probability that a random value falls between the given low and high limit.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List containing the dataset to consider',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'Probabilities',
    description: 'List containing probabilities corresponding to data. Sum of probabilities must be 1',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Low limit',
    description: 'The lower bound on the value range for which to calculate the probability',
  }, {
    type: 'number',
    title: 'High limit',
    description: 'The lower bound on the value range for which to calculate the probability',
    default: null,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 4],
        probs: [0.5, 0.25, 0.25],
      },
      expression: 'PROB(numbers, probs, 2)',
      result: 0.25,
    },
    {
      context: {
        numbersLarge: [1, 2, 4, 5, 6],
        probsLarge: [0.1, 0.1, 0.1, 0.3, 0.4],
      },
      expression: 'PROB(numbersLarge, probsLarge, 2)',
      result: 0.1,
    },
    {
      context: {
        numbersLarge: [1, 2, 4, 5, 6],
        probsLarge: [0.1, 0.1, 0.1, 0.3, 0.4],
      },
      expression: 'PROB(numbersLarge, probsLarge, 3)',
      result: 0,
    },
  ],
};
