// @sapappgyver/appgyver-formulas/functions/statistical/QUARTILE_EXC.js
const execute = require('./QUARTILE_EXC.runtime');

const quartileSet = new Set([1, 2, 3]);

module.exports = {
  name: 'QUARTILE_EXC',
  category: 'Statistical',
  title: 'Quartile of dataset, exclusive',
  description: 'Given a list of numbers and a quartile (integer 1, 2 or 3), returns the value in the list nearest to the given quartile.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'The list or range containing the dataset to consider.',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    enum: [...quartileSet],
    title: 'Quartile number',
    description: 'Determines the quartile value to return:\n'
      + '\n- `1` returns the value closest to the first quartile (25% mark)\n'
      + '\n- `2` returns the value closest to the median (50% mark)\n'
      + '\n- `3` returns the value closest to the third quartile (75% mark)',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'QUARTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 1)', result: 2.75 },
    { expression: 'QUARTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2)', result: 5.5 },
    { expression: 'QUARTILE_EXC([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3)', result: 8.25 },
  ],
};
