// @sapappgyver/appgyver-formulas/functions/statistical/QUARTILE.js
const execute = require('./QUARTILE.runtime');

module.exports = {
  name: 'QUARTILE',
  category: 'Statistical',
  title: 'Quartile of dataset',
  description: 'Given a list of numbers and a quartile (integer between 0 and 4), returns the value in the list nearest to the given quartile.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'A list of numbers describing the dataset.',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    enum: [0, 1, 2, 3, 4],
    title: 'Quartile number',
    description: 'Determines the quartile value to return:\n'
      + '\n- `0` returns the minimum value (0% mark)\n'
      + '\n- `1` returns the value closest to the first quartile (25% mark)\n'
      + '\n- `2` returns the value closest to the median (50% mark)\n'
      + '\n- `3` returns the value closest to the third quartile (75% mark)\n'
      + '\n- `4` returns the maximum value (100% mark).',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'QUARTILE([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 0)', result: 1 },
    { expression: 'QUARTILE([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 1)', result: 3.25 },
    { expression: 'QUARTILE([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2)', result: 5.5 },
    { expression: 'QUARTILE([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3)', result: 7.75 },
    { expression: 'QUARTILE([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 4)', result: 10 },
  ],
};
