// @sapappgyver/appgyver-formulas/functions/statistical/RANK.js
const execute = require('./RANK.runtime');

module.exports = {
  name: 'RANK',
  category: 'Statistical',
  title: 'Rank of the given value in the dataset',
  description: 'Given a list of numbers, returns the given value\'s rank in the list.',
  parameters: [{
    type: 'number',
    title: 'Value',
    description: 'The value whose rank will be determined',
  }, {
    type: 'array',
    title: 'Values',
    description: 'The list or range containing the dataset to consider',
    items: {
      type: 'number',
    },
  }, {
    type: 'boolean',
    title: 'Ordering',
    description: 'Whether to consider the values in data in descending or ascending order',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'RANK(5, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10])', result: 6 },
    { expression: 'RANK(5, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], false)', result: 6 },
    { expression: 'RANK(5, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], true)', result: 5 },
    { expression: 'RANK(5, [1, 4, 3, 2, 5, 10, 7, 8, 9, 6])', result: 6 },
    { expression: 'RANK(5, [1, 4, 3, 2, 5, 10, 7, 8, 9, 6], true)', result: 5 },
  ],
};
