// @sapappgyver/appgyver-formulas/functions/statistical/RANK_AVG.js
const execute = require('./RANK_AVG.runtime');

module.exports = {
  name: 'RANK_AVG',
  category: 'Statistical',
  title: 'Average rank of the given value in the dataset',
  description: 'Given a list of numbers, returns the given value\'s rank in the list. If there are multiple matching values in the list, returns their average rank.',
  parameters: [{
    type: 'number',
    title: 'Value',
    description: 'The value whose rank will be determined',
  }, {
    type: 'array',
    title: 'Values',
    description: 'The list or range containing the dataset to consider',
    items: {
      type: 'number',
    },
  }, {
    type: 'boolean',
    title: 'Ordering',
    description: 'Whether to consider the values in data in descending or ascending order',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'RANK_AVG(4, [1, 2, 3, 4, 4, 5, 7, 8, 9, 10])', result: 6.5 },
    { expression: 'RANK_AVG(4, [1, 2, 3, 4, 4, 5, 7, 8, 9, 10], false)', result: 6.5 },
    { expression: 'RANK_AVG(4, [1, 2, 3, 4, 4, 6, 7, 8, 9, 10], true)', result: 4.5 },
    { expression: 'RANK_AVG(4, [1, 4, 3, 2, 5, 10, 7, 8, 9, 4])', result: 6.5 },
    { expression: 'RANK_AVG(4, [1, 4, 3, 2, 5, 10, 7, 8, 9, 4], true)', result: 4.5 },
  ],
};
