// @sapappgyver/appgyver-formulas/functions/statistical/RSQ.js
const execute = require('./RSQ.runtime');

module.exports = {
  name: 'RSQ',
  category: 'Statistical',
  title: 'Pearson\'s R-Square function',
  description: 'Given two lists of numbers as a dataset, calculates the square of r, the Pearson product moment correlation coefficient.',
  parameters: [{
    type: 'array',
    title: 'Y dataset',
    description: 'List of dependent data.',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'X dataset',
    description: 'List of independent data.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 4],
        numbers2: [1, 3, 9],
      },
      expression: 'RSQ(numbers, numbers2)',
      result: 0.99175824,
    },
    {
      context: {
        numbers: [1, 2, 4],
        numbersReverse: [4, 2, 1],
      },
      expression: 'RSQ(numbers, numbersReverse)',
      result: 0.8622449,
    },
    {
      context: {
        numbers2: [1, 3, 9],
      },
      expression: 'RSQ(numbers2, numbers2)',
      result: 1,
    },
    {
      context: {
        numbersLarge: [1, 2, 4, 5, 6],
        numbersNoOrder: [23, 19, 50, 1, 1],
      },
      expression: 'RSQ(numbersLarge, numbersNoOrder)',
      result: 0.15776396,
    },
  ],
};
