// @sapappgyver/appgyver-formulas/functions/statistical/SKEW.js
const execute = require('./SKEW.runtime');

module.exports = {
  name: 'SKEW',
  category: 'Statistical',
  title: 'Skewness sample',
  description: 'Given a list of numbers, calculates their sample skewness.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List of numeric values',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbersSym: [1, 2, 3, 4],
      },
      expression: 'SKEW(numbersSym)',
      result: 0,
    },
    {
      context: {
        numbersNonSym: [1, 2, 3, 10],
      },
      expression: 'SKEW(numbersNonSym)',
      result: 1.76363261,
    },
    {
      context: {
        numbersNonSymNeg: [-10, 1, 2, 3],
      },
      expression: 'SKEW(numbersNonSymNeg)',
      result: -1.89165592,
    },
    {
      context: {
        numbers: [2, 5, 8, 13, 10, 18, 23, 26],
      },
      expression: 'SKEW(numbers)',
      result: 0.34461547,
    },
  ],
};
