// @sapappgyver/appgyver-formulas/functions/statistical/SKEW_P.js
const execute = require('./SKEW_P.runtime');

module.exports = {
  name: 'SKEW_P',
  category: 'Statistical',
  title: 'Skewness population',
  description: 'Given a list of numbers, calculates their population skewness.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List of numeric values',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbersSym: [1, 2, 3, 4],
      },
      expression: 'SKEW_P(numbersSym)',
      result: 0,
    },
    {
      context: {
        numbersNonSym: [1, 2, 3, 10],
      },
      expression: 'SKEW_P(numbersNonSym)',
      result: 1.01823376,
    },
    {
      context: {
        numbersNonSymNeg: [-10, 1, 2, 3],
      },
      expression: 'SKEW_P(numbersNonSymNeg)',
      result: -1.09214806,
    },
    {
      context: {
        numbers: [2, 5, 8, 13, 10, 18, 23, 26],
      },
      expression: 'SKEW_P(numbers)',
      result: 0.27630708,
    },
  ],
};
