// @sapappgyver/appgyver-formulas/functions/statistical/SLOPE.js
const execute = require('./SLOPE.runtime');

module.exports = {
  name: 'SLOPE',
  category: 'Statistical',
  title: 'Slope',
  description: 'Given two lists of numbers as a dataset, calculates the slope of the line from linear regression.',
  parameters: [{
    type: 'array',
    title: 'Y dataset',
    description: 'List of dependent data.',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'X dataset',
    description: 'List of independent data.',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [1, 2, 3],
        numbers2: [1, 3, 6],
      },
      expression: 'SLOPE(numbers, numbers2)',
      result: 0.39473684,
    },
    {
      context: {
        numbers: [1, 2, 3],
      },
      expression: 'SLOPE(numbers, numbers)',
      result: 1,
    },
    {
      context: {
        numbersReverse: [3, 2, 1],
      },
      expression: 'SLOPE(numbersReverse, numbersReverse)',
      result: 1,
    },
    {
      context: {
        numbers2: [1, 3, 6],
      },
      expression: 'SLOPE(numbers2, numbers2)',
      result: 1,
    },
    {
      context: {
        numbersLarge: [1, 2, 3, 4, 5],
        numbersNoOrder: [23, 19, 50, 1, 1],
      },
      expression: 'SLOPE(numbersLarge, numbersNoOrder)',
      result: -0.03815854,
    },
    {
      context: {
        numbersLarge: [1, 2, 3, 4, 5],
        numbersLargeReverse: [5, 4, 3, 2, 1],
      },
      expression: 'SLOPE(numbersLarge, numbersLargeReverse)',
      result: -1,
    },
    {
      context: {
        numbersConstant: [1, 1, 1],
        numbers: [1, 2, 3],
      },
      expression: 'SLOPE(numbersConstant, numbers)',
      result: 0,
    },
  ],
};
