// @sapappgyver/appgyver-formulas/functions/statistical/STDEV_P.js
const execute = require('./STDEV_P.runtime');

module.exports = {
  name: 'STDEV_P',
  category: 'Statistical',
  title: 'Standard deviation based on a population',
  description: 'Given a list of numbers as a population, calculates standard deviation.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List of numeric values',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        numbers: [2, 4, 6],
      },
      expression: 'STDEV_P(numbers)',
      result: 1.63299316,
    },
    {
      context: {
        numbers2: [2, 5, 8, 13, 10],
      },
      expression: 'STDEV_P(numbers2)',
      result: 3.82622529,
    },
    {
      context: {
        numbers3: [2, 4, 4, 4, 5, 5, 7, 9],
      },
      expression: 'STDEV_P(numbers3)',
      result: 2,
    },
    {
      context: {
        singleNumber: [1],
      },
      expression: 'STDEV_P(singleNumber)',
      result: 0,
    },
  ],
};
