// @sapappgyver/appgyver-formulas/functions/statistical/T_DIST.js
const execute = require('./T_DIST.runtime');

module.exports = {
  name: 'T_DIST',
  category: 'Statistical',
  title: 'Student\'s t-distribution',
  description: 'Calculates a Student\'s t-distribution probability for the given input.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the distribution function',
  }, {
    type: 'number',
    title: 'Degrees of freedom.',
    description: 'The number of degrees of freedom.',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'T_DIST(1, 2)', result: 0.19245009 },
    { expression: 'T_DIST(1, 2, false)', result: 0.19245009 },
    { expression: 'T_DIST(-1, 2)', result: 0.19245009 },
    { expression: 'T_DIST(2, 2)', result: 0.06804138 },
    { expression: 'T_DIST(1, 2, true)', result: 0.78867513 },
    { expression: 'T_DIST(-1, 2, true)', result: 0.21132487 },
    { expression: 'T_DIST(2, 2, true)', result: 0.90824829 },
  ],
};
