// @sapappgyver/appgyver-formulas/functions/statistical/T_TEST.js
const execute = require('./T_TEST.runtime');

module.exports = {
  name: 'T_TEST',
  category: 'Statistical',
  title: 'T-test',
  description: 'Calculates the probability associated with t-test for the given two lists of numbers.',
  parameters: [{
    type: 'array',
    title: 'First sample',
    description: 'The first list of numbers to consider for the t-test',
    items: {
      type: 'number',
    },
  }, {
    type: 'array',
    title: 'Second sample',
    description: 'The second list of numbers to consider for the t-test',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Tails',
    description: 'Specifies the number of distribution tails: '
      + 'If 1: uses a one-tailed distribution; '
      + 'If 2: uses a two-tailed distribution.',
    enum: [1, 2],
  }, {
    type: 'number',
    title: 'Type',
    description: 'Specifies the type of t-Test: '
      + 'If 1: a paired test is performed; '
      + 'If 2: a two-sample equal variance (homoscedastic) test is performed; '
      + 'If 3: a two-sample unequal variance (heteroscedastic) test is performed.',
    enum: [1, 2, 3],
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        num1: [1, 2, 3, 4, 5, 6],
        num2: [1, 3, 9, 12, 2, 4],
      },
      expression: 'T_TEST(num1, num2, 1, 1)',
      result: 0.1986383,
    },
    {
      context: {
        num1: [1, 2, 3, 4, 5, 6],
        num2: [1, 3, 9, 12, 2, 4],
      },
      expression: 'T_TEST(num1, num2, 1, 2)',
      result: 0.20462468,
    },
    {
      context: {
        numShort1: [1, 2, 4],
        numShort2: [1, 3, 9],
      },
      expression: 'T_TEST(numShort1, numShort2, 1, 1)',
      result: 0.16031689,
    },
  ],
};
