// @sapappgyver/appgyver-formulas/functions/statistical/TRIM_MEAN.js
const execute = require('./TRIM_MEAN.runtime');

module.exports = {
  name: 'TRIM_MEAN',
  category: 'Statistical',
  title: 'Trim mean',
  description: 'Given a list of numbers, calculates the mean, excluding a specified range of numbers from the start and end of the list.',
  parameters: [{
    type: 'array',
    title: 'Values',
    description: 'List containing the dataset to consider',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Number',
    description: 'The proportion of the dataset to exclude, from the extremities of the set',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'TRIM_MEAN([1, 2, 3, 4], 0.3)', result: 2.5 },
    {
      context: {
        numbers: [29, 44, 61, 1, 36, -3, 14, 10, -24],
      },
      expression: 'TRIM_MEAN(numbers, 0.5)',
      result: 18,
    },
    {
      context: {
        numbers: [29, 44, 61, 1, 36, -3, 14, 10, -24],
      },
      expression: 'TRIM_MEAN(numbers, 0.99)',
      result: 14,
    },
  ],
};
