// @sapappgyver/appgyver-formulas/functions/statistical/WEIBULL_DIST.js
const execute = require('./WEIBULL_DIST.runtime');

module.exports = {
  name: 'WEIBULL_DIST',
  category: 'Statistical',
  title: 'Weibull distribution',
  description: 'Calculates the Weibull distribution function\'s value for the given input number, shape and scale. Optionally, the cumulative distribution function can be used.',
  parameters: [{
    type: 'number',
    title: 'Input',
    description: 'The input to the distribution function',
  }, {
    type: 'number',
    title: 'Shape',
    description: 'The shape parameter of the Weibull distribution function',
  }, {
    type: 'number',
    title: 'Scale',
    description: 'The scale parameter of the Weibull distribution function.',
  }, {
    type: 'boolean',
    title: 'Cumulative',
    description: 'Whether to use the cumulative distribution function rather than the distribution function',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'WEIBULL_DIST(0, 2, 3)', result: 0 },
    { expression: 'WEIBULL_DIST(0.1, 2, 3)', result: 0.02219754 },
    { expression: 'WEIBULL_DIST(0.1, 2, 3, false)', result: 0.02219754 },
    { expression: 'WEIBULL_DIST(1, 2, 3)', result: 0.19885318 },
    { expression: 'WEIBULL_DIST(0.1, 2, 3, true)', result: 0.00111049 },
    { expression: 'WEIBULL_DIST(1, 2, 3, true)', result: 0.10516068 },
  ],
};
