// @sapappgyver/appgyver-formulas/functions/statistical/WEIBULL_INV.js
const execute = require('./WEIBULL_INV.runtime');

module.exports = {
  name: 'WEIBULL_INV',
  category: 'Statistical',
  title: 'Weibull inverse distribution',
  description: 'Given a probability, calculates the value of the inverse Weibull distribution function.',
  parameters: [{
    type: 'number',
    title: 'Probability',
    description: 'Probability to calculate the inverse for, given as a decimal between 0 and 1.',
  }, {
    type: 'number',
    title: 'Shape',
    description: 'The shape parameter of the Weibull distribution function.',
  }, {
    type: 'number',
    title: 'Scale',
    description: 'The scale parameter of the Weibull distribution function.',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'WEIBULL_INV(0, 2, 3)', result: 0 },
    { expression: 'WEIBULL_INV(0.1, 2, 3)', result: 0.97377854 },
    { expression: 'WEIBULL_INV(0.9, 2, 3)', result: 4.55228139 },
    { expression: 'WEIBULL_INV(0.5, 2, 4)', result: 3.33021844 },
    { expression: 'WEIBULL_INV(0.5, 3, 3)', result: 2.65499113 },
  ],
};
