// @sapappgyver/appgyver-formulas/functions/statistical/Z_TEST.js
const execute = require('./Z_TEST.runtime');

module.exports = {
  name: 'Z_TEST',
  category: 'Statistical',
  title: 'Z test',
  description: 'Given a list of numbers, calculates the one-tailed P-value of a Z-test. Optionally, the standard distribution can be explicitly given.',
  parameters: [{
    type: 'array',
    title: 'Data',
    description: 'List of numbers to test.',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Value',
    description: 'The test statistic to use in the Z-test',
  }, {
    type: 'number',
    title: 'Standard deviation',
    description: 'The standard deviation to assume for the Z-test; if it is not provided, it will be calculated from the data',
    default: null,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'Z_TEST([1, 2, 3, 4, 5], 2)', result: 0.07864960353 },
    { expression: 'Z_TEST([1, 2, 3, 4, 5], 2, 3)', result: 0.2280282701 },
    { expression: 'Z_TEST([1, 2, 3, 4, 5], 10)', result: 1 },
    { expression: 'Z_TEST([1, 2, 3, 4, 5], 10, 3)', result: 0.9999999093 },
    { expression: 'Z_TEST([1, 2, 3, 4, 5], 10, 10)', result: 0.941237566 },
    { expression: 'Z_TEST([1, 2, 3, 4, 5], 5)', result: 0.9976611325 },
    { expression: 'Z_TEST([1, 2, 3, 4, 5], 5, 10)', result: 0.672639577 },
  ],
};
