// @sapappgyver/appgyver-formulas/functions/date/ADD_DURATION.js
const execute = require('./ADD_DURATION.runtime');

const defaultUnit = 'milliseconds';

module.exports = {
  name: 'ADD_DURATION',
  category: 'Date',
  title: 'Add duration to datetime',
  description: `Returns a new datetime by adding a number of given units to the original datetime, such as 1 day or 7 hours.

Valid units are: \`years\`,  \`months\`, \`days\`, \`hours\`, \`minutes\`, \`seconds\` or \`milliseconds\`.`,
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date',
    description: 'Date',
  }, {
    type: 'number',
    title: 'Duration',
    description: 'Duration to be added to date',
  }, {
    type: 'string',
    title: 'Units',
    description: 'Units of duration("years", "months", "weeks", "quarters", "days", "hours", "minutes", "seconds", "milliseconds"). Default is "milliseconds".',
    enum: ['years', 'months', 'weeks', 'quarters', 'days', 'hours', 'minutes', 'seconds', 'milliseconds', null],
    default: defaultUnit,
  }],
  analyze: () => ({ schema: { type: 'string', format: 'date-time' } }),
  execute,
  examples: [
    { expression: 'ADD_DURATION("2019-07-11T13:06:15+03:00", 0)', result: '2019-07-11T13:06:15+03:00' },
    { expression: 'ADD_DURATION("2019-07-11T13:06:15+03:00", 1, "days")', result: '2019-07-12T13:06:15+03:00' },
    { expression: 'ADD_DURATION("2019-07-11T13:06:15Z", -2, "hours")', result: '2019-07-11T11:06:15Z' },
  ],
};
