// @sapappgyver/appgyver-formulas/functions/date/DATETIME_IS_SAME_OR_AFTER.js
const execute = require('./DATETIME_IS_SAME_OR_AFTER.runtime');

module.exports = {
  name: 'DATETIME_IS_SAME_OR_AFTER',
  category: 'Date',
  title: 'Check if a datetime is the same or after another',
  description: 'Returns `true` if the first datetime is the same or after the second datetime, otherwise returns `false`.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date 1',
    description: 'Date 1',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Date 2',
    description: 'Date 2',
  }, {
    type: 'string',
    title: 'Precision',
    description: 'Precision for comparing ("year","month","week","isoWeek","day","hour","minute","second", null)',
    enum: ['year', 'month', 'week', 'isoWeek', 'day', 'hour', 'minute', 'second', null],
    default: null,
  }],
  analyze: () => ({ schema: { type: 'boolean' } }),
  execute,
  examples: [
    { expression: 'DATETIME_IS_SAME_OR_AFTER("2011-10-10T12:13:14Z", "2011-10-09T12:13:14Z")', result: true },
    { expression: 'DATETIME_IS_SAME_OR_AFTER("2011-10-09T12:13:14Z", "2011-10-10T12:13:14Z")', result: false },
    { expression: 'DATETIME_IS_SAME_OR_AFTER("2011-10-10T12:13:14Z", "2011-10-10T12:13:14Z")', result: true },
    { expression: 'DATETIME_IS_SAME_OR_AFTER("2012-10-10T12:13:14Z", "2011-10-09T12:13:14Z", "year")', result: true },
  ],
};
