// @sapappgyver/appgyver-formulas/functions/date/DATETIME_IS_BEFORE.js
const execute = require('./DATETIME_IS_BEFORE.runtime');

module.exports = {
  name: 'DATETIME_IS_BEFORE',
  category: 'Date',
  title: 'Check if a datetime is before another',
  description: 'Returns `true` if the first datetime is before the second datetime, otherwise returns `false`.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date 1',
    description: 'First date',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Date 2',
    description: 'Second date',
  }, {
    type: 'string',
    title: 'Precision',
    description: 'Optional precision for the comparison, one of `year`, `month`, `week`, `isoWeek`, `day`, `hour`, `minute` or `second`',
    enum: ['year', 'month', 'week', 'isoWeek', 'day', 'hour', 'minute', 'second', null],
    default: null,
  }],
  analyze: () => ({ schema: { type: 'boolean' } }),
  execute,
  examples: [
    { expression: 'DATETIME_IS_BEFORE("2011-10-10T12:13:14Z", "2011-10-09T12:13:14Z")', result: false },
    { expression: 'DATETIME_IS_BEFORE("2011-10-09T12:13:14Z", "2011-10-10T12:13:14Z")', result: true },
    { expression: 'DATETIME_IS_BEFORE("2011-10-10T12:13:14Z", "2011-10-10T12:13:14Z")', result: false },
    { expression: 'DATETIME_IS_BEFORE("2012-10-10T12:13:14Z", "2011-10-09T12:13:14Z", "year")', result: false },
  ],
};
