// @sapappgyver/appgyver-formulas/functions/date/DATETIME.js
const execute = require('./DATETIME.runtime');

module.exports = {
  name: 'DATETIME',
  category: 'Date',
  title: 'Convert text to ISO 8601 datetime',
  description: 'Convert a valid date/datetime text, or a number representing milliseconds since the Unix Epoch (1 January, 1970 UTC), to an ISO 8601 formatted text representation.',
  parameters: [{
    title: 'Timestamp',
    description: 'Date text or number of milliseconds since 1 January, 1970 UTC',
    anyOf: [{ type: 'number' }, { type: 'string' }],
  }, {
    title: 'Date Format',
    description: 'If you know the format of an input text, you can use that to parse a date',
    type: 'string',
    default: null,
  }],
  analyze: () => ({ schema: { type: 'string', format: 'date-time' } }),
  execute,
  examples: [
    { expression: 'DATETIME(0)', result: '1970-01-01T00:00:00Z' },
    { expression: 'DATETIME(5000)', result: '1970-01-01T00:00:05Z' },
    { expression: 'DATETIME("2019-12-31T23:59:55+05:00")', result: '2019-12-31T23:59:55+05:00' },
  ],
};
