// @sapappgyver/appgyver-formulas/functions/date/DATETIME_IS_SAME.js
const execute = require('./DATETIME_IS_SAME.runtime');

module.exports = {
  name: 'DATETIME_IS_SAME',
  category: 'Date',
  title: 'Check if datetimes are equal',
  description: 'Returns `true` if the two given datetimes are exactly the same, otherwise returns `false`.',
  parameters: [{
    title: 'Date 1',
    description: 'Date 1',
    type: 'string',
    format: 'date-time',
  }, {
    title: 'Date 2',
    description: 'Date 2',
    type: 'string',
    format: 'date-time',
  }, {
    type: 'string',
    title: 'Precision',
    default: null,
    description: 'Precision for comparing ("year","month","week","isoWeek","day","hour","minute","second", null)',
    enum: ['year', 'month', 'week', 'isoWeek', 'day', 'hour', 'minute', 'second', null],
  }],
  analyze: () => ({ schema: { type: 'boolean' } }),
  execute,
  examples: [
    { expression: 'DATETIME_IS_SAME("2011-10-10T12:13:14Z", "2011-10-09T12:13:14Z")', result: false },
    { expression: 'DATETIME_IS_SAME("2011-10-10T12:13:14Z", "2011-10-10T12:13:14Z")', result: true },
    { expression: 'DATETIME_IS_SAME("2011-10-10T12:13:14Z", "2011-10-09T12:13:14Z", "year")', result: true },
  ],
};
