// @sapappgyver/appgyver-formulas/functions/date/DATETIME_DIFFERENCE.js
const execute = require('./DATETIME_DIFFERENCE.runtime');

const defaultUnit = 'milliseconds';

module.exports = {
  name: 'DATETIME_DIFFERENCE',
  category: 'Date',
  title: 'Calculate difference between two datetimes',
  description: 'Returns the difference between the given two datetimes in the given unit of time.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date 1',
    description: 'Date 1',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Date 2',
    description: 'Date 2',
  }, {
    type: 'string',
    title: 'Units',
    description: 'Returned units ("years", "quarters", "months", "weeks", "days", "hours", "minutes", "seconds" , "milliseconds")',
    default: defaultUnit,
    enum: ['years', 'quarters', 'months', 'weeks', 'days', 'hours', 'minutes', 'seconds', 'milliseconds'],
  }, {
    type: 'boolean',
    title: 'In decimal',
    description: 'Prevent truncate of returned result (true, false)',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'DATETIME_DIFFERENCE("2019-07-11T13:06:15+03:00", "2019-07-11T13:06:15+03:00")', result: 0 },
    { expression: 'DATETIME_DIFFERENCE("2019-07-11T13:06:15+03:00", "2018-07-11T13:06:15+03:00", "hours")', result: 8760 },
    { expression: 'DATETIME_DIFFERENCE("2019-07-11T13:06:15+03:00", "2018-07-11T13:06:15+03:00", "weeks", true)', result: 52.14285714 },
  ],
};
