// @sapappgyver/appgyver-formulas/functions/date/FORMAT_DATETIME_WITH_TIMEZONE.js
const moment = require('moment-timezone');
const execute = require('./FORMAT_DATETIME_WITH_TIMEZONE.runtime');

const timezones = moment.tz.names();

module.exports = {
  name: 'FORMAT_DATETIME_WITH_TIMEZONE',
  category: 'Date',
  title: 'Format datetime into human-readable text, using an explicit timezone',
  description: `Returns the given datetime formatted into a human readable text.

The formatted text takes the given timezone into account, adjusting the output accordingly.`,
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date',
    description: 'Date',
  }, {
    type: 'string',
    title: 'Format',
    description: 'Format to display date',
  }, {
    type: 'string',
    title: 'Timezone',
    description: 'Timezone in which the date will be formatted. Defaults to the timezone defined in the date/time itself, or UTC if undefined.',
    enum: [null, ...timezones],
    default: null,
  }],
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    { expression: 'FORMAT_DATETIME_WITH_TIMEZONE("2019-07-11T13:06:15+03:00", "DD-MM-YYYY HH:mm:ss")', result: '11-07-2019 13:06:15' },
    { expression: 'FORMAT_DATETIME_WITH_TIMEZONE("2019-07-11T13:06:15Z", "DD-MM-YYYY HH:mm:ss", "Europe/Helsinki")', result: '11-07-2019 16:06:15' },
  ],
};
