// @sapappgyver/appgyver-formulas/functions/date/FORMAT_DATETIME_LOCAL.js
const execute = require('./FORMAT_DATETIME_LOCAL.runtime');

module.exports = {
  name: 'FORMAT_DATETIME_LOCAL',
  category: 'Date',
  title: 'Format datetime into human-readable text, with local timezone taken into account',
  description: `Returns the given datetime formatted into a human readable text. Follows the latest Moment-js syntax.

The formatted text takes the current timezone of the device into account, so e.g. a date returned by a server in UTC +0 is formatted correctly to take the local time difference into account.`,
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date/time',
    description: 'Date/time to format',
  }, {
    type: 'string',
    title: 'Format',
    description: 'Format in which the date/time is returned',
  }],
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    // NOTE: The examples assume the local timezone US/Central! See functions.test.js
    { expression: 'FORMAT_DATETIME_LOCAL("2019-07-11T13:06:15+03:00", "DD.MM.YYYY HH:mm:ss")', result: '11.07.2019 05:06:15' },
    { expression: 'FORMAT_DATETIME_LOCAL("2019-02-11T13:06:15Z", "DD.MM.YYYY HH:mm:ss")', result: '11.02.2019 07:06:15' },
    { expression: 'FORMAT_DATETIME_LOCAL("2022-01-13T07:30:30Z", "DD.MM.YYYY HH:mm:ss")', result: '13.01.2022 01:30:30' },
  ],
};
