// @sapappgyver/appgyver-formulas/functions/date/GET_DATETIME_COMPONENT.js
const execute = require('./GET_DATETIME_COMPONENT.runtime');

module.exports = {
  name: 'GET_DATETIME_COMPONENT',
  category: 'Date',
  title: 'Get datetime component',
  description: 'For the given datetime, returns the given component: `year`,  `month`, `date`, `hour`, `minute`, `second` or `millisecond`.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date',
    description: 'Date',
  }, {
    type: 'string',
    title: 'Component',
    description: 'Component to get from the date',
    enum: ['year', 'month', 'date', 'hour', 'minute', 'second', 'millisecond'],
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'GET_DATETIME_COMPONENT("2012-07-11T13:06:15+03:00", "year")', result: 2012 },
    { expression: 'GET_DATETIME_COMPONENT("2012-07-11T13:06:15+03:00", "month")', result: 7 },
    { expression: 'GET_DATETIME_COMPONENT("2012-07-11T13:06:15+03:00", "date")', result: 11 },
    { expression: 'GET_DATETIME_COMPONENT("2012-07-11T13:06:15+02:00", "hour")', result: 13 },
    { expression: 'GET_DATETIME_COMPONENT("2012-07-11T13:06:15+01:00", "minute")', result: 6 },
    { expression: 'GET_DATETIME_COMPONENT("2012-07-11T13:06:15+01:00", "second")', result: 15 },
    { expression: 'GET_DATETIME_COMPONENT("2015-12-02T21:45:22.279Z", "millisecond")', result: 279 },
  ],
};
