// @sapappgyver/appgyver-formulas/functions/date/NOW.js
const moment = require('moment-timezone');
const execute = require('./NOW.runtime');

const exampleMoment = moment.parseZone('2019-06-27T15:46:12.240Z');

module.exports = {
  name: 'NOW',
  category: 'Date',
  title: 'Get current datetime',
  description: `Returns the current datetime as an ISO 8601 text.

Optionally, the current datetime can be returned as a human-readable text by providing a datetime format parameter.`,
  deterministic: false, // Depends on the current time
  parameters: [
    {
      type: 'string',
      title: 'Format',
      description: 'Format to display date',
      default: null,
    },
  ],
  analyze: ([formatParam]) => {
    const formatSchema = formatParam && formatParam.schema;
    const schema = { type: 'string' };
    if (!formatParam) {
      schema.format = 'date-time';
    }
    let formatExamples = (formatSchema && (formatSchema.examples || formatSchema.enum)) || [];
    formatExamples = formatExamples.filter(example => typeof example === 'string');
    if (!formatExamples.length) {
      formatExamples = [null];
    }
    schema.examples = [];
    formatExamples.forEach((format) => {
      try {
        schema.examples.push(exampleMoment.format(format));
      } catch (error) {
        // Ignore formatting errors
      }
    });
    return { schema };
  },
  execute,
  examples: [
    { expression: 'NOW("DD-MM-YYYY")', result: '13-08-2019' },
    { expression: 'NOW("")', result: '2019-08-13T17:33:41Z' },
  ],
};
