// @sapappgyver/appgyver-formulas/functions/date/SET_TIMEZONE.js
const moment = require('moment-timezone');
const execute = require('./SET_TIMEZONE.runtime');

const timezones = moment.tz.names();

module.exports = {
  name: 'SET_TIMEZONE',
  category: 'Date',
  title: 'Convert datetime to different timezone',
  description: 'Returns a datetime that represents the given datetime in another timezone.',
  parameters: [{
    title: 'Date',
    description: 'Date to convert',
    type: 'string',
    format: 'date-time',
  }, {
    title: 'Timezone',
    description: 'Timezone to convert.',
    type: 'string',
    enum: timezones,
  }],
  analyze: () => ({ schema: { type: 'string', format: 'date-time' } }),
  execute,
  examples: [
    { expression: 'SET_TIMEZONE("2019-07-11T13:06:15+03:00", "America/Los_Angeles")', result: '2019-07-11T03:06:15-07:00' },
    { expression: 'SET_TIMEZONE("2019-07-11T13:06:15+03:00", "Europe/Madrid")', result: '2019-07-11T12:06:15+02:00' },
    { expression: 'SET_TIMEZONE("2019-07-11T13:06:15+03:00", "Etc/GMT+3")', result: '2019-07-11T07:06:15-03:00' },
    { expression: 'SET_TIMEZONE("2019-07-11T13:06:15+03:00", "Etc/GMT-3")', result: '2019-07-11T13:06:15+03:00' },
  ],
};
