// @sapappgyver/appgyver-formulas/functions/date/SET_DATETIME_COMPONENT.js
const execute = require('./SET_DATETIME_COMPONENT.runtime');

module.exports = {
  name: 'SET_DATETIME_COMPONENT',
  category: 'Date',
  title: 'Set datetime component to new value',
  description: `Returns the given datetime with a given component changed to a new value. Valid components are: \`year\`,  \`month\`, \`date\`, \`hour\`, \`minute\`, \`second\` or \`millisecond\`.

Note that the value is changed in the timezone in which the given datetime is specified!`,
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date',
    description: 'Date',
  }, {
    type: 'number',
    title: 'Value',
    description: 'New value for the component',
  }, {
    type: 'string',
    title: 'Component',
    description: 'Component to change',
    enum: ['year', 'month', 'date', 'hour', 'minute', 'second', 'millisecond'],
  }],
  analyze: () => ({ schema: { type: 'string', format: 'date-time' } }),
  execute,
  examples: [
    { expression: 'SET_DATETIME_COMPONENT("2012-07-11T13:06:15+03:00", 2013, "year")', result: '2013-07-11T13:06:15+03:00' },
    { expression: 'SET_DATETIME_COMPONENT("2012-07-11T13:06:15+03:00", 3, "month")', result: '2012-03-11T13:06:15+03:00' },
    { expression: 'SET_DATETIME_COMPONENT("2012-07-11T13:06:15+03:00", 5, "date")', result: '2012-07-05T13:06:15+03:00' },
    { expression: 'SET_DATETIME_COMPONENT("2012-07-11T13:06:15+02:00", 10, "hour")', result: '2012-07-11T10:06:15+02:00' },
    { expression: 'SET_DATETIME_COMPONENT("2012-07-11T13:06:15+01:00", 30, "minute")', result: '2012-07-11T13:30:15+01:00' },
    { expression: 'SET_DATETIME_COMPONENT("2015-12-02T21:45:22.279Z", 12, "millisecond")', result: '2015-12-02T21:45:22.012Z' },
  ],
};
