// @sapappgyver/appgyver-formulas/functions/date/SUBTRACT_DURATION.js
const execute = require('./SUBTRACT_DURATION.runtime');

const defaultUnit = 'milliseconds';

module.exports = {
  name: 'SUBTRACT_DURATION',
  category: 'Date',
  title: 'Subtract duration from datetime',
  description: `Returns a new datetime by subtracting a number of given units from the original datetime, such as 1 day or 7 hours.

Valid units are: \`years\`,  \`months\`, \`dates\`, \`hours\`, \`minutes\`, \`seconds\` or \`milliseconds\`.`,
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date',
    description: 'Date',
  }, {
    type: 'number',
    title: 'Duration',
    description: 'Duration to be subtracted from date',
  }, {
    type: 'string',
    title: 'Units',
    description: 'Units of duration("years", "months", "weeks", "quarters", "days", "hours", "minutes", "seconds", "milliseconds"). Default is "milliseconds".',
    enum: ['years', 'months', 'weeks', 'quarters', 'days', 'hours', 'minutes', 'seconds', 'milliseconds', null],
    default: defaultUnit,
  }],
  analyze: () => ({ schema: { type: 'string', format: 'date-time' } }),
  execute,
  examples: [
    { expression: 'SUBTRACT_DURATION("2019-07-11T13:06:15+03:00", 3600000)', result: '2019-07-11T12:06:15+03:00' },
    { expression: 'SUBTRACT_DURATION("2019-07-11T13:06:15+03:00", 1, "hours")', result: '2019-07-11T12:06:15+03:00' },
    { expression: 'SUBTRACT_DURATION("2019-07-11T13:06:15+03:00", 1, "months")', result: '2019-06-11T13:06:15+03:00' },
    { expression: 'SUBTRACT_DURATION("2019-07-11T13:06:15Z", -2, "hours")', result: '2019-07-11T15:06:15Z' },
  ],
};
