// @sapappgyver/appgyver-formulas/functions/date/YEARFRAC.js
const execute = require('./YEARFRAC.runtime');

module.exports = {
  name: 'YEARFRAC',
  category: 'Date',
  title: 'Number of years between dates',
  description: `Returns the difference in years between the given two datetimes. Fractional year differences are represented as a decimal number.

Optionally, you can change the convention for how the number of days in a year is counted.`,
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date start',
    description: 'The start date to consider in the calculation. Must be a reference to a cell containing a date, a function returning a date type, or a number.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Date end',
    description: 'The end date to consider in the calculation. Must be a reference to a cell containing a date, a function returning a date type, or a number.',
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'YEARFRAC("2010-01-22T13:06:15+03:00", "2000-01-01T13:06:15+03:00")', result: 10.05833333 },
    { expression: 'YEARFRAC("2013-01-23T13:06:15+03:00", "2004-02-03T13:06:15+03:00", "actual_360")', result: 9.10277778 },
    { expression: 'YEARFRAC("2016-01-25T13:06:15+03:00", "2008-03-07T13:06:15+03:00", "actual_365")', result: 7.89041096 },
    { expression: 'YEARFRAC("2019-01-26T13:06:15+03:00", "2012-04-09T13:06:15+03:00", "european")', result: 6.79722222 },
  ],
};
