// @sapappgyver/appgyver-formulas/functions/financial/ACCRINT.js
const execute = require('./ACCRINT.runtime');

module.exports = {
  name: 'ACCRINT',
  category: 'Financial',
  title: 'Accrued interest of a security',
  description: 'Returns the accrued interest of a security with periodic payments.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Date start',
    description: 'The date the security was initially issued.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Date first payment',
    description: 'The first date interest will be paid.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Date end',
    description: 'The security\'s settlement date.',
  }, {
    type: 'number',
    title: 'Rate',
    description: 'The annualized rate of interest.',
  }, {
    type: 'number',
    title: 'Redemption',
    description: 'The redemption amount per 100 face value, or par.',
  }, {
    type: 'number',
    title: 'Frequency',
    description: 'Yearly interest/coupon payments count (1, 2, or 4).',
    enum: [1, 2, 4],
  }, {
    type: 'string',
    title: 'basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        startDate: '2000-01-01T13:06:15+03:00',
        firstDate: '2009-01-22T13:06:15+03:00',
        endDate: '2010-01-22T13:06:15+03:00',
      },
      expression: 'ACCRINT(startDate, firstDate, endDate, 0.8, 200, 2)',
      result: 1609.33333333,
    },
    {
      context: {
        startDate: '2000-01-01T13:06:15+03:00',
        firstDate: '2009-01-22T13:06:15+03:00',
        endDate: '2010-01-22T13:06:15+03:00',
      },
      expression: 'ACCRINT(startDate, firstDate, endDate, 0.8, 200, 2, "us_nasd")',
      result: 1609.33333333,
    },
    {
      context: {
        startDate: '2000-01-01T13:06:15+03:00',
        firstDate: '2009-01-22T13:06:15+03:00',
        endDate: '2010-01-22T13:06:15+03:00',
      },
      expression: 'ACCRINT(startDate, firstDate, endDate, 0.8, 200, 2, "actual_365")',
      result: 1610.52054795,
    },
  ],
};
