// @sapappgyver/appgyver-formulas/functions/financial/AMORLINC.js
const execute = require('./AMORLINC.runtime');

module.exports = {
  name: 'AMORLINC',
  category: 'Financial',
  title: 'Depreciation for accounting period',
  description: 'Returns the depreciation for an accounting period. If the asset was purchased during the period, returns the prorated depreciation. This provision is for users of the French accounting system.',
  parameters: [{
    type: 'number',
    title: 'Cost',
    description: 'The asset\'s purchase cost',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Purchase date',
    description: 'The date the asset was purchased',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'First period end',
    description: 'The end date of the first period.',
  }, {
    type: 'number',
    title: 'Salvage',
    description: 'The asset\'s value at the end of its life (i.e. its salvage value)',
  }, {
    type: 'number',
    title: 'Period',
    description: 'The period for which to calculate depreciation',
  }, {
    type: 'number',
    title: 'Rate',
    description: 'The annual depreciation rate.',
  }, {
    type: 'string',
    title: 'basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      context: {
        date1: '1969-07-20T13:06:15+03:00',
        date2: '1969-08-20T13:06:15+03:00',
      },
      expression: 'AMORLINC(1000, date1, date2, 100, 1.9, 0.15)',
      result: 150,
    },
    {
      context: {
        date3: '2000-07-20T13:06:15+03:00',
        date4: '2002-08-20T13:06:15+03:00',
      },
      expression: 'AMORLINC(2000, date3, date4, 300, 1.9, 0.12, "actual")',
      result: 240,
    },
  ],
};
