// @sapappgyver/appgyver-formulas/functions/financial/COUPDAYS.js
const execute = require('./COUPDAYS.runtime');

module.exports = {
  name: 'COUPDAYS',
  category: 'Financial',
  title: 'Number of days in coupon period for settlement date',
  description: 'Given a settlement date, calculates how many days are in the containing interest payment/coupon period.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Frequency',
    description: 'Yearly interest/coupon payments count (1, 2, or 4).',
    enum: [1, 2, 4],
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'number' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2021-02-01T13:06:15+03:00',
        date2: '2023-12-31T13:06:15+03:00',
      },
      expression: 'COUPDAYS(date1, date2, 2, "actual")',
      result: 181,
    },
    {
      context: {
        date1: '2021-02-01T13:06:15+03:00',
        date2: '2023-12-31T13:06:15+03:00',
      },
      expression: 'COUPDAYS(date1, date2, 4, "actual")',
      result: 90,
    },
    {
      context: {
        date1: '2021-02-01T13:06:15+03:00',
        date2: '2023-12-31T13:06:15+03:00',
      },
      expression: 'COUPDAYS(date1, date2, 2)',
      result: 180,
    },
  ],
};
