// @sapappgyver/appgyver-formulas/functions/financial/COUPNUM.js
const execute = require('./COUPNUM.runtime');

module.exports = {
  name: 'COUPNUM',
  category: 'Financial',
  title: 'Number of coupons between dates',
  description: 'Calculates the number of interest payments/coupons between the given settlement data and the investment\'s maturity date.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Frequency',
    description: 'Yearly interest/coupon payments count (1, 2, or 4).',
    enum: [1, 2, 4],
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'number' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2020-02-01T13:06:15+03:00',
        date2: '2023-12-31T13:06:15+03:00',
      },
      expression: 'COUPNUM(date1, date2, 2, "actual")',
      result: 8,
    },
    {
      context: {
        date3: '2020-01-31T13:06:15+03:00',
        date4: '2023-10-31T13:06:15+03:00',
      },
      expression: 'COUPNUM(date3, date4, 1, "us_nasd")',
      result: 4,
    },
    {
      context: {
        date5: '2020-01-30T13:06:15+03:00',
        date6: '2023-10-30T13:06:15+03:00',
      },
      expression: 'COUPNUM(date5, date6, 1)',
      result: 4,
    },
    {
      context: {
        date7: '2020-01-31T13:06:15+03:00',
        date8: '2023-10-30T13:06:15+03:00',
      },
      expression: 'COUPNUM(date7, date8, 4, "european")',
      result: 15,
    },
  ],
};
