// @sapappgyver/appgyver-formulas/functions/financial/COUPNCD.js
const execute = require('./COUPNCD.runtime');

module.exports = {
  name: 'COUPNCD',
  category: 'Financial',
  title: 'Next coupon date after settlement date',
  description: 'Calculates the next interest payment/coupon date after the given settlement date.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Frequency',
    description: 'Yearly interest/coupon payments count (1, 2, or 4).',
    enum: [1, 2, 4],
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'string', format: 'date-time' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2010-02-01T13:06:15+03:00',
        date2: '2019-12-31T13:06:15+03:00',
      },
      expression: 'COUPNCD(date1, date2, 1)',
      result: '2010-12-31T13:06:15+03:00',
    },
    {
      context: {
        date1: '2010-02-01T13:06:15+03:00',
        date2: '2019-12-31T13:06:15+03:00',
      },
      expression: 'COUPNCD(date1, date2, 1, "us_nasd")',
      result: '2010-12-31T13:06:15+03:00',
    },
    {
      context: {
        date3: '2014-01-10T13:06:15+03:00',
        date4: '2017-05-20T13:06:15+03:00',
      },
      expression: 'COUPNCD(date3, date4, 4, "actual_360")',
      result: '2014-02-20T13:06:15+03:00',
    },
  ],
};
