// @sapappgyver/appgyver-formulas/functions/financial/CUMIPMT.js
const execute = require('./CUMIPMT.runtime');

module.exports = {
  name: 'CUMIPMT',
  category: 'Financial',
  title: 'Cumulative interest',
  description: 'Calculates the cumulative interest for an investment over multiple payment periods. This is done based on a constant interest rate and periodic payments with constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Number of periods',
    description: 'The number of payments to be made',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The current value of the annuity.',
  }, {
    type: 'number',
    title: 'First period',
    description: 'The number of the payment period to begin the cumulative calculation.',
  }, {
    type: 'number',
    title: 'Last period',
    description: 'The number of the payment period to end the cumulative calculation.',
  }, {
    type: 'boolean',
    title: 'Payments at the beginning of period',
    description: 'Whether payments are due at the end (false) or beginning (true) of each period',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'CUMIPMT(0.012, 12, 100, 1, 5, false)', result: -5.05299881 },
    { expression: 'CUMIPMT(0.02, 12, 100, 1, 5, false)', result: -8.47868480 },
    { expression: 'CUMIPMT(0.01, 12, 100, 1, 5, true)', result: -3.17186904 },
  ],
};
