// @sapappgyver/appgyver-formulas/functions/financial/CUMPRINC.js
const execute = require('./CUMPRINC.runtime');

module.exports = {
  name: 'CUMPRINC',
  category: 'Financial',
  title: 'Cumulative principal paid',
  description: 'Calculates the cumulative principle paid on an investment over multiple payment periods. This is done based on a constant interest rate and periodic payments with constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Number of periods',
    description: 'The number of payments to be made',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The current value of the annuity.',
  }, {
    type: 'number',
    title: 'First period',
    description: 'The number of the payment period to begin the cumulative calculation.',
  }, {
    type: 'number',
    title: 'Last period',
    description: 'The number of the payment period to end the cumulative calculation.',
  }, {
    type: 'boolean',
    title: 'Payments at the beginning of period',
    description: 'Whether payments are due at the end (false) or beginning (true) of each period',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'CUMPRINC(0.01, 12, 100, 1, 5, true)', result: -40.81267981 },
    { expression: 'CUMPRINC(0.01, 12, 100, 1, 5, false)', result: -40.22080661 },
    { expression: 'CUMPRINC(0.12, 12, 100, 1, 5, false)', result: -26.32417137 },
  ],
};
