// @sapappgyver/appgyver-formulas/functions/financial/DECLINING_BALANCE.js
const execute = require('./DECLINING_BALANCE.runtime');

const months = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];

module.exports = {
  name: 'DECLINING_BALANCE',
  category: 'Financial',
  title: 'Arithmetic declining balance method',
  description: 'Calculates an assets depreciation over a period of time utilising the arithmetic declining balance method.',
  parameters: [{
    type: 'number',
    title: 'Cost',
    description: 'The initial cost of the asset',
  }, {
    type: 'number',
    title: 'Salvage',
    description: 'The value of the asset at the end of depreciation',
  }, {
    type: 'number',
    title: 'Life',
    description: 'The number of periods over which the asset is depreciated',
  }, {
    type: 'number',
    title: 'Period',
    description: 'The single period within life for which to calculate depreciation',
  }, {
    type: 'number',
    title: 'Month',
    description: 'The number of months in the first year of depreciation',
    default: 12,
    enum: months,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'DECLINING_BALANCE(750, 60, 5, 4)', result: 65.28 },
    { expression: 'DECLINING_BALANCE(750, 60, 5, 5)', result: 39.37 },
    { expression: 'DECLINING_BALANCE(750, 60, 6, 5)', result: 47.78 },
    { expression: 'DECLINING_BALANCE(750, 100, 5, 4)', result: 74.22 },
    { expression: 'DECLINING_BALANCE(1000, 100, 5, 4)', result: 92.71 },
    { expression: 'DECLINING_BALANCE(1000, 100, 5, 4, 12)', result: 92.71 },
    { expression: 'DECLINING_BALANCE(1000, 100, 5, 4, 1)', result: 142.40 },
  ],
};
