// @sapappgyver/appgyver-formulas/functions/financial/DOUBLE_DECLINING_BALANCE.js
const execute = require('./DOUBLE_DECLINING_BALANCE.runtime');

module.exports = {
  name: 'DOUBLE_DECLINING_BALANCE',
  category: 'Financial',
  title: 'Double-declining balance method',
  description: 'Calculates an assets depreciation over a period of time utilising the double-declining balance method.',
  parameters: [{
    type: 'number',
    title: 'Cost',
    description: 'The initial cost of the asset',
  }, {
    type: 'number',
    title: 'Salvage',
    description: 'The value of the asset at the end of depreciation',
  }, {
    type: 'number',
    title: 'Life',
    description: 'The number of periods over which the asset is depreciated',
  }, {
    type: 'number',
    title: 'Period',
    description: 'The single period within life for which to calculate depreciation',
  }, {
    type: 'number',
    title: 'Factor',
    description: 'The factor by which depreciation decreases (default is 2)',
    default: 2,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'DOUBLE_DECLINING_BALANCE(750, 60, 5, 4)', result: 64.8 },
    { expression: 'DOUBLE_DECLINING_BALANCE(750, 60, 5, 5)', result: 37.20 },
    { expression: 'DOUBLE_DECLINING_BALANCE(750, 60, 6, 5)', result: 49.38 },
    { expression: 'DOUBLE_DECLINING_BALANCE(750, 100, 5, 4)', result: 62 },
    { expression: 'DOUBLE_DECLINING_BALANCE(1000, 100, 5, 4)', result: 86.40 },
    { expression: 'DOUBLE_DECLINING_BALANCE(1000, 100, 5, 4, 2)', result: 86.40 },
    { expression: 'DOUBLE_DECLINING_BALANCE(1000, 100, 5, 4, 1)', result: 102.40 },
  ],
};
