// @sapappgyver/appgyver-formulas/functions/financial/FUTURE_VALUE_SCHEDULE.js
const execute = require('./FUTURE_VALUE_SCHEDULE.runtime');

module.exports = {
  name: 'FUTURE_VALUE_SCHEDULE',
  category: 'Financial',
  title: 'Future value by schedule',
  description: 'Calculates a given principal\'s future value using a specified collection of potentially varying interest rates.',
  parameters: [{
    type: 'number',
    title: 'Principal',
    description: 'The amount of initial capital or value to compound against',
  }, {
    type: 'array',
    title: 'Rate schedule',
    description: 'A series of interest rates to compound against the principal',
    items: {
      type: 'number',
    },
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'FUTURE_VALUE_SCHEDULE(1, [1, 2, 3, 4, 5])', result: 720 },
    { expression: 'FUTURE_VALUE_SCHEDULE(2, [1, 2, 3, 4, 5])', result: 1440 },
    { expression: 'FUTURE_VALUE_SCHEDULE(-1, [1, 2, 3, 4, 5])', result: -720 },
    { expression: 'FUTURE_VALUE_SCHEDULE(1, [1, -2, 3, 4, 5])', result: -240 },
  ],
};
