// @sapappgyver/appgyver-formulas/functions/financial/FUTURE_VALUE.js
const execute = require('./FUTURE_VALUE.runtime');

module.exports = {
  name: 'FUTURE_VALUE',
  category: 'Financial',
  title: 'Future value',
  description: 'Calculates an annuity investment\'s future value. This is done based on a constant interest rate and periodic payments of constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Periods',
    description: 'The number of payments to be made',
  }, {
    type: 'number',
    title: 'Payment amount',
    description: 'The amount per period to be paid',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The current value of the annuity. Default is 0.',
    default: 0,
  }, {
    type: 'boolean',
    title: 'Payments at the beginning of period',
    description: 'Whether payments are due at the end (false) or beginning (true) of each period',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'FUTURE_VALUE(0.1, 4, 100)', result: -464.10 },
    { expression: 'FUTURE_VALUE(0.5, 4, 100)', result: -812.50 },
    { expression: 'FUTURE_VALUE(0.5, 6, 100)', result: -2078.13 },
    { expression: 'FUTURE_VALUE(0.5, 6, 100, 200)', result: -4356.25 },
    { expression: 'FUTURE_VALUE(0.5, 6, 100, 200, true)', result: -5395.31 },
  ],
};
