// @sapappgyver/appgyver-formulas/functions/financial/INTRATE.js
const execute = require('./INTRATE.runtime');

module.exports = {
  name: 'INTRATE',
  category: 'Financial',
  title: 'Effective interest rate',
  description: 'Calculates the effective interest rate when an investment is bought for one value and sold for another where no interest or dividends were generated by the investment.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Buy date',
    description: 'The date of purchase of the investment.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Sell date',
    description: 'The date of sale of the investment.',
  }, {
    type: 'number',
    title: 'Buy price',
    description: 'The price at which the investment was purchased.',
  }, {
    type: 'number',
    title: 'Sell price',
    description: 'The price at which the investment was sold.',
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'number' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2019-12-31T13:06:15+03:00',
      },
      expression: 'INTRATE(date1, date2, 90, 140, "us_nasd")',
      result: 0.05557099,
    },
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2019-12-31T13:06:15+03:00',
      },
      expression: 'INTRATE(date1, date2, 90, 140)',
      result: 0.05557099,
    },
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2019-12-31T13:06:15+03:00',
      },
      expression: 'INTRATE(date1, date2, 90, 140, "actual")',
      result: 0.055586,
    },
    {
      context: {
        date3: '2009-01-31T13:06:15+03:00',
        date4: '2023-10-31T13:06:15+03:00',
      },
      expression: 'INTRATE(date3, date4, 290, 300, "us_nasd")',
      result: 0.00233781,
    },
    {
      context: {
        date5: '2009-01-30T13:06:15+03:00',
        date6: '2023-10-30T13:06:15+03:00',
      },
      expression: 'INTRATE(date5, date6, 290, 300, "actual_360")',
      result: 0.00230483,
    },
    {
      context: {
        date7: '2009-01-31T13:06:15+03:00',
        date8: '2023-10-30T13:06:15+03:00',
      },
      expression: 'INTRATE(date7, date8, 290, 300, "actual")',
      result: 0.00233855,
    },
  ],
};
