// @sapappgyver/appgyver-formulas/functions/financial/INVESTMENT_DURATION.js
const execute = require('./INVESTMENT_DURATION.runtime');

module.exports = {
  name: 'INVESTMENT_DURATION',
  category: 'Financial',
  title: 'Number of compounding periods required for investment',
  description: 'Calculates how many compounding periods are required for an investment with a given current value that appreciates at a given rate in order to reach a specified target value.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Rate',
    description: 'The annualized rate of interest.',
  }, {
    type: 'number',
    title: 'Yield',
    description: 'The expected annual yield of the security.',
  }, {
    type: 'number',
    title: 'Frequency',
    description: 'Yearly interest/coupon payments count (1, 2, or 4).',
    enum: [1, 2, 4],
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'number' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2010-01-01T13:06:15+03:00',
        date2: '2039-01-01T13:06:15+03:00',
      },
      expression: 'INVESTMENT_DURATION(date1, date2, 0.03, 0.12, 2, "us_nasd")',
      result: 10.4300076,
    },
    {
      context: {
        date1: '2010-01-01T13:06:15+03:00',
        date2: '2039-01-01T13:06:15+03:00',
      },
      expression: 'INVESTMENT_DURATION(date1, date2, 0.03, 0.12, 2)',
      result: 10.4300076,
    },
    {
      context: {
        date3: '2009-01-30T13:06:15+03:00',
        date4: '2023-07-30T13:06:15+03:00',
      },
      expression: 'INVESTMENT_DURATION(date3, date4, 0.05, 1.5, 4, "us_nasd")',
      result: 0.91667041,
    },
    {
      context: {
        date3: '2009-01-30T13:06:15+03:00',
        date4: '2023-07-30T13:06:15+03:00',
      },
      expression: 'INVESTMENT_DURATION(date3, date4, 0.05, 1.5, 4, "actual_360")',
      result: 1.12222596,
    },
  ],
};
