// @sapappgyver/appgyver-formulas/functions/financial/INVESTMENT_PDURATION.js
const execute = require('./INVESTMENT_PDURATION.runtime');

module.exports = {
  name: 'INVESTMENT_PDURATION',
  category: 'Financial',
  title: 'Periods required for investment to grow to given value',
  description: 'Calculates how many periods an investment needs to grow to a given value at a given rate.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The rate at which the investment grows each period',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The investment\'s current value.',
  }, {
    type: 'number',
    title: 'Future value',
    description: 'The investment\'s desired future value.',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'INVESTMENT_PDURATION(0.1, 10, 20)', result: 7.272540897 },
    { expression: 'INVESTMENT_PDURATION(1, 10, 20)', result: 1 },
    { expression: 'INVESTMENT_PDURATION(1, 10, 30)', result: 1.584962501 },
    { expression: 'INVESTMENT_PDURATION(1, 20, 20)', result: 0 },
  ],
};
