// @sapappgyver/appgyver-formulas/functions/financial/IPMT.js
const execute = require('./IPMT.runtime');

module.exports = {
  name: 'IPMT',
  category: 'Financial',
  title: 'Payment on interest',
  description: 'Calculates the cumulative interest payments on an investment. This is done based on a constant interest rate and periodic payments of constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Period',
    description: 'The amortization period, between 1 and **Number of periods**.',
  }, {
    type: 'number',
    title: 'Number of periods',
    description: 'The number of payments to be made',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The current value of the annuity.',
  }, {
    type: 'number',
    title: 'Future value',
    description: 'Future value left when the final payment has been completed, defaults to 0.',
    default: 0,
  }, {
    type: 'boolean',
    title: 'Payments at the beginning of period',
    description: 'Whether payments are due at the end (false) or beginning (true) of each period',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'IPMT(0.1, 4, 12, 100)', result: -8.45 },
    { expression: 'IPMT(0.5, 4, 12, 100)', result: -49.08 },
    { expression: 'IPMT(0.5, 6, 12, 100)', result: -47.44 },
    { expression: 'IPMT(0.5, 6, 12, 100, 200)', result: -42.32 },
    { expression: 'IPMT(0.5, 6, 12, 100, 200, true)', result: -28.21 },
  ],
};
