// @sapappgyver/appgyver-formulas/functions/financial/IRR.js
const execute = require('./IRR.runtime');

module.exports = {
  name: 'IRR',
  category: 'Financial',
  title: 'Investment\'s internal rate of return',
  description: 'Calculates an investment\'s internal rate of return using a series of periodic cash flows.',
  parameters: [{
    type: 'array',
    title: 'Сashflow amounts',
    description: 'A list containing the income or payments associated with the investment.',
  }, {
    type: 'number',
    title: 'Rate guess',
    description: 'An estimate for what the internal rate of return will be.',
    default: 0.1,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      expression: 'IRR(array1)',
      context: {
        array1: [-20000, -1000, 9000, 15000],
      },
      result: 0.05327951,
    },
    {
      expression: 'IRR(array1, 0.1)',
      context: {
        array1: [-20000, -1000, 9000, 15000],
      },
      result: 0.05327951,
    },
    {
      expression: 'IRR(array2, 0.8)',
      context: {
        array2: [-100000, -80000, -40000, 10000, 80000, 100000, 150000],
      },
      result: 0.10378603,
    },
  ],
};
