// @sapappgyver/appgyver-formulas/functions/financial/ISPMT.js
const execute = require('./ISPMT.runtime');

module.exports = {
  name: 'ISPMT',
  category: 'Financial',
  title: 'Investment\'s interest paid',
  description: 'Given a particular period, calculates an investment\'s interest paid. This is done based on a constant interest rate and periodic payments of constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Period',
    description: 'The time frame for which you want to view the interest payment.',
  }, {
    type: 'number',
    title: 'Number of periods',
    description: 'The number of payments to be made.',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The current value of the annuity.',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'ISPMT(0.1, 3, 12, 100)', result: -7.5 },
    { expression: 'ISPMT(1, 3, 12, 100)', result: -75 },
    { expression: 'ISPMT(1, 6, 12, 100)', result: -50 },
    { expression: 'ISPMT(1, 3, 12, 200)', result: -150 },
  ],
};
