// @sapappgyver/appgyver-formulas/functions/financial/MIRR.js
const execute = require('./MIRR.runtime');

module.exports = {
  name: 'MIRR',
  category: 'Financial',
  title: 'Modified internal rate',
  description: 'Given a series of periodic cash flows, and the difference between the interest rate paid on financing versus the return received on reinvested income, calculates the modified internal rate of return as a decimal number.',
  parameters: [{
    type: 'array',
    title: 'Сashflow amounts',
    description: 'A list containing the income or payments associated with the investment.',
    items: {
      type: 'number',
    },
  }, {
    type: 'number',
    title: 'Financing rate',
    description: 'The interest rate paid on funds invested.',
  }, {
    type: 'number',
    title: 'Reinvestment return rate',
    description: 'The return (as a percentage) earned on reinvestment of income received from the investment',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'MIRR([1, -1, 2, 3, 4, 5], 1, 1)', result: 1.709384058 },
    { expression: 'MIRR([1, -1, 2, 3, 4, 5], 1, 2.5)', result: 3.21691195 },
    { expression: 'MIRR([2, -2, 2, 3, 4, 5], 0, 1)', result: 1.208166722 },
    { expression: 'MIRR([1, -1, 2, 3, 4, 5], 2, 1)', result: 1.938250528 },
  ],
};
