// @sapappgyver/appgyver-formulas/functions/financial/NPER.js
const execute = require('./NPER.runtime');

module.exports = {
  name: 'NPER',
  category: 'Financial',
  title: 'Number of payment periods',
  description: 'Calculates how many payment periods are needed for an investment. This is done based on a constant interest rate and periodic payments of constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Payment amount',
    description: 'The amount of each payment made',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The current value of the annuity',
  }, {
    type: 'number',
    title: 'Future value',
    description: 'Future value left when the final payment has been completed, defaults to 0.',
    default: 0,
  }, {
    type: 'boolean',
    title: 'Payments at the beginning of period',
    description: 'Whether payments are due at the end (false) or beginning (true) of each period',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'NPER(0.1, 4, 100)', result: -13.14406259 },
    { expression: 'NPER(0.5, 4, 100)', result: -6.419022583 },
    { expression: 'NPER(0.5, 6, 100)', result: -5.508716229 },
    { expression: 'NPER(0.5, 6, 100, -200)', result: 1.573717172 },
    { expression: 'NPER(0.5, 6, 100, -200, true)', result: 1.513842809 },
  ],
};
