// @sapappgyver/appgyver-formulas/functions/financial/PMT.js
const execute = require('./PMT.runtime');

module.exports = {
  name: 'PMT',
  category: 'Financial',
  title: 'Annuity investment\'s periodic payment',
  description: 'Calculates an annuity investment\'s periodic payment. This is done based on a constant interest rate and periodic payments of constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Number of periods',
    description: 'The number of payments to be made',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The current value of the annuity.',
  }, {
    type: 'number',
    title: 'Future value',
    description: 'The future value remaining after the final payment has been made (Default is 0).',
    default: 0,
  }, {
    type: 'boolean',
    title: 'Payments at the beginning of period',
    description: 'Whether payments are due at the end (false) or beginning (true) of each period',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'PMT(0.1, 4, 100)', result: -31.55 },
    { expression: 'PMT(0.5, 4, 100)', result: -62.31 },
    { expression: 'PMT(0.5, 6, 100)', result: -54.81 },
    { expression: 'PMT(0.5, 6, 100, 200)', result: -64.44 },
    { expression: 'PMT(0.5, 6, 100, 200, true)', result: -42.96 },
  ],
};
