// @sapappgyver/appgyver-formulas/functions/financial/PPMT.js
const execute = require('./PPMT.runtime');

module.exports = {
  name: 'PPMT',
  category: 'Financial',
  title: 'Investment\'s payment on interest',
  description: 'Calculates an investment\'s payment on interest. This is done based on a constant interest rate and periodic payments of constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Period',
    description: 'The amortization period, between 1 and **Number of periods**. ',
  }, {
    type: 'number',
    title: 'Number of periods',
    description: 'The number of payments to be made',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The current value of the annuity.',
  }, {
    type: 'number',
    title: 'Future value',
    description: 'The future value remaining after the final payment has been made (Default is 0).',
    default: 0,
  }, {
    type: 'boolean',
    title: 'Payments at the beginning of period',
    description: 'Whether payments are due at the end (false) or beginning (true) of each period',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'PPMT(0.1, 4, 12, 100)', result: -6.22 },
    { expression: 'PPMT(0.5, 4, 12, 100)', result: -1.31 },
    { expression: 'PPMT(0.5, 6, 12, 100)', result: -2.95 },
    { expression: 'PPMT(0.5, 6, 12, 100, 200)', result: -8.85 },
    { expression: 'PPMT(0.5, 6, 12, 100, 200, true)', result: -5.9 },
  ],
};
