// @sapappgyver/appgyver-formulas/functions/financial/PRESENT_VALUE.js
const execute = require('./PRESENT_VALUE.runtime');

module.exports = {
  name: 'PRESENT_VALUE',
  category: 'Financial',
  title: 'Annuity investment\'s present value',
  description: 'Calculates an annuity investment\'s present value. This is done based on a constant interest rate and periodic payments of constant amounts.',
  parameters: [{
    type: 'number',
    title: 'Rate',
    description: 'The interest rate',
  }, {
    type: 'number',
    title: 'Periods',
    description: 'The number of payments to be made',
  }, {
    type: 'number',
    title: 'Payment amount',
    description: 'The amount per period to be paid',
  }, {
    type: 'number',
    title: 'Future value',
    description: 'Future value left when the final payment has been completed, defaults to 0.',
    default: 0,
  }, {
    type: 'boolean',
    title: 'Payments at the beginning of period',
    description: 'Whether payments are due at the end (false) or beginning (true) of each period',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'PRESENT_VALUE(0.1, 4, 100)', result: -316.99 },
    { expression: 'PRESENT_VALUE(0.5, 4, 100)', result: -160.49 },
    { expression: 'PRESENT_VALUE(0.5, 6, 100)', result: -182.44 },
    { expression: 'PRESENT_VALUE(0.5, 6, 100, 200)', result: -200.00 },
    { expression: 'PRESENT_VALUE(0.5, 6, 100, 200, true)', result: -291.22 },
  ],
};
