// @sapappgyver/appgyver-formulas/functions/financial/PRICE.js
const execute = require('./PRICE.runtime');

module.exports = {
  name: 'PRICE',
  category: 'Financial',
  title: 'Price of security that pays periodic interest',
  description: 'For a security that pays periodic interest, given the expected yield, calculates the security\'s price.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Rate',
    description: 'The annualized rate of interest.',
  }, {
    type: 'number',
    title: 'Yield',
    description: 'The expected annual yield of the security.',
  }, {
    type: 'number',
    title: 'Redemption',
    description: 'The redemption value of the security.',
  }, {
    type: 'number',
    title: 'Frequency',
    description: 'Yearly interest/coupon payments count (1, 2, or 4).',
    enum: [1, 2, 4],
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'number' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2039-12-31T13:06:15+03:00',
      },
      expression: 'PRICE(date1, date2, 0.03, 1.2, 100, 2, "us_nasd")',
      result: 2.49642314,
    },
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2039-12-31T13:06:15+03:00',
      },
      expression: 'PRICE(date1, date2, 0.03, 1.2, 100, 2)',
      result: 2.49642314,
    },
    {
      context: {
        date3: '2009-01-31T13:06:15+03:00',
        date4: '2023-10-31T13:06:15+03:00',
      },
      expression: 'PRICE(date3, date4, 0.05, 1.5, 300, 4, "us_nasd")',
      result: 3.33333539,
    },
    {
      context: {
        date5: '2009-01-30T13:06:15+03:00',
        date6: '2023-10-30T13:06:15+03:00',
      },
      expression: 'PRICE(date5, date6, 0.05, 1.5, 300, 4, "actual_360")',
      result: 3.33333539,
    },
    {
      context: {
        date7: '2009-01-31T13:06:15+03:00',
        date8: '2023-10-30T13:06:15+03:00',
      },
      expression: 'PRICE(date7, date8, 0.05, 1.5, 300, 1, "actual")',
      result: 2.93632873,
    },
  ],
};
