// @sapappgyver/appgyver-formulas/functions/financial/PRICEMAT.js
const execute = require('./PRICEMAT.runtime');

module.exports = {
  name: 'PRICEMAT',
  category: 'Financial',
  title: 'Price of security that pays interest at maturity',
  description: 'For a security that pays interest at maturity, given the expected yield, calculates the security\'s price.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Issue',
    description: 'The date the security was initially issued.',
  }, {
    type: 'number',
    title: 'Rate',
    description: 'The annualized rate of interest.',
  }, {
    type: 'number',
    title: 'Yield',
    description: 'The expected annual yield of the security.',
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'number' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2039-12-31T13:06:15+03:00',
        date3: '2010-01-01T13:06:15+03:00',
      },
      expression: 'PRICEMAT(date1, date2, date3, 3, 1.2, "us_nasd")',
      result: 245.13477190,
    },
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2039-12-31T13:06:15+03:00',
        date3: '2010-01-01T13:06:15+03:00',
      },
      expression: 'PRICEMAT(date1, date2, date3, 3, 1.2)',
      result: 245.13477190,
    },
    {
      context: {
        date4: '2009-01-31T13:06:15+03:00',
        date5: '2023-10-31T13:06:15+03:00',
        date6: '2009-01-30T13:06:15+03:00',
      },
      expression: 'PRICEMAT(date4, date5, date6, 5, 1, "actual")',
      result: 473.31715882,
    },
    {
      context: {
        date6: '2009-01-30T13:06:15+03:00',
        date7: '2023-10-30T13:06:15+03:00',
        date8: '2008-10-30T13:06:15+03:00',
      },
      expression: 'PRICEMAT(date6, date7, date8, 5, 1, "actual_360")',
      result: 355.16687938,
    },
  ],
};
