// @sapappgyver/appgyver-formulas/functions/financial/PRICEDISC.js
const execute = require('./PRICEDISC.runtime');

module.exports = {
  name: 'PRICEDISC',
  category: 'Financial',
  title: 'Price of discount security',
  description: 'For a discount security (i.e. one that doesn\'t bear interest), given the expected yield, calculates the security\'s price.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Discount',
    description: 'The discount rate of the security at time of purchase',
  }, {
    type: 'number',
    title: 'Redemption',
    description: 'The redemption value of the security.',
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'number' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2039-12-31T13:06:15+03:00',
      },
      expression: 'PRICEDISC(date1, date2, 0.03, 100, "us_nasd")',
      result: 10.00833333,
    },
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2039-12-31T13:06:15+03:00',
      },
      expression: 'PRICEDISC(date1, date2, 0.03, 100)',
      result: 10.00833333,
    },
    {
      context: {
        date3: '2009-01-31T13:06:15+03:00',
        date4: '2023-10-31T13:06:15+03:00',
      },
      expression: 'PRICEDISC(date3, date4, 0.05, 300, "us_nasd")',
      result: 78.75,
    },
    {
      context: {
        date5: '2009-01-30T13:06:15+03:00',
        date6: '2023-10-30T13:06:15+03:00',
      },
      expression: 'PRICEDISC(date5, date6, 0.05, 300, "actual_360")',
      result: 75.58333333,
    },
    {
      context: {
        date7: '2009-01-31T13:06:15+03:00',
        date8: '2023-10-30T13:06:15+03:00',
      },
      expression: 'PRICEDISC(date7, date8, 0.05, 300, "actual")',
      result: 78.81982475,
    },
  ],
};
