// @sapappgyver/appgyver-formulas/functions/financial/RATE.js
const execute = require('./RATE.runtime');

module.exports = {
  name: 'RATE',
  category: 'Financial',
  title: 'Interest rate for annuity investment',
  description: 'Calculates an annuity investment\'s interest rate. This is done based on a constant interest rate and constant-rate periodic payments.',
  parameters: [{
    type: 'number',
    title: 'Number of periods',
    description: 'The total number of payment periods in an annuity',
  }, {
    type: 'number',
    title: 'Payment per period',
    description: 'The payment made each period, does not change over the life of the annuity',
  }, {
    type: 'number',
    title: 'Present value',
    description: 'The present value — the total amount that a series of future payments is worth now',
  }, {
    type: 'number',
    title: 'Future value',
    description: 'Future value left when the final payment has been completed, defaults to 0.',
    default: 0,
  }, {
    type: 'boolean',
    title: 'End or beginning',
    description: 'Whether payments are due at the end (false) or beginning (true)',
    default: false,
  }, {
    type: 'number',
    title: 'Rate guess',
    description: 'An estimate for what the interest rate will be',
    default: 0.1,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'RATE(12, -100, 400)', result: 0.22893307 },
    { expression: 'RATE(12, -100, 400, 2)', result: 0.22881153 },
    { expression: 'RATE(12, -100, 400, 2, false)', result: 0.22881153 },
    { expression: 'RATE(12, -100, 400, 2, true)', result: -0.98039216 },
    { expression: 'RATE(12, -100, 400, 2, true, 0.1)', result: -0.98039216 },
  ],
};
