// @sapappgyver/appgyver-formulas/functions/financial/RECEIVED.js
const execute = require('./RECEIVED.runtime');

module.exports = {
  name: 'RECEIVED',
  category: 'Financial',
  title: 'Amount received at maturity for fixed-income security investment',
  description: 'For a fixed-income security investment, given the purchase date, calculates the amount received at maturity.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Investment',
    description: 'The amount invested (irrespective of face value of each security).',
  }, {
    type: 'number',
    title: 'Discount',
    description: 'The discount rate of the security invested in.',
  }, {
    type: 'string',
    title: 'Basis',
    description: 'Method to use for counting days per month/year:\n'
    + '\n- `us_nasd` (default): 30-day month, 360-day year, i.e. US (NASD).'
    + '\n- `actual`: Actual number of days per month/year.'
    + '\n- `actual_360`: 360 day year, actual number of days per month.'
    + '\n- `actual_365`: 365 day year, actual number of days per month. This means leap years are not taken into account.'
    + '\n- `european`: Like `us_nasd`, but for months that are not 30 days long, European standards are used.',
    enum: ['us_nasd', 'actual', 'actual_360', 'actual_365', 'european'],
    default: 'us_nasd',
  }],
  analyze: () => ({
    schema: { type: 'number' },
  }),
  execute,
  examples: [
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2039-12-31T13:06:15+03:00',
      },
      expression: 'RECEIVED(date1, date2, 100, 0.05, "us_nasd")',
      result: -200.05557099,
    },
    {
      context: {
        date1: '2010-01-02T13:06:15+03:00',
        date2: '2039-12-31T13:06:15+03:00',
      },
      expression: 'RECEIVED(date1, date2, 100, 0.05)',
      result: -200.05557099,
    },
    {
      context: {
        date3: '2009-01-31T13:06:15+03:00',
        date4: '2023-10-31T13:06:15+03:00',
      },
      expression: 'RECEIVED(date3, date4, 2400, 0.03, "us_nasd")',
      result: 4304.93273543,
    },
    {
      context: {
        date5: '2009-01-30T13:06:15+03:00',
        date6: '2023-10-30T13:06:15+03:00',
      },
      expression: 'RECEIVED(date5, date6, 2800, 0.03, "actual_360")',
      result: 5080.1330511,
    },
    {
      context: {
        date7: '2009-01-31T13:06:15+03:00',
        date8: '2023-10-30T13:06:15+03:00',
      },
      expression: 'RECEIVED(date7, date8, 2500, 0.03, "actual")',
      result: 4483.18192978,
    },
  ],
};
