// @sapappgyver/appgyver-formulas/functions/financial/SYD.js
const execute = require('./SYD.runtime');

module.exports = {
  name: 'SYD',
  category: 'Financial',
  title: 'Sum of years digits method',
  description: 'Calculates an asset\'s depreciation using the sum of years digits method for a specified period.',
  parameters: [{
    type: 'number',
    title: 'Cost',
    description: 'The initial cost of the asset',
  }, {
    type: 'number',
    title: 'Salvage',
    description: 'The value of the asset at the end of depreciation',
  }, {
    type: 'number',
    title: 'Life',
    description: 'The number of periods over which the asset is depreciated',
  }, {
    type: 'number',
    title: 'Period',
    description: 'The single period within life for which to calculate depreciation',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'SYD(100, 50, 5, 1)', result: 16.67 },
    { expression: 'SYD(100, 50, 5, 2)', result: 13.33 },
    { expression: 'SYD(100, 100, 5, 1)', result: 0 },
    { expression: 'SYD(150, 50, 5, 1)', result: 33.33 },
  ],
};
