// @sapappgyver/appgyver-formulas/functions/financial/TBILLYIELD.js
const execute = require('./TBILLYIELD.runtime');

module.exports = {
  name: 'TBILLYIELD',
  category: 'Financial',
  title: 'US Treasury Bill yield based on price',
  description: 'Returns a US Treasury Bill\'s yield based on price.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Price',
    description: 'The price at which the security is bought',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'TBILLYIELD("2010-01-02T13:06:15+03:00", "2010-12-31T13:06:15+03:00", 98.45)', result: 0.01561392 },
    { expression: 'TBILLYIELD("2011-01-01T13:06:15+03:00", "2011-07-16T13:06:15+03:00", 51.78)', result: 1.71045475 },
    { expression: 'TBILLYIELD("2011-09-25T13:06:15+03:00", "2012-02-15T13:06:15+03:00", 48)', result: 2.72727273 },
  ],
};
