// @sapappgyver/appgyver-formulas/functions/financial/TBILLPRICE.js
const execute = require('./TBILLPRICE.runtime');

module.exports = {
  name: 'TBILLPRICE',
  category: 'Financial',
  title: 'US Treasury Bill price based on discount rate',
  description: 'Returns a US Treasury Bill\'s price based on discount rate.',
  parameters: [{
    type: 'string',
    format: 'date-time',
    title: 'Settlement',
    description: 'The security\'s settlement date.',
  }, {
    type: 'string',
    format: 'date-time',
    title: 'Maturity',
    description: 'The security\'s maturity or end date.',
  }, {
    type: 'number',
    title: 'Discount',
    description: 'The discount rate of the bill at time of purchase.',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'TBILLPRICE("2010-01-01T13:06:15+03:00", "2010-12-30T13:06:15+03:00", 0.3)', result: 69.75 },
    { expression: 'TBILLPRICE("2010-01-01T13:06:15+03:00", "2010-06-12T13:06:15+03:00", 0.1)', result: 95.5 },
    { expression: 'TBILLPRICE("2010-01-01T13:06:15+03:00", "2010-03-30T13:06:15+03:00", 0.5)', result: 87.77777778 },
  ],
};
