// @sapappgyver/appgyver-formulas/functions/financial/VDB.js
const execute = require('./VDB.runtime');

module.exports = {
  name: 'VDB',
  category: 'Financial',
  title: 'Depreciation of asset for period',
  description: 'Calculates an asset\'s depreciation for a given period.',
  parameters: [{
    type: 'number',
    title: 'Cost',
    description: 'Initial cost of the asset.',
  }, {
    type: 'number',
    title: 'Salvage',
    description: 'Value of the asset at the end of depreciation (salvage value).',
  }, {
    type: 'number',
    title: 'Life',
    description: 'Number of periods of depreciation (useful life of the asset).',
  }, {
    type: 'number',
    title: 'Start period',
    description: 'Starting period to calculate depreciation.',
  }, {
    type: 'number',
    title: 'End period',
    description: 'Ending period to calculate depreciation.',
  }, {
    type: 'number',
    title: 'Factor',
    description: 'Rate of balance decline (the default is 2: the double-declining balance method).',
    default: 2,
  }, {
    type: 'boolean',
    title: 'No switch',
    description: 'Whether to switch to straight-line depreciation when the depreciation is greater than the declining balance calculation',
    default: false,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    { expression: 'VDB(100, 10, 20, 10, 11, 2, true)', result: 3.48678440 },
    { expression: 'VDB(100, 33, 20, 10, 11, 2, false)', result: 1.86784401 },
    { expression: 'VDB(100, 33, 20, 10, 11, 2)', result: 1.86784401 },
    { expression: 'VDB(100, 33, 20, 10, 11)', result: 1.86784401 },
  ],
};
