// @sapappgyver/appgyver-formulas/functions/financial/XIRR.js
const execute = require('./XIRR.runtime');

module.exports = {
  name: 'XIRR',
  category: 'Financial',
  title: 'Investment internal rate of return',
  description: 'Calculates an investment\'s internal rate of return, using a series of cash flows that are potentially irregular.',
  parameters: [{
    type: 'array',
    items: {
      type: 'number',
    },
    title: 'Сashflow amounts',
    description: 'A list or range containing the income or payments associated with the investment.',
  }, {
    type: 'array',
    items: {
      type: 'string',
      format: 'date-time',
    },
    title: 'Сashflow dates',
    description: 'A list or range with dates corresponding to the cash flows in cashflow amounts.',
  }, {
    type: 'number',
    title: 'Rate guess',
    description: 'An estimate for what the internal rate of return will be.',
    default: 0.1,
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      expression: 'XIRR(values1, dates1)',
      context: {
        values1: [-1000, 1000, 2000, 3000, 5000],
        dates1: ['2000-01-01T13:06:15+03:00', '2004-02-03T13:06:15+03:00', '2008-03-07T13:06:15+03:00', '2012-04-09T13:06:15+03:00', '2016-05-12T13:06:15+03:00'],
      },
      result: 0.2567962754481638,
    },
    {
      expression: 'XIRR(values1, dates1, 0.2)',
      context: {
        values1: [-1000, 1000, 2000, 3000, 5000],
        dates1: ['2000-01-01T13:06:15+03:00', '2004-02-03T13:06:15+03:00', '2008-03-07T13:06:15+03:00', '2012-04-09T13:06:15+03:00', '2016-05-12T13:06:15+03:00'],
      },
      result: 0.25679628,
    },
    {
      expression: 'XIRR(values1, dates2, 0.2)',
      context: {
        values1: [-1000, 1000, 2000, 3000, 5000],
        dates2: ['2000-01-01T13:06:15+03:00', '2001-02-03T13:06:15+03:00', '2002-03-07T13:06:15+03:00', '2003-04-09T13:06:15+03:00', '2004-05-12T13:06:15+03:00'],
      },
      result: 1.35712007,
    },
  ],
};
