// @sapappgyver/appgyver-formulas/functions/financial/XNPV.js
const execute = require('./XNPV.runtime');

module.exports = {
  name: 'XNPV',
  category: 'Financial',
  title: 'Investment net present value',
  description: 'Calculates an investment\'s net present value, using a series of cash flows that are potentially irregular.',
  parameters: [{
    type: 'number',
    title: 'Discount',
    description: 'The discount rate of the investment over one period.',
  }, {
    type: 'array',
    items: {
      type: 'number',
    },
    title: 'Сashflow amounts',
    description: 'A range of cells containing the income or payments associated with the investment.',
  }, {
    type: 'array',
    items: {
      type: 'string',
      format: 'date-time',
    },
    title: 'Сashflow dates',
    description: 'A range of cells with dates corresponding to the cash flows in сashflow amounts',
  }],
  analyze: () => ({ schema: { type: 'number' } }),
  execute,
  examples: [
    {
      expression: 'XNPV(1, cash1, dates1)',
      context: {
        cash1: [1, 2, 3, 4, 5],
        dates1: [
          '2019-07-11T13:06:15+03:00',
          '2019-07-12T13:06:15+03:00',
          '2019-07-13T13:06:15+03:00',
          '2019-07-14T13:06:15+03:00',
          '2019-07-15T13:06:15+03:00',
        ],
      },
      result: 14.92427256,
    },
    {
      expression: 'XNPV(2, cash1, dates1)',
      context: {
        cash1: [1, 2, 3, 4, 5],
        dates1: [
          '2019-07-11T13:06:15+03:00',
          '2019-07-12T13:06:15+03:00',
          '2019-07-13T13:06:15+03:00',
          '2019-07-14T13:06:15+03:00',
          '2019-07-15T13:06:15+03:00',
        ],
      },
      result: 14.88019094,
    },
    {
      expression: 'XNPV(1, cashSmall, datesSmall)',
      context: {
        cashSmall: [1, 2, 3],
        datesSmall: [
          '2019-07-11T13:06:15+03:00',
          '2019-07-12T13:06:15+03:00',
          '2019-07-13T13:06:15+03:00',
        ],
      },
      result: 5.984832948,
    },
  ],
};
