// @sapappgyver/appgyver-formulas/functions/color/HSL.js

const execute = require('./HSL.runtime');

module.exports = {
  name: 'HSL',
  category: 'Color',
  title: 'Generate a color based on given of hue, saturation, lightness and alpha components',
  description: 'Generate a color based on the values of the given hue, saturation, lightness '
    + 'and alpha components. Returns a text representing a valid color, or `null` if the color '
    + 'cannot be generated.',
  parameters: [{
    title: 'Hue',
    description: 'Hue component (`0…360`)',
    type: 'number',
  }, {
    title: 'Saturation',
    description: 'Saturation component (`0…1`)',
    type: 'number',
  }, {
    title: 'Lightness',
    description: 'Lightness component (`0…1`)',
    type: 'number',
  }, {
    title: 'Alpha',
    description: 'Alpha component (`0…1`)',
    type: 'number',
    default: 1,
  }],
  analyze: () => ({ schema: { type: 'string', format: 'color' } }),
  execute,
  examples: [
    { expression: 'HSL(10, 0.2, 0.3)', result: 'rgb(92,66,61)' },
    { expression: 'HSL(10, 0.2, 0.3, 0.5)', result: 'rgba(92,66,61,0.5)' },
    { expression: 'HSL(500, 500, 500, 500)', result: 'rgb(255,0,255)' },
  ],
};
