// @sapappgyver/appgyver-formulas/functions/color/MIX.js

const execute = require('./MIX.runtime');

const defaultRatio = 0.5;

module.exports = {
  name: 'MIX',
  category: 'Color',
  title: 'Mixes the two colors together',
  description: 'Returns a color that is a mix of the given two colors. By default the color is the average of the two colors. You may optionally define a ratio how much the returned color resembles the second color.',
  parameters: [{
    title: 'Color 1',
    description: 'The first color to mix',
    type: 'string',
    format: 'color',
  }, {
    title: 'Color 2',
    description: 'The second color to mix',
    type: 'string',
    format: 'color',
  }, {
    title: 'Ratio',
    description: 'Mix ratio (0...1)',
    type: 'number',
    default: defaultRatio,
  }],
  analyze: () => ({ schema: { type: 'string', format: 'color' } }),
  execute,
  examples: [
    { expression: 'MIX("#fff", "#000", 0.5)', result: 'rgb(180,180,180)' },
    { expression: 'MIX("rgb(10, 20, 40)", "rgb(100, 100, 100)",  0.5)', result: 'rgb(71,72,76)' },
    { expression: 'MIX("hsl(80, 20%, 40%)", "hsl(80, 40%, 20%)")', result: 'rgb(87,100,62)' },
    { expression: 'MIX("red", "blue", 0.5)', result: 'rgb(180,0,180)' },
  ],
};
