// @sapappgyver/appgyver-formulas/functions/color/RGB.js

const execute = require('./RGB.runtime');

module.exports = {
  name: 'RGB',
  category: 'Color',
  title: 'Generate color based on given red, green, blue and alpha values',
  description: 'Generate a color based on the given values for the red, green, blue and alpha components. Returns a text representing the color if valid, or `null` if invalid.',
  parameters: [{
    title: 'Red',
    description: 'Red component',
    type: 'number',
  }, {
    title: 'Green',
    description: 'Green component',
    type: 'number',
  }, {
    title: 'Blue',
    description: 'Red component',
    type: 'number',
  }, {
    title: 'Alpha',
    description: 'Alpha component',
    type: 'number',
    default: 1,
  }],
  analyze: () => ({ schema: { type: 'string', format: 'color' } }),
  execute,
  examples: [
    { expression: 'RGB(125, 125, 125)', result: 'rgb(125,125,125)' },
    { expression: 'RGB(125, 125, 125, 0.5)', result: 'rgba(125,125,125,0.5)' },
    { expression: 'RGB(333, -333, 0, -1)', result: 'rgba(255,0,0,0)' },
  ],
};
