// ts-patch/lib/system/errors.js
"use strict";
/* ********************************************************************************************************************
 * Errors Classes
 * ********************************************************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileWriteError = exports.BackupError = exports.RestoreError = exports.NPMError = exports.OptionsError = exports.PersistenceError = exports.PatchError = exports.PackageError = exports.FileNotFound = exports.WrongTSVersion = void 0;
var WrongTSVersion = /** @class */ (function (_super) {
    __extends(WrongTSVersion, _super);
    function WrongTSVersion() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'WrongTSVersion';
        return _this;
    }
    return WrongTSVersion;
}(Error));
exports.WrongTSVersion = WrongTSVersion;
var FileNotFound = /** @class */ (function (_super) {
    __extends(FileNotFound, _super);
    function FileNotFound() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'FileNotFound';
        return _this;
    }
    return FileNotFound;
}(Error));
exports.FileNotFound = FileNotFound;
var PackageError = /** @class */ (function (_super) {
    __extends(PackageError, _super);
    function PackageError() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'PackageError';
        return _this;
    }
    return PackageError;
}(Error));
exports.PackageError = PackageError;
var PatchError = /** @class */ (function (_super) {
    __extends(PatchError, _super);
    function PatchError() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'PatchError';
        return _this;
    }
    return PatchError;
}(Error));
exports.PatchError = PatchError;
var PersistenceError = /** @class */ (function (_super) {
    __extends(PersistenceError, _super);
    function PersistenceError() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'PersistenceError';
        return _this;
    }
    return PersistenceError;
}(Error));
exports.PersistenceError = PersistenceError;
var OptionsError = /** @class */ (function (_super) {
    __extends(OptionsError, _super);
    function OptionsError() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'OptionsError';
        return _this;
    }
    return OptionsError;
}(Error));
exports.OptionsError = OptionsError;
var NPMError = /** @class */ (function (_super) {
    __extends(NPMError, _super);
    function NPMError() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'NPMError';
        return _this;
    }
    return NPMError;
}(Error));
exports.NPMError = NPMError;
var RestoreError = /** @class */ (function (_super) {
    __extends(RestoreError, _super);
    function RestoreError(filename, message) {
        var _this = _super.call(this, "Error restoring: " + filename + (message ? ' - ' + message : '')) || this;
        _this.filename = filename;
        _this.name = 'RestoreError';
        return _this;
    }
    return RestoreError;
}(Error));
exports.RestoreError = RestoreError;
var BackupError = /** @class */ (function (_super) {
    __extends(BackupError, _super);
    function BackupError(filename, message) {
        var _this = _super.call(this, "Error backing up " + filename + (message ? ' - ' + message : '')) || this;
        _this.filename = filename;
        _this.name = 'BackupError';
        return _this;
    }
    return BackupError;
}(Error));
exports.BackupError = BackupError;
var FileWriteError = /** @class */ (function (_super) {
    __extends(FileWriteError, _super);
    function FileWriteError(filename, message) {
        var _this = _super.call(this, "Error while trying to write to " + filename + (message && ": " + message)) || this;
        _this.filename = filename;
        return _this;
    }
    return FileWriteError;
}(Error));
exports.FileWriteError = FileWriteError;
