// @sapappgyver/appgyver-formulas/functions/object/REVERSE_LOOKUP_ALL.js
const { getPossibleObjectSchemaProperties, getSchemaAssignmentErrors, resolveObjectMemberAccessor } = require('@sapappgyver/appgyver-schemas');
const execute = require('./REVERSE_LOOKUP_ALL.runtime');

module.exports = {
  name: 'REVERSE_LOOKUP_ALL',
  category: 'Object',
  title: 'Get key of the object by stored value',
  description: 'Given a value and a target object, return a list of all the keys under which the given value can be found.',
  parameters: [{
    type: 'object',
    title: 'Object',
    description: 'Object whose key will be get',
  }, {
    title: 'Value',
    anyOf: [
      { type: 'string' },
      { type: 'number' },
      { type: 'boolean' },
      { type: 'array' },
      { type: 'object' },
    ],
    description: 'Value to find key',
  }],
  analyze: ([objectParam, itemParam], offset, parameterErrors) => {
    const objectSchema = (objectParam && objectParam.schema) || {};
    const itemSchema = (itemParam && itemParam.schema) || {};
    const { schema: keysSchema } = resolveObjectMemberAccessor(
      objectSchema,
      {
        type: 'string',
      },
      true,
    );
    const possibleProps = getPossibleObjectSchemaProperties(objectSchema);
    const assignmentErrors = getSchemaAssignmentErrors(itemSchema, keysSchema);
    const schema = possibleProps
      ? { type: 'array', items: { type: 'string', enum: possibleProps } }
      : { type: 'array', items: { type: 'string' } };
    return {
      schema,
      errors: parameterErrors.length ? [] : [...assignmentErrors].map(error => ({
        ...error, offset: itemParam.offset,
      })),
    };
  },
  execute,
  examples: [
    {
      expression: 'REVERSE_LOOKUP_ALL(animal, 123)',
      context: {
        animal: {
          name: 'dog', weight: 123, active: false, children: 0, height: 123,
        },
      },
      result: ['weight', 'height'],
    },
    {
      expression: 'REVERSE_LOOKUP_ALL(animals, "Richi")',
      context: {
        animals: {
          dog: 'Richi', cat: 'Tom', mouse: 'Jerry', fox: 'Alice', horse: 'Richi',
        },
      },
      result: ['dog', 'horse'],
    },
  ],
};
