// @sapappgyver/appgyver-formulas/functions/utility/DEFAULT.js
const { getUnionSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./DEFAULT.runtime');

module.exports = {
  name: 'DEFAULT',
  category: 'Utility',
  title: 'Returns default value in case the original value is empty',
  description: 'If the given value is non-empty (as per the definitions in the `EMPTY` function), returns the given value, otherwise return the given default.',
  parameters: [{
    title: 'Value',
    description: 'Value to check',
  }, {
    title: 'Default',
    description: 'Default value',
  }],
  analyze: (parameters) => {
    const results = parameters.slice(0, 2);
    const resultSchemas = results.map(({ schema }) => schema);
    return {
      schema: getUnionSchema(resultSchemas),
    };
  },
  execute,
  examples: [
    { expression: 'DEFAULT(true, 1)', result: true },
    { expression: 'DEFAULT(null, 1)', result: 1 },
    { expression: 'DEFAULT({ a: 1 }, 1)', result: { a: 1 } },
    { expression: 'DEFAULT([], 1)', result: [] },
  ],
};
