// @sapappgyver/appgyver-formulas/functions/utility/ENCODE_JSON.js
const execute = require('./ENCODE_JSON.runtime');

module.exports = {
  name: 'ENCODE_JSON',
  category: 'Utility',
  title: 'Convert JSON input to text',
  description: 'Converts a given JSON-format input (number, true/false, list, object) into a text.',
  parameters: [{
    title: 'Value',
    description: 'Value to convert',
    anyOf: [{ type: 'string' }, { type: 'number' }, { type: 'boolean' }, { type: 'object' }, { type: 'array' }],
  }, {
    title: 'Spaces',
    description: 'A Number that\'s used to insert white space into the output JSON text for readability purposes',
    type: 'number',
    default: 0,
  }],
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    { expression: 'ENCODE_JSON("foobar")', result: '"foobar"' },
    { expression: 'ENCODE_JSON([1, 2, null, "123", true])', result: '[1,2,null,"123",true]' },
    { expression: 'ENCODE_JSON({a: 1, b: null, c: "123", d: true})', result: '{"a":1,"b":null,"c":"123","d":true}' },
    { expression: 'ENCODE_JSON([1, 2, null, "123", true], 2)', result: '[\n  1,\n  2,\n  null,\n  "123",\n  true\n]' },
    { expression: 'ENCODE_JSON({a: 1, b: null, c: "123", d: true}, 2)', result: '{\n  "a": 1,\n  "b": null,\n  "c": "123",\n  "d": true\n}' },
  ],
};
