// @sapappgyver/appgyver-formulas/functions/utility/ENCODE_OBJECT_FOR_URL.js
const execute = require('./ENCODE_OBJECT_FOR_URL.runtime');

module.exports = {
  name: 'ENCODE_OBJECT_FOR_URL',
  category: 'Utility',
  title: 'Encodes an object for use in a query string or an `x-www-form-urlencoded` HTTP request body',
  description: 'Encodes an object as `key=value` pairs joined by an ampersand (`&`) character. The keys and values are percent encoded. Only keys with string, number or boolean values are included. This format is typically used in URLs to specify query parameters (after the `?` character) and in HTTP request bodies using the `application/x-www-form-urlencoded` content type.',
  parameters: [{
    title: 'Object',
    description: 'Object to convert',
    type: 'object',
    additionalProperties: {
      anyOf: [
        { type: 'string' },
        { type: 'number' },
        { type: 'boolean' },
      ],
    },
  }],
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    { expression: 'ENCODE_OBJECT_FOR_URL({key: "value"})', result: 'key=value' },
    { expression: 'ENCODE_OBJECT_FOR_URL({sum: 23, currency: "€"})', result: 'sum=23&currency=%E2%82%AC' },
    { expression: 'ENCODE_OBJECT_FOR_URL({redirect_uri: "https://my-domain.tld/oauth?nonce=abc"})', result: 'redirect_uri=https%3A%2F%2Fmy-domain.tld%2Foauth%3Fnonce%3Dabc' },
  ],
};
