// @sapappgyver/appgyver-formulas/functions/utility/IF.js
const { getUnionSchema } = require('@sapappgyver/appgyver-schemas');
const execute = require('./IF.runtime');

module.exports = {
  name: 'IF',
  category: 'Utility',
  title: 'If condition',
  description: 'If the first parameter evaluates to `true`, returns the second parameter. Otherwise, returns the third parameter.',
  parameters: [{
    type: 'boolean',
    title: 'Condition',
    description: 'Value deciding the return value',
  }, {
    title: 'True value',
    description: 'Returned if the first parameter is truthy',
  }, {
    title: 'False value',
    description: 'Returned if the first parameter is falsy',
  }],
  analyze: (parameters) => {
    // Union of the 2nd and 3rd params schemas
    const results = parameters.slice(1, 3);
    const resultSchemas = results.map(({ schema }) => schema);
    const schema = getUnionSchema(resultSchemas);
    return { schema };
  },
  execute,
  examples: [
    { expression: 'IF(true, "yes", "no")', result: 'yes' },
    { expression: 'IF(false, "yes", "no")', result: 'no' },
  ],
};
