// @sapappgyver/appgyver-formulas/functions/utility/NUMERIC_BASE.js
const execute = require('./NUMERIC_BASE.runtime');

module.exports = {
  name: 'NUMERIC_BASE',
  category: 'Utility',
  title: 'Convert number to given base',
  description: 'Converts the given number into a text representation in another base.',
  parameters: [{
    type: 'number',
    title: 'Number',
    description: 'Number to convert',
  }, {
    type: 'number',
    title: 'Base',
    description: 'Base to convert',
  }],
  analyze: () => ({ schema: { type: 'string' } }),
  execute,
  examples: [
    { expression: 'NUMERIC_BASE(123, 2)', result: '1111011' },
    { expression: 'NUMERIC_BASE(123, 4)', result: '1323' },
    { expression: 'NUMERIC_BASE(123, 8)', result: '173' },
    { expression: 'NUMERIC_BASE(123, 10)', result: '123' },
    { expression: 'NUMERIC_BASE(123, 16)', result: '7b' },
    { expression: 'NUMERIC_BASE(123, 2.5)', result: '1111011' },
    { expression: 'NUMERIC_BASE(123.625, 2)', result: '1111011.101' },
    { expression: 'NUMERIC_BASE(-123.625, 2)', result: '-1111011.101' },
  ],
};
